// @loaders.gl/gis/dist/index.js
// Types from `@loaders.gl/schema`
export { getGeometryInfo } from "./lib/geometry-api/geometry-info.js";
export { getBinaryGeometryInfo } from "./lib/binary-geometry-api/binary-geometry-info.js";
export { transformBinaryCoords, transformGeoJsonCoords } from "./lib/binary-geometry-api/transform-coordinates.js";
// TABLE CONVERSION
export { convertGeoArrowToTable } from "./lib/table-converters/convert-geoarrow-table.js";
export { convertWKBTableToGeoJSON } from "./lib/table-converters/convert-wkb-table-to-geojson.js";
// FEATURE COLLECTION CONVERSION
export { convertFlatGeojsonToBinaryFeatureCollection, 
// deprecated
convertFlatGeojsonToBinaryFeatureCollection as flatGeojsonToBinary } from "./lib/feature-collection-converters/convert-flat-geojson-to-binary-features.js";
export { convertGeojsonToBinaryFeatureCollection, 
// deprecated
convertGeojsonToBinaryFeatureCollection as geojsonToBinary } from "./lib/feature-collection-converters/convert-geojson-to-binary-features.js";
export { convertGeojsonToFlatGeojson, 
// deprecated
convertGeojsonToFlatGeojson as geojsonToFlatGeojson } from "./lib/feature-collection-converters/convert-geojson-to-flat-geojson.js";
export { convertBinaryFeatureCollectionToGeojson, convertBinaryFeatureCollectionToGeojson as binaryToGeojson } from "./lib/feature-collection-converters/convert-binary-features-to-geojson.js";
// GEOMETRY ENCODING DETECTION
export { isWKB, isTWKB, isWKT } from "./lib/geometry-converters/wkb/helpers/parse-wkb-header.js";
export { WKT_MAGIC_STRINGS } from "./lib/geometry-converters/wkb/helpers/wkb-types.js";
// GEOMETRY CONVERSION
export { convertBinaryGeometryToGeometry } from "./lib/geometry-converters/convert-binary-geometry-to-geojson.js";
export { convertWKTToGeometry } from "./lib/geometry-converters/wkb/convert-wkt-to-geometry.js";
export { convertWKBToGeometry } from "./lib/geometry-converters/wkb/convert-wkb-to-geometry.js";
export { convertWKBToBinaryGeometry } from "./lib/geometry-converters/wkb/convert-wkb-to-binary-geometry.js";
export { convertTWKBToGeometry } from "./lib/geometry-converters/wkb/convert-twkb-to-geometry.js";
export { convertGeometryToWKT } from "./lib/geometry-converters/wkb/convert-geometry-to-wkt.js";
export { convertGeometryToWKB } from "./lib/geometry-converters/wkb/convert-geometry-to-wkb.js";
export { convertGeometryToTWKB } from "./lib/geometry-converters/wkb/convert-geometry-to-twkb.js";
export { parseWKTCRS } from "./lib//wkt-crs/parse-wkt-crs.js";
export { encodeWKTCRS } from "./lib//wkt-crs/encode-wkt-crs.js";
export { convertGeoArrowToBinaryFeatureCollection, 
// deprecated
convertGeoArrowToBinaryFeatureCollection as getBinaryGeometriesFromArrow, getBinaryGeometryTemplate, getTriangleIndices, getMeanCentersFromBinaryGeometries } from "./lib/feature-collection-converters/convert-geoarrow-to-binary-features.js";
export { convertGeoArrowGeometryToGeoJSON } from "./lib/geometry-converters/convert-geoarrow-to-geojson.js";
// EXPERIMENTAL APIs
export { encodeHex, decodeHex } from "./lib/utils/hex-transcoder.js";
export { extractNumericPropTypes as _extractNumericPropTypes } from "./lib/feature-collection-converters/convert-flat-geojson-to-binary-features.js";
