// @luma.gl/gltf/dist/gltf/create-scenegraph-from-gltf.js
// luma.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { parseGLTF } from "../parsers/parse-gltf.js";
import { parseGLTFLights } from "../parsers/parse-gltf-lights.js";
import { GLTFAnimator } from "./gltf-animator.js";
import { parseGLTFAnimations } from "../parsers/parse-gltf-animations.js";
import { getGLTFExtensionSupport } from "./gltf-extension-support.js";
/** Converts a post-processed glTF asset into luma.gl scenegraph nodes and animation helpers. */
export function createScenegraphsFromGLTF(device, gltf, options) {
    const { scenes, materials, gltfMeshIdToNodeMap, gltfNodeIdToNodeMap, gltfNodeIndexToNodeMap } = parseGLTF(device, gltf, options);
    const animations = parseGLTFAnimations(gltf);
    const animator = new GLTFAnimator({ animations, gltfNodeIdToNodeMap, materials });
    const lights = parseGLTFLights(gltf, { useByteColors: options?.useByteColors ?? true });
    const extensionSupport = getGLTFExtensionSupport(gltf);
    const sceneBounds = scenes.map(scene => getScenegraphBounds(scene.getBounds()));
    const modelBounds = getCombinedScenegraphBounds(sceneBounds);
    return {
        scenes,
        materials,
        animator,
        lights,
        extensionSupport,
        sceneBounds,
        modelBounds,
        gltfMeshIdToNodeMap,
        gltfNodeIdToNodeMap,
        gltfNodeIndexToNodeMap,
        gltf
    };
}
function getScenegraphBounds(bounds) {
    if (!bounds) {
        return {
            bounds: null,
            center: [0, 0, 0],
            size: [0, 0, 0],
            radius: 0.5,
            recommendedOrbitDistance: 1
        };
    }
    const normalizedBounds = [
        [bounds[0][0], bounds[0][1], bounds[0][2]],
        [bounds[1][0], bounds[1][1], bounds[1][2]]
    ];
    const size = [
        normalizedBounds[1][0] - normalizedBounds[0][0],
        normalizedBounds[1][1] - normalizedBounds[0][1],
        normalizedBounds[1][2] - normalizedBounds[0][2]
    ];
    const center = [
        normalizedBounds[0][0] + size[0] * 0.5,
        normalizedBounds[0][1] + size[1] * 0.5,
        normalizedBounds[0][2] + size[2] * 0.5
    ];
    const maxHalfExtent = Math.max(size[0], size[1], size[2]) * 0.5;
    const radius = Math.max(0.5 * Math.hypot(size[0], size[1], size[2]), 0.001);
    return {
        bounds: normalizedBounds,
        center,
        size,
        radius,
        recommendedOrbitDistance: Math.max((Math.max(maxHalfExtent, 0.001) / Math.tan(Math.PI / 6)) * 1.15, radius * 1.1)
    };
}
function getCombinedScenegraphBounds(sceneBounds) {
    let combinedBounds = null;
    for (const sceneBoundInfo of sceneBounds) {
        if (!sceneBoundInfo.bounds) {
            continue;
        }
        if (!combinedBounds) {
            combinedBounds = [
                [...sceneBoundInfo.bounds[0]],
                [...sceneBoundInfo.bounds[1]]
            ];
            continue;
        }
        for (let axis = 0; axis < 3; axis++) {
            combinedBounds[0][axis] = Math.min(combinedBounds[0][axis], sceneBoundInfo.bounds[0][axis]);
            combinedBounds[1][axis] = Math.max(combinedBounds[1][axis], sceneBoundInfo.bounds[1][axis]);
        }
    }
    return getScenegraphBounds(combinedBounds);
}
