// @luma.gl/gltf/dist/gltf/gltf-animator.js
// luma.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { log } from '@luma.gl/core';
import { evaluateSampler, interpolate } from "./animations/interpolate.js";
import { getTextureTransformDeltaMatrix, getTextureTransformSlotDefinition } from "../pbr/texture-transform.js";
/** Evaluates one glTF animation against the generated scenegraph. */
class GLTFSingleAnimator {
    /** Animation definition being played. */
    animation;
    /** Target scenegraph lookup table. */
    gltfNodeIdToNodeMap;
    /** Materials aligned with the source glTF materials array. */
    materials;
    /** Playback start time in seconds. */
    startTime = 0;
    /** Whether playback is currently enabled. */
    playing = true;
    /** Playback speed multiplier. */
    speed = 1;
    /** Mutable runtime texture-transform state for animated material slots. */
    materialTextureTransformState = new Map();
    /** Creates a single-animation controller. */
    constructor(props) {
        this.animation = props.animation;
        this.gltfNodeIdToNodeMap = props.gltfNodeIdToNodeMap;
        this.materials = props.materials || [];
        this.animation.name ||= 'unnamed';
        Object.assign(this, props);
        if (this.animation.channels.some(channel => channel.type !== 'node') &&
            !this.materials.length) {
            throw new Error(`Animation ${this.animation.name} targets materials, but GLTFAnimator was created without a materials array`);
        }
    }
    /** Advances the animation to the supplied wall-clock time in milliseconds. */
    setTime(timeMs) {
        if (!this.playing) {
            return;
        }
        const absTime = timeMs / 1000;
        const time = (absTime - this.startTime) * this.speed;
        this.animation.channels.forEach(channel => {
            if (channel.type === 'node') {
                const { sampler, targetNodeId, path } = channel;
                const targetNode = this.gltfNodeIdToNodeMap.get(targetNodeId);
                if (!targetNode) {
                    throw new Error(`Cannot find animation target node ${targetNodeId}`);
                }
                interpolate(time, sampler, targetNode, path);
                return;
            }
            const material = this.materials[channel.targetMaterialIndex];
            if (!material) {
                throw new Error(`Cannot find animation target material ${channel.targetMaterialIndex} for ${channel.pointer}`);
            }
            const value = evaluateSampler(time, channel.sampler);
            if (value) {
                if (channel.type === 'material') {
                    applyMaterialAnimationValue(material, channel, value);
                }
                else {
                    applyTextureTransformAnimationValue(material, channel, value, this.materialTextureTransformState);
                }
            }
        });
    }
}
/** Coordinates playback of every animation found in a glTF scene. */
export class GLTFAnimator {
    /** Individual animation controllers. */
    animations;
    /** Creates an animator for the supplied glTF scenegraph. */
    constructor(props) {
        this.animations = props.animations.map((animation, index) => {
            const name = animation.name || `Animation-${index}`;
            return new GLTFSingleAnimator({
                gltfNodeIdToNodeMap: props.gltfNodeIdToNodeMap,
                materials: props.materials,
                animation: { name, channels: animation.channels }
            });
        });
    }
    /** @deprecated Use .setTime(). Will be removed (deck.gl is using this) */
    animate(time) {
        log.warn('GLTFAnimator#animate is deprecated. Use GLTFAnimator#setTime instead')();
        this.setTime(time);
    }
    /** Advances every animation to the supplied wall-clock time in milliseconds. */
    setTime(time) {
        this.animations.forEach(animation => animation.setTime(time));
    }
    /** Returns the per-animation controllers managed by this animator. */
    getAnimations() {
        return this.animations;
    }
}
function applyMaterialAnimationValue(material, channel, value) {
    const pbrMaterial = channel.component !== undefined
        ? {
            [channel.property]: updateMaterialArrayComponent(getCurrentMaterialValue(material, channel.property), channel.component, value[0])
        }
        : {
            [channel.property]: value.length === 1 ? value[0] : value
        };
    material.setProps({ pbrMaterial });
}
function getCurrentMaterialValue(material, property) {
    const uniformValues = material.shaderInputs.getUniformValues();
    const currentValue = uniformValues['pbrMaterial']?.[property];
    return Array.isArray(currentValue) ? [...currentValue] : [];
}
function updateMaterialArrayComponent(currentValue, component, nextValue) {
    const updatedValue = [...currentValue];
    updatedValue[component] = nextValue;
    return updatedValue;
}
function applyTextureTransformAnimationValue(material, channel, value, materialTextureTransformState) {
    const slotDefinition = getTextureTransformSlotDefinition(channel.textureSlot);
    const currentTransform = getCurrentTextureTransform(materialTextureTransformState, material, channel);
    switch (channel.path) {
        case 'offset':
            if (channel.component !== undefined) {
                currentTransform.offset[channel.component] = value[0];
            }
            else {
                currentTransform.offset = [value[0], value[1]];
            }
            break;
        case 'rotation':
            currentTransform.rotation = value[0];
            break;
        case 'scale':
            if (channel.component !== undefined) {
                currentTransform.scale[channel.component] = value[0];
            }
            else {
                currentTransform.scale = [value[0], value[1]];
            }
            break;
    }
    material.setProps({
        pbrMaterial: {
            [slotDefinition.uvTransformUniform]: getTextureTransformDeltaMatrix(channel.baseTransform, currentTransform)
        }
    });
}
function getCurrentTextureTransform(materialTextureTransformState, material, channel) {
    const materialState = materialTextureTransformState.get(material) || {};
    let textureTransformState = materialState[channel.textureSlot];
    if (!textureTransformState) {
        textureTransformState = {
            offset: [...channel.baseTransform.offset],
            rotation: channel.baseTransform.rotation,
            scale: [...channel.baseTransform.scale]
        };
        materialState[channel.textureSlot] = textureTransformState;
        materialTextureTransformState.set(material, materialState);
    }
    return textureTransformState;
}
