// @luma.gl/gltf/dist/parsers/parse-gltf-lights.js
import { Matrix4 } from '@math.gl/core';
import { normalizeByteColor3 } from '@luma.gl/shadertools';
/** Parse KHR_lights_punctual extension into luma.gl light definitions */
export function parseGLTFLights(gltf, options = {}) {
    const lightDefs = 
    // `postProcessGLTF()` moves KHR_lights_punctual into `gltf.lights`.
    gltf.lights ||
        gltf.extensions?.['KHR_lights_punctual']?.['lights'];
    if (!lightDefs || !Array.isArray(lightDefs) || lightDefs.length === 0) {
        return [];
    }
    const lights = [];
    const parentNodeById = createParentNodeMap(gltf.nodes || []);
    const worldMatrixByNodeId = new Map();
    for (const node of gltf.nodes || []) {
        const lightIndex = node.light ??
            node.extensions?.KHR_lights_punctual?.light;
        if (typeof lightIndex !== 'number') {
            // eslint-disable-next-line no-continue
            continue;
        }
        const gltfLight = lightDefs[lightIndex];
        if (!gltfLight) {
            // eslint-disable-next-line no-continue
            continue;
        }
        const color = normalizeGLTFLightColor((gltfLight.color || [1, 1, 1]), options.useByteColors ?? true);
        const intensity = gltfLight.intensity ?? 1;
        const range = gltfLight.range;
        const worldMatrix = getNodeWorldMatrix(node, parentNodeById, worldMatrixByNodeId);
        switch (gltfLight.type) {
            case 'directional':
                lights.push(parseDirectionalLight(worldMatrix, color, intensity));
                break;
            case 'point':
                lights.push(parsePointLight(worldMatrix, color, intensity, range));
                break;
            case 'spot':
                lights.push(parseSpotLight(worldMatrix, color, intensity, range, gltfLight.spot));
                break;
            default:
                // Unsupported light type
                break;
        }
    }
    return lights;
}
/**
 * Converts glTF colors from the 0-1 spec range to the configured luma light convention.
 */
function normalizeGLTFLightColor(color, useByteColors) {
    if (useByteColors) {
        return color.map(component => component * 255);
    }
    return normalizeByteColor3(color, false);
}
/**
 * Converts a glTF punctual light attached to a node into a point light.
 */
function parsePointLight(worldMatrix, color, intensity, range) {
    const position = getLightPosition(worldMatrix);
    let attenuation = [1, 0, 0];
    if (range !== undefined && range > 0) {
        attenuation = [1, 0, 1 / (range * range)];
    }
    return {
        type: 'point',
        position,
        color,
        intensity,
        attenuation
    };
}
/**
 * Converts a glTF punctual light attached to a node into a directional light.
 */
function parseDirectionalLight(worldMatrix, color, intensity) {
    const direction = getLightDirection(worldMatrix);
    return {
        type: 'directional',
        direction,
        color,
        intensity
    };
}
/**
 * Converts a glTF punctual light attached to a node into a spot light.
 */
function parseSpotLight(worldMatrix, color, intensity, range, spot = {}) {
    const position = getLightPosition(worldMatrix);
    const direction = getLightDirection(worldMatrix);
    let attenuation = [1, 0, 0];
    if (range !== undefined && range > 0) {
        attenuation = [1, 0, 1 / (range * range)];
    }
    return {
        type: 'spot',
        position,
        direction,
        color,
        intensity,
        attenuation,
        innerConeAngle: spot.innerConeAngle ?? 0,
        outerConeAngle: spot.outerConeAngle ?? Math.PI / 4
    };
}
/**
 * Builds a parent lookup so punctual lights can be resolved in world space.
 */
function createParentNodeMap(nodes) {
    const parentNodeById = new Map();
    for (const node of nodes) {
        for (const childNode of node.children || []) {
            parentNodeById.set(childNode.id, node);
        }
    }
    return parentNodeById;
}
/**
 * Resolves a glTF node's world matrix from its local transform and parent chain.
 */
function getNodeWorldMatrix(node, parentNodeById, worldMatrixByNodeId) {
    const cachedWorldMatrix = worldMatrixByNodeId.get(node.id);
    if (cachedWorldMatrix) {
        return cachedWorldMatrix;
    }
    const localMatrix = getNodeLocalMatrix(node);
    const parentNode = parentNodeById.get(node.id);
    const worldMatrix = parentNode
        ? new Matrix4(getNodeWorldMatrix(parentNode, parentNodeById, worldMatrixByNodeId)).multiplyRight(localMatrix)
        : localMatrix;
    worldMatrixByNodeId.set(node.id, worldMatrix);
    return worldMatrix;
}
/**
 * Resolves a glTF node's local matrix from its matrix or TRS components.
 */
function getNodeLocalMatrix(node) {
    if (node.matrix) {
        return new Matrix4(node.matrix);
    }
    const matrix = new Matrix4();
    if (node.translation) {
        matrix.translate(node.translation);
    }
    if (node.rotation) {
        matrix.multiplyRight(new Matrix4().fromQuaternion(node.rotation));
    }
    if (node.scale) {
        matrix.scale(node.scale);
    }
    return matrix;
}
/**
 * Resolves the world-space position of a glTF light node.
 */
function getLightPosition(worldMatrix) {
    return worldMatrix.transformAsPoint([0, 0, 0]);
}
/**
 * Resolves the world-space forward direction of a glTF light node.
 */
function getLightDirection(worldMatrix) {
    return worldMatrix.transformDirection([0, 0, -1]);
}
