// @luma.gl/core/dist/shadertypes/shader-types/shader-block-layout.js
// luma.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { alignTo } from "../data-types/decode-data-types.js";
import { getVariableShaderTypeInfo, resolveVariableShaderTypeAlias } from "./shader-type-decoder.js";
/**
 * Builds a deterministic shader-block layout from composite shader type declarations.
 *
 * The returned value is pure layout metadata. It records the packed field
 * offsets and exact packed byte length, but it does not allocate buffers or
 * serialize values.
 */
export function makeShaderBlockLayout(uniformTypes, options = {}) {
    const copiedUniformTypes = { ...uniformTypes };
    const layout = options.layout ?? 'std140';
    const fields = {};
    let size = 0;
    for (const [key, uniformType] of Object.entries(copiedUniformTypes)) {
        size = addToLayout(fields, key, uniformType, size, layout);
    }
    size = alignTo(size, getTypeAlignment(copiedUniformTypes, layout));
    return {
        layout,
        byteLength: size * 4,
        uniformTypes: copiedUniformTypes,
        fields
    };
}
/**
 * Returns the layout metadata for a scalar, vector, or matrix leaf type.
 *
 * The result includes both the occupied size in 32-bit words and the alignment
 * requirement that must be applied before placing the value in a shader block.
 */
export function getLeafLayoutInfo(type, layout) {
    const resolvedType = resolveVariableShaderTypeAlias(type);
    const decodedType = getVariableShaderTypeInfo(resolvedType);
    const matrixMatch = /^mat(\d)x(\d)<.+>$/.exec(resolvedType);
    if (matrixMatch) {
        const columns = Number(matrixMatch[1]);
        const rows = Number(matrixMatch[2]);
        const columnInfo = getVectorLayoutInfo(rows, resolvedType, decodedType.type, layout);
        const columnStride = getMatrixColumnStride(columnInfo.size, columnInfo.alignment, layout);
        return {
            alignment: columnInfo.alignment,
            size: columns * columnStride,
            components: columns * rows,
            columns,
            rows,
            columnStride,
            shaderType: resolvedType,
            type: decodedType.type
        };
    }
    const vectorMatch = /^vec(\d)<.+>$/.exec(resolvedType);
    if (vectorMatch) {
        return getVectorLayoutInfo(Number(vectorMatch[1]), resolvedType, decodedType.type, layout);
    }
    return {
        alignment: 1,
        size: 1,
        components: 1,
        columns: 1,
        rows: 1,
        columnStride: 1,
        shaderType: resolvedType,
        type: decodedType.type
    };
}
/**
 * Type guard for composite struct declarations.
 */
export function isCompositeShaderTypeStruct(value) {
    return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}
/**
 * Recursively adds a composite type to the flattened field map.
 *
 * @returns The next free 32-bit-word offset after the inserted type.
 */
function addToLayout(fields, name, type, offset, layout) {
    if (typeof type === 'string') {
        const info = getLeafLayoutInfo(type, layout);
        const alignedOffset = alignTo(offset, info.alignment);
        fields[name] = {
            offset: alignedOffset,
            ...info
        };
        return alignedOffset + info.size;
    }
    if (Array.isArray(type)) {
        if (Array.isArray(type[0])) {
            throw new Error(`Nested arrays are not supported for ${name}`);
        }
        const elementType = type[0];
        const length = type[1];
        const stride = getArrayStride(elementType, layout);
        const arrayOffset = alignTo(offset, getTypeAlignment(type, layout));
        for (let i = 0; i < length; i++) {
            addToLayout(fields, `${name}[${i}]`, elementType, arrayOffset + i * stride, layout);
        }
        return arrayOffset + stride * length;
    }
    if (isCompositeShaderTypeStruct(type)) {
        const structAlignment = getTypeAlignment(type, layout);
        let structOffset = alignTo(offset, structAlignment);
        for (const [memberName, memberType] of Object.entries(type)) {
            structOffset = addToLayout(fields, `${name}.${memberName}`, memberType, structOffset, layout);
        }
        return alignTo(structOffset, structAlignment);
    }
    throw new Error(`Unsupported CompositeShaderType for ${name}`);
}
/**
 * Returns the occupied size of a composite type in 32-bit words.
 */
function getTypeSize(type, layout) {
    if (typeof type === 'string') {
        return getLeafLayoutInfo(type, layout).size;
    }
    if (Array.isArray(type)) {
        const elementType = type[0];
        const length = type[1];
        if (Array.isArray(elementType)) {
            throw new Error('Nested arrays are not supported');
        }
        return getArrayStride(elementType, layout) * length;
    }
    let size = 0;
    for (const memberType of Object.values(type)) {
        const compositeMemberType = memberType;
        size = alignTo(size, getTypeAlignment(compositeMemberType, layout));
        size += getTypeSize(compositeMemberType, layout);
    }
    return alignTo(size, getTypeAlignment(type, layout));
}
/**
 * Returns the required alignment of a composite type in 32-bit words.
 */
function getTypeAlignment(type, layout) {
    if (typeof type === 'string') {
        return getLeafLayoutInfo(type, layout).alignment;
    }
    if (Array.isArray(type)) {
        const elementType = type[0];
        const elementAlignment = getTypeAlignment(elementType, layout);
        return uses16ByteArrayAlignment(layout)
            ? Math.max(elementAlignment, 4)
            : elementAlignment;
    }
    let maxAlignment = 1;
    for (const memberType of Object.values(type)) {
        const memberAlignment = getTypeAlignment(memberType, layout);
        maxAlignment = Math.max(maxAlignment, memberAlignment);
    }
    return uses16ByteStructAlignment(layout)
        ? Math.max(maxAlignment, 4)
        : maxAlignment;
}
/**
 * Returns the layout metadata for a vector leaf type.
 */
function getVectorLayoutInfo(components, shaderType, type, layout) {
    return {
        alignment: components === 2 ? 2 : 4,
        size: components === 3 ? 3 : components,
        components,
        columns: 1,
        rows: components,
        columnStride: components === 3 ? 3 : components,
        shaderType,
        type
    };
}
/**
 * Returns the stride of an array element in 32-bit words.
 *
 * This includes any layout-specific padding between adjacent array elements.
 */
function getArrayStride(elementType, layout) {
    const elementSize = getTypeSize(elementType, layout);
    const elementAlignment = getTypeAlignment(elementType, layout);
    return getArrayLikeStride(elementSize, elementAlignment, layout);
}
/**
 * Returns the common stride rule shared by array-like elements in the target layout.
 */
function getArrayLikeStride(size, alignment, layout) {
    return alignTo(size, uses16ByteArrayAlignment(layout) ? 4 : alignment);
}
/**
 * Returns the stride of a matrix column in 32-bit words.
 */
function getMatrixColumnStride(size, alignment, layout) {
    return layout === 'std140' ? 4 : alignTo(size, alignment);
}
/**
 * Returns `true` when arrays must be rounded up to 16-byte boundaries.
 */
function uses16ByteArrayAlignment(layout) {
    return layout === 'std140' || layout === 'wgsl-uniform';
}
/**
 * Returns `true` when structs must be rounded up to 16-byte boundaries.
 */
function uses16ByteStructAlignment(layout) {
    return layout === 'std140' || layout === 'wgsl-uniform';
}
