// @deck.gl/json/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// dist/index.js
var dist_exports = {};
__export(dist_exports, {
  JSONConfiguration: () => JSONConfiguration,
  JSONConverter: () => JSONConverter,
  Transport: () => Transport,
  _convertFunctions: () => convertFunctions,
  _parseExpressionString: () => parseExpressionString,
  _shallowEqualObjects: () => shallowEqualObjects
});
module.exports = __toCommonJS(dist_exports);

// dist/syntactic-sugar.js
var FUNCTION_IDENTIFIER = "@@=";
var CONSTANT_IDENTIFIER = "@@#";
var TYPE_KEY = "@@type";
var FUNCTION_KEY = "@@function";

// dist/utils/get.js
function get(container, compositeKey) {
  const keyList = getKeys(compositeKey);
  let value = container;
  for (const key of keyList) {
    value = isObject(value) ? value[key] : void 0;
  }
  return value;
}
function isObject(value) {
  return value !== null && typeof value === "object";
}
var keyMap = {};
function getKeys(compositeKey) {
  if (typeof compositeKey === "string") {
    let keyList = keyMap[compositeKey];
    if (!keyList) {
      keyList = compositeKey.split(".");
      keyMap[compositeKey] = keyList;
    }
    return keyList;
  }
  return Array.isArray(compositeKey) ? compositeKey : [compositeKey];
}

// dist/expression-eval/expression-eval.js
var import_jsep = __toESM(require("jsep"), 1);
var binops = {
  "||": (a, b) => {
    return a || b;
  },
  "&&": (a, b) => {
    return a && b;
  },
  "|": (a, b) => {
    return a | b;
  },
  "^": (a, b) => {
    return a ^ b;
  },
  "&": (a, b) => {
    return a & b;
  },
  "==": (a, b) => {
    return a == b;
  },
  "!=": (a, b) => {
    return a != b;
  },
  "===": (a, b) => {
    return a === b;
  },
  "!==": (a, b) => {
    return a !== b;
  },
  "<": (a, b) => {
    return a < b;
  },
  ">": (a, b) => {
    return a > b;
  },
  "<=": (a, b) => {
    return a <= b;
  },
  ">=": (a, b) => {
    return a >= b;
  },
  "<<": (a, b) => {
    return a << b;
  },
  ">>": (a, b) => {
    return a >> b;
  },
  ">>>": (a, b) => {
    return a >>> b;
  },
  "+": (a, b) => {
    return a + b;
  },
  "-": (a, b) => {
    return a - b;
  },
  "*": (a, b) => {
    return a * b;
  },
  "/": (a, b) => {
    return a / b;
  },
  "%": (a, b) => {
    return a % b;
  }
};
var unops = {
  "-": (a) => {
    return -a;
  },
  "+": (a) => {
    return +a;
  },
  "~": (a) => {
    return ~a;
  },
  "!": (a) => {
    return !a;
  }
};
function evaluateArray(list, context) {
  return list.map(function(v) {
    return evaluate(v, context);
  });
}
function evaluateMember(node, context) {
  const object = evaluate(node.object, context);
  let key;
  if (node.computed) {
    key = evaluate(node.property, context);
  } else {
    key = node.property.name;
  }
  if (/^__proto__|prototype|constructor$/.test(key)) {
    throw Error(`Access to member "${key}" disallowed.`);
  }
  return [object, object[key]];
}
function evaluate(_node, context) {
  const node = _node;
  switch (node.type) {
    case "ArrayExpression":
      return evaluateArray(node.elements, context);
    case "BinaryExpression":
      return binops[node.operator](evaluate(node.left, context), evaluate(node.right, context));
    case "CallExpression":
      let caller;
      let fn;
      let assign;
      if (node.callee.type === "MemberExpression") {
        assign = evaluateMember(node.callee, context);
        caller = assign[0];
        fn = assign[1];
      } else {
        fn = evaluate(node.callee, context);
      }
      if (typeof fn !== "function") {
        return void 0;
      }
      return fn.apply(caller, evaluateArray(node.arguments, context));
    case "ConditionalExpression":
      return evaluate(node.test, context) ? evaluate(node.consequent, context) : evaluate(node.alternate, context);
    case "Identifier":
      return context[node.name];
    case "Literal":
      return node.value;
    case "LogicalExpression":
      if (node.operator === "||") {
        return evaluate(node.left, context) || evaluate(node.right, context);
      } else if (node.operator === "&&") {
        return evaluate(node.left, context) && evaluate(node.right, context);
      }
      return binops[node.operator](evaluate(node.left, context), evaluate(node.right, context));
    case "MemberExpression":
      return evaluateMember(node, context)[1];
    case "ThisExpression":
      return context;
    case "UnaryExpression":
      return unops[node.operator](evaluate(node.argument, context));
    default:
      return void 0;
  }
}

// dist/helpers/parse-expression-string.js
var cachedExpressionMap = {
  // Identity function
  "-": (object) => object
};
function parseExpressionString(propValue, configuration) {
  if (propValue in cachedExpressionMap) {
    return cachedExpressionMap[propValue];
  }
  let func;
  const ast = (0, import_jsep.default)(propValue);
  if (ast.type === "Identifier") {
    func = (row) => {
      return get(row, propValue);
    };
  } else {
    traverse(ast, (node) => {
      if (node.type === "CallExpression") {
        throw new Error("Function calls not allowed in JSON expressions");
      }
    });
    func = (row) => {
      return evaluate(ast, row);
    };
  }
  cachedExpressionMap[propValue] = func;
  return func;
}
function traverse(node, visitor) {
  if (Array.isArray(node)) {
    node.forEach((element) => traverse(element, visitor));
  } else if (node && typeof node === "object") {
    if (node.type) {
      visitor(node);
    }
    for (const key in node) {
      traverse(node[key], visitor);
    }
  }
}

// dist/json-configuration.js
var isObject2 = (value) => value && typeof value === "object";
var JSONConfiguration = class {
  /**
   * Creates a configuration from a single plain object.
   * @param configuration Configuration catalogs and hooks to register.
   */
  constructor(configuration) {
    this.config = { ...JSONConfiguration.defaultProps };
    this.convertFunction = parseExpressionString;
    this.preProcessClassProps = (_Class, props) => props;
    this.postProcessConvertedJson = (json) => json;
    this.merge(configuration);
  }
  /**
   * Merges additional configuration catalogs and hooks into the current instance.
   * @param configuration Additional configuration values to merge.
   */
  merge(configuration) {
    for (const key in configuration) {
      if (key in this.config) {
        const value = configuration[key];
        this.config[key] = isObject2(this.config[key]) ? Object.assign(this.config[key], value) : value;
      } else if (key === "convertFunction" && configuration.convertFunction) {
        this.convertFunction = configuration.convertFunction;
      } else if (key === "preProcessClassProps" && configuration.preProcessClassProps) {
        this.preProcessClassProps = configuration.preProcessClassProps;
      } else if (key === "postProcessConvertedJson" && configuration.postProcessConvertedJson) {
        this.postProcessConvertedJson = configuration.postProcessConvertedJson;
      }
    }
  }
  /**
   * Returns a plain-object snapshot of the current configuration, including hooks.
   * @returns A complete configuration object suitable for cloning or re-use.
   */
  getProps() {
    return {
      ...this.config,
      convertFunction: this.convertFunction,
      preProcessClassProps: this.preProcessClassProps,
      postProcessConvertedJson: this.postProcessConvertedJson
    };
  }
};
JSONConfiguration.defaultProps = {
  log: console,
  // eslint-disable-lin,
  typeKey: TYPE_KEY,
  functionKey: FUNCTION_KEY,
  classes: {},
  reactComponents: {},
  enumerations: {},
  constants: {},
  functions: {},
  React: void 0
};

// dist/helpers/convert-functions.js
function hasFunctionIdentifier(value) {
  return typeof value === "string" && value.startsWith(FUNCTION_IDENTIFIER);
}
function trimFunctionIdentifier(value) {
  return value.replace(FUNCTION_IDENTIFIER, "");
}
function convertFunctions(props, configuration) {
  const replacedProps = {};
  for (const propName in props) {
    let propValue = props[propName];
    if (hasFunctionIdentifier(propValue)) {
      const trimmedFunctionIdentifier = trimFunctionIdentifier(propValue);
      propValue = parseExpressionString(trimmedFunctionIdentifier, configuration);
    }
    replacedProps[propName] = propValue;
  }
  return replacedProps;
}

// dist/helpers/instantiate-class.js
function instantiateClass(type, props, configuration) {
  const Class = configuration.config.classes[type];
  const Component = configuration.config.reactComponents[type];
  if (!Class && !Component) {
    const { log } = configuration.config;
    if (log) {
      const stringProps = JSON.stringify(props, null, 0).slice(0, 40);
      log.warn(`JSON converter: No registered class of type ${type}(${stringProps}...)  `);
    }
    return null;
  }
  if (Class) {
    return instantiateJavaScriptClass(Class, props, configuration);
  }
  return instantiateReactComponent(Component, props, configuration);
}
function instantiateJavaScriptClass(Class, props, configuration) {
  if (configuration.preProcessClassProps) {
    props = configuration.preProcessClassProps(Class, props);
  }
  props = convertFunctions(props, configuration);
  return new Class(props);
}
function instantiateReactComponent(Component, props, configuration) {
  const { React } = configuration.config;
  const { children = [] } = props;
  delete props.children;
  if (configuration.preProcessClassProps) {
    props = configuration.preProcessClassProps(Component, props);
  }
  props = convertFunctions(props, configuration);
  return React.createElement(Component, props, children);
}

// dist/helpers/execute-function.js
function executeFunction(targetFunction, props, configuration) {
  const matchedFunction = configuration.config.functions[targetFunction];
  if (!matchedFunction) {
    const { log } = configuration.config;
    if (log) {
      const stringProps = JSON.stringify(props, null, 0).slice(0, 40);
      log.warn(`JSON converter: No registered function ${targetFunction}(${stringProps}...)  `);
    }
    return null;
  }
  return matchedFunction(props);
}

// dist/helpers/parse-json.js
function parseJSON(json) {
  return typeof json === "string" ? JSON.parse(json) : json;
}

// dist/json-converter.js
function isObject3(value) {
  return Boolean(value) && typeof value === "object";
}
var JSONConverter = class {
  /**
   * Creates a converter for a configuration object or `JSONConfiguration` instance.
   * @param props Converter properties.
   */
  constructor(props) {
    this.log = console;
    this.onJSONChange = () => {
    };
    this.json = null;
    this.convertedJson = null;
    this.setProps(props);
  }
  /** Releases resources held by the converter. Present for API symmetry. */
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  finalize() {
  }
  /**
   * Updates the active configuration and callbacks.
   * @param props Converter properties or an already-created configuration instance.
   */
  setProps(props) {
    if ("configuration" in props) {
      this.configuration = props.configuration instanceof JSONConfiguration ? props.configuration : new JSONConfiguration(props.configuration);
    }
    if ("onJSONChange" in props) {
      this.onJSONChange = props.onJSONChange;
    }
  }
  /**
   * Merges additional configuration into the current converter.
   * @param config Additional catalogs or hooks to merge.
   */
  mergeConfiguration(config) {
    this.configuration.merge(config instanceof JSONConfiguration ? config.getProps() : config);
  }
  /**
   * Converts a JSON object or JSON string into runtime props.
   * @param json JSON payload to convert.
   * @returns The converted runtime value.
   */
  convert(json) {
    if (!json || json === this.json) {
      return this.convertedJson;
    }
    this.json = json;
    const parsedJSON = parseJSON(json);
    if (!isObject3(parsedJSON)) {
      throw new Error("JSONConverter: expected an object");
    }
    let convertedJson = convertJSON(parsedJSON, this.configuration);
    convertedJson = this.configuration.postProcessConvertedJson(convertedJson);
    this.convertedJson = convertedJson;
    return convertedJson;
  }
  /**
   * Backwards-compatible alias for `convert()`.
   * @param json JSON payload to convert.
   * @returns The converted runtime value.
   */
  convertJson(json) {
    return this.convert(json);
  }
};
function convertJSON(json, configuration) {
  configuration = new JSONConfiguration(configuration.getProps());
  return convertJSONRecursively(json, "", configuration);
}
function convertJSONRecursively(json, key, configuration) {
  if (Array.isArray(json)) {
    return json.map((element, i) => convertJSONRecursively(element, String(i), configuration));
  }
  if (isObject3(json)) {
    if (isClassInstance(json, configuration)) {
      return convertClassInstance(json, configuration);
    }
    if (FUNCTION_KEY in json) {
      return convertFunctionObject(json, configuration);
    }
    return convertPlainObject(json, configuration);
  }
  if (typeof json === "string") {
    return convertString(json, key, configuration);
  }
  return json;
}
function isClassInstance(json, configuration) {
  const { typeKey } = configuration.config;
  const isClass = isObject3(json) && Boolean(json[typeKey]);
  return isClass;
}
function convertClassInstance(json, configuration) {
  const { typeKey } = configuration.config;
  const type = json[typeKey];
  let props = { ...json };
  delete props[typeKey];
  props = convertPlainObject(props, configuration);
  return instantiateClass(type, props, configuration);
}
function convertFunctionObject(json, configuration) {
  const { functionKey } = configuration.config;
  const targetFunction = json[functionKey];
  let props = { ...json };
  delete props[functionKey];
  props = convertPlainObject(props, configuration);
  return executeFunction(targetFunction, props, configuration);
}
function convertPlainObject(json, configuration) {
  if (!isObject3(json)) {
    throw new Error("convertPlainObject: expected an object");
  }
  const result = {};
  for (const [key, value] of Object.entries(json)) {
    result[key] = convertJSONRecursively(value, key, configuration);
  }
  return result;
}
function convertString(string, key, configuration) {
  if (string.startsWith(FUNCTION_IDENTIFIER) && configuration.convertFunction) {
    string = string.replace(FUNCTION_IDENTIFIER, "");
    return configuration.convertFunction(string, configuration);
  }
  if (string.startsWith(CONSTANT_IDENTIFIER)) {
    string = string.replace(CONSTANT_IDENTIFIER, "");
    if (configuration.config.constants[string]) {
      return configuration.config.constants[string];
    }
    const [enumVarName, enumValName] = string.split(".");
    return configuration.config.enumerations[enumVarName][enumValName];
  }
  return string;
}

// dist/transports/transport.js
var state = {
  onInitialize: (_) => _,
  onFinalize: (_) => _,
  onMessage: (_) => _
};
var Transport = class {
  /**
   * Registers lifecycle callbacks shared by all transport instances.
   * @param callbacks Partial callback set to install.
   */
  static setCallbacks({ onInitialize, onFinalize, onMessage }) {
    if (onInitialize) {
      state.onInitialize = onInitialize;
    }
    if (onFinalize) {
      state.onFinalize = onFinalize;
    }
    if (onMessage) {
      state.onMessage = onMessage;
    }
  }
  /**
   * Creates a transport instance.
   * @param name Human-readable transport name.
   */
  constructor(name = "Transport") {
    this._messageQueue = [];
    this.userData = {};
    this._destroyed = false;
    this.name = name;
  }
  /**
   * Return a root DOM element for this transport connection
   * @returns default implementation returns document.body
   * Jupyter Notebook transports will return an element associated with the notebook cell
   */
  getRootDOMElement() {
    return typeof document !== "undefined" ? document.body : null;
  }
  /**
   * Sends a JSON message through the transport back-channel.
   */
  sendJSONMessage() {
    console.error("Back-channel not implemented for this transport");
  }
  /**
   * Sends a binary message through the transport back-channel.
   */
  sendBinaryMessage() {
    console.error("Back-channel not implemented for this transport");
  }
  //
  // API for transports (not intended for apps)
  //
  /**
   * Marks the transport as initialized and emits the initialize callback.
   * @param options Additional payload fields merged into the callback message.
   */
  _initialize(options = {}) {
    const message = { transport: this, ...options };
    state.onInitialize(message);
  }
  /**
   * Marks the transport as finalized and emits the finalize callback.
   * @param options Additional payload fields merged into the callback message.
   */
  _finalize(options = {}) {
    const message = { transport: this, ...options };
    state.onFinalize(message);
    this._destroyed = true;
  }
  /**
   * Delivers a received message to the registered message callback.
   * @param message Additional payload fields merged into the callback message.
   */
  _messageReceived(message = {}) {
    message = { transport: this, ...message };
    console.debug("Delivering transport message", message);
    state.onMessage(message);
  }
  /*
    // This tries to handle the case that a transport connection initializes before the application
    // has set the callbacks.
    // Note: It is not clear that this can actually happen in the in initial Jupyter widget transport
    _flushQueuedConnections() {
      if (onInitialize) {
        state._initPromise.then(initArgs => {
          onInitialize(initArgs);
  
          if (state._onMessage) {
            // Send any queued messages
            let message;
            while ((message = this._messageQueue.pop())) {
              console.debug('Delivering queued transport message', message); // eslint-disable-line
              this._onMessage(message);
            }
          }
        });
      }
    }
    */
  /**
   * Serializes an object to JSON while tolerating circular references where possible.
   * @param v Value to serialize.
   * @returns A JSON string with cycles removed or deduplicated where possible.
   */
  static _stringifyJSONSafe(v) {
    const cache = /* @__PURE__ */ new Set();
    return JSON.stringify(v, (key, value) => {
      if (typeof value === "object" && value !== null) {
        if (cache.has(value)) {
          try {
            return JSON.parse(JSON.stringify(value));
          } catch (err) {
            return void 0;
          }
        }
        cache.add(value);
      }
      return value;
    });
  }
};

// dist/utils/shallow-equal-objects.js
function shallowEqualObjects(a, b) {
  if (a === b) {
    return true;
  }
  if (typeof a !== "object" || a === null || typeof b !== "object" || b === null) {
    return false;
  }
  if (Object.keys(a).length !== Object.keys(b).length) {
    return false;
  }
  for (const key in a) {
    if (!(key in b) || a[key] !== b[key]) {
      return false;
    }
  }
  for (const key in b) {
    if (!(key in a)) {
      return false;
    }
  }
  return true;
}
