// @deck.gl/json/dist/json-configuration.js
// deck.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { TYPE_KEY, FUNCTION_KEY } from "./syntactic-sugar.js";
// TODO - default parsing code should not be part of the configuration.
import { parseExpressionString } from "./helpers/parse-expression-string.js";
const isObject = value => value && typeof value === 'object';
/**
 * Stores the catalogs and hooks used by `JSONConverter` to resolve JSON into runtime values.
 */
export class JSONConfiguration {
    /**
     * Creates a configuration from a single plain object.
     * @param configuration Configuration catalogs and hooks to register.
     */
    constructor(configuration) {
        /** Normalized configuration catalogs consumed by the conversion helpers. */
        this.config = { ...JSONConfiguration.defaultProps };
        /** Hook used to compile `@@=` accessor strings into executable functions. */
        this.convertFunction = parseExpressionString;
        /** Hook used to rewrite props before class/component instantiation. */
        this.preProcessClassProps = (_Class, props) => props;
        /** Hook used to rewrite the converted JSON result before it is returned. */
        this.postProcessConvertedJson = json => json;
        this.merge(configuration);
    }
    /**
     * Merges additional configuration catalogs and hooks into the current instance.
     * @param configuration Additional configuration values to merge.
     */
    merge(configuration) {
        for (const key in configuration) {
            if (key in this.config) {
                const value = configuration[key];
                this.config[key] = isObject(this.config[key])
                    ? Object.assign(this.config[key], value)
                    : value;
            }
            else if (key === 'convertFunction' && configuration.convertFunction) {
                this.convertFunction = configuration.convertFunction;
            }
            else if (key === 'preProcessClassProps' && configuration.preProcessClassProps) {
                this.preProcessClassProps = configuration.preProcessClassProps;
            }
            else if (key === 'postProcessConvertedJson' && configuration.postProcessConvertedJson) {
                this.postProcessConvertedJson = configuration.postProcessConvertedJson;
            }
        }
    }
    /**
     * Returns a plain-object snapshot of the current configuration, including hooks.
     * @returns A complete configuration object suitable for cloning or re-use.
     */
    getProps() {
        return {
            ...this.config,
            convertFunction: this.convertFunction,
            preProcessClassProps: this.preProcessClassProps,
            postProcessConvertedJson: this.postProcessConvertedJson
        };
    }
}
/** Default values used when a configuration key is omitted. */
JSONConfiguration.defaultProps = {
    log: console, // eslint-disable-lin,
    typeKey: TYPE_KEY,
    functionKey: FUNCTION_KEY,
    classes: {},
    reactComponents: {},
    enumerations: {},
    constants: {},
    functions: {},
    React: undefined
};
