// @deck.gl/extensions/dist/path-style/path-style-extension.js
// deck.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { LayerExtension, _mergeShaders as mergeShaders } from '@deck.gl/core';
import { vec3 } from '@math.gl/core';
import { dashShaders, offsetShaders, scatterplotDashShaders, textBackgroundDashShaders } from "./shaders.glsl.js";
const defaultProps = {
    getDashArray: { type: 'accessor', value: [0, 0] },
    getOffset: { type: 'accessor', value: 0 },
    dashJustified: false,
    dashGapPickable: false
};
/** Adds selected features to the `PathLayer`, `ScatterplotLayer`, `TextBackgroundLayer`, and composite layers that render them. */
class PathStyleExtension extends LayerExtension {
    constructor({ dash = false, offset = false, highPrecisionDash = false } = {}) {
        super({ dash: dash || highPrecisionDash, offset, highPrecisionDash });
    }
    getLayerType(layer) {
        if ('pathTesselator' in layer.state) {
            return 'path';
        }
        const layerName = layer.constructor.layerName;
        if (layerName === 'ScatterplotLayer') {
            return 'scatterplot';
        }
        if (layerName === 'TextBackgroundLayer') {
            return 'textBackground';
        }
        return null;
    }
    isEnabled(layer) {
        return this.getLayerType(layer) !== null;
    }
    getShaders(extension) {
        const layerType = extension.getLayerType(this);
        if (!layerType) {
            return null;
        }
        if (layerType === 'scatterplot' || layerType === 'textBackground') {
            if (!extension.opts.dash) {
                return null;
            }
            const inject = layerType === 'scatterplot'
                ? scatterplotDashShaders.inject
                : textBackgroundDashShaders.inject;
            const pathStyle = {
                name: 'pathStyle',
                inject,
                uniformTypes: {
                    dashGapPickable: 'i32'
                }
            };
            return { modules: [pathStyle] };
        }
        // PathLayer: existing logic
        let result = {};
        const defines = {};
        if (extension.opts.dash) {
            result = mergeShaders(result, dashShaders);
            if (extension.opts.highPrecisionDash) {
                defines.HIGH_PRECISION_DASH = true;
            }
        }
        if (extension.opts.offset) {
            result = mergeShaders(result, offsetShaders);
        }
        const { inject } = result;
        const pathStyle = {
            name: 'pathStyle',
            inject,
            uniformTypes: {
                dashAlignMode: 'f32',
                dashGapPickable: 'i32'
            }
        };
        return {
            modules: [pathStyle],
            defines
        };
    }
    initializeState(context, extension) {
        const attributeManager = this.getAttributeManager();
        const layerType = extension.getLayerType(this);
        if (!attributeManager || !layerType) {
            return;
        }
        if (extension.opts.dash) {
            attributeManager.addInstanced({
                instanceDashArrays: { size: 2, accessor: 'getDashArray' },
                ...(layerType === 'path' && extension.opts.highPrecisionDash
                    ? {
                        instanceDashOffsets: {
                            size: 1,
                            accessor: 'getPath',
                            transform: extension.getDashOffsets.bind(this)
                        }
                    }
                    : {})
            });
        }
        if (layerType === 'path' && extension.opts.offset) {
            attributeManager.addInstanced({
                instanceOffsets: { size: 1, accessor: 'getOffset' }
            });
        }
    }
    updateState(params, extension) {
        if (!extension.isEnabled(this)) {
            return;
        }
        if (extension.opts.dash) {
            const layerType = extension.getLayerType(this);
            if (layerType === 'scatterplot' || layerType === 'textBackground') {
                const pathStyleProps = {
                    dashGapPickable: Boolean(this.props.dashGapPickable)
                };
                this.setShaderModuleProps({ pathStyle: pathStyleProps });
            }
            else {
                const pathStyleProps = {
                    dashAlignMode: this.props.dashJustified ? 1 : 0,
                    dashGapPickable: Boolean(this.props.dashGapPickable)
                };
                this.setShaderModuleProps({ pathStyle: pathStyleProps });
            }
        }
    }
    getDashOffsets(path) {
        const result = [0];
        const positionSize = this.props.positionFormat === 'XY' ? 2 : 3;
        const isNested = Array.isArray(path[0]);
        const geometrySize = isNested ? path.length : path.length / positionSize;
        let p;
        let prevP;
        for (let i = 0; i < geometrySize - 1; i++) {
            p = isNested ? path[i] : path.slice(i * positionSize, i * positionSize + positionSize);
            p = this.projectPosition(p);
            if (i > 0) {
                result[i] = result[i - 1] + vec3.dist(prevP, p);
            }
            prevP = p;
        }
        result[geometrySize - 1] = 0;
        return result;
    }
}
PathStyleExtension.defaultProps = defaultProps;
PathStyleExtension.extensionName = 'PathStyleExtension';
export default PathStyleExtension;
