// @loaders.gl/wms/dist/index.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
export { CSWCapabilitiesLoader } from "./csw-capabilities-loader.js";
export { CSWDomainLoader } from "./csw-domain-loader.js";
export { CSWRecordsLoader } from "./csw-records-loader.js";
// WMS - Web Map Service
export { WMSErrorLoader } from "./wms-error-loader.js";
export { WMSCapabilitiesLoader } from "./wms-capabilities-loader.js";
export { WMSFeatureInfoLoader as _WMSFeatureInfoLoader } from "./wip/wms-feature-info-loader.js";
export { WMSLayerDescriptionLoader as _WMSLayerDescriptionLoader } from "./wip/wms-layer-description-loader.js";
export { WFSCapabilitiesLoader as _WFSCapabilitiesLoader } from "./wfs-capabilities-loader.js";
export { GMLLoader as _GMLLoader } from "./gml-loader.js";
// EXPERIMENTAL: DATA SOURCES
// OGC Services
// export {CSWSource} from './csw-source';
export { WMSSource, WMSImageSource } from "./wms-source.js";
// ArcGIS SourceLoaders
export { getArcGISServices as _getArcGISServices } from "./arcgis/arcgis-server.js";
export { ArcGISImageServerSource as _ArcGISImageServerSource } from "./arcgis/arcgis-image-source.js";
export { ImageSource } from '@loaders.gl/loader-utils';
export { createImageSource } from "./lib/deprecated/create-image-source.js";
