// @luma.gl/core/dist/index.js
// luma.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
export { luma } from "./adapter/luma.js";
// ADAPTER (DEVICE AND GPU RESOURCE INTERFACES)
export { Adapter } from "./adapter/adapter.js";
export { Device, DeviceFeatures, DeviceLimits } from "./adapter/device.js";
export { CanvasContext } from "./adapter/canvas-context.js";
export { PresentationContext } from "./adapter/presentation-context.js";
// GPU RESOURCES
export { Resource } from "./adapter/resources/resource.js";
export { Buffer } from "./adapter/resources/buffer.js";
export { Texture } from "./adapter/resources/texture.js";
export { TextureView } from "./adapter/resources/texture-view.js";
export { ExternalTexture } from "./adapter/resources/external-texture.js";
export { Shader } from "./adapter/resources/shader.js";
export { Sampler } from "./adapter/resources/sampler.js";
export { Framebuffer } from "./adapter/resources/framebuffer.js";
export { RenderPipeline } from "./adapter/resources/render-pipeline.js";
export { SharedRenderPipeline } from "./adapter/resources/shared-render-pipeline.js";
export { PipelineFactory } from "./factories/pipeline-factory.js";
export { ShaderFactory } from "./factories/shader-factory.js";
export { _getDefaultBindGroupFactory } from "./factories/bind-group-factory.js";
export { RenderPass } from "./adapter/resources/render-pass.js";
export { ComputePipeline } from "./adapter/resources/compute-pipeline.js";
export { ComputePass } from "./adapter/resources/compute-pass.js";
export { CommandEncoder } from "./adapter/resources/command-encoder.js";
export { CommandBuffer } from "./adapter/resources/command-buffer.js";
export { VertexArray } from "./adapter/resources/vertex-array.js";
export { TransformFeedback } from "./adapter/resources/transform-feedback.js";
export { QuerySet } from "./adapter/resources/query-set.js";
export { Fence } from "./adapter/resources/fence.js";
export { PipelineLayout } from "./adapter/resources/pipeline-layout.js";
// PORTABLE API - UNIFORM BUFFERS
export { makeShaderBlockLayout } from "./shadertypes/shader-types/shader-block-layout.js";
export { ShaderBlockWriter } from "./portable/shader-block-writer.js";
export { UniformBlock } from "./portable/uniform-block.js";
export { UniformStore } from "./portable/uniform-store.js";
export { dataTypeDecoder } from "./shadertypes/data-types/data-type-decoder.js";
export { getTypedArrayConstructor } from "./shadertypes/data-types/decode-data-types.js";
export { shaderTypeDecoder, getAttributeShaderTypeInfo, getVariableShaderTypeInfo } from "./shadertypes/shader-types/shader-type-decoder.js";
export { vertexFormatDecoder } from "./shadertypes/vertex-types/vertex-format-decoder.js";
export { textureFormatDecoder } from "./shadertypes/texture-types/texture-format-decoder.js";
export { getTextureImageView, setTextureImageData } from "./shadertypes/texture-types/texture-layout.js";
export { readPixel, writePixel } from "./shadertypes/texture-types/pixel-utils.js";
export { isExternalImage, getExternalImageSize } from "./shadertypes/image-types/image-types.js";
// INTERNAL UTILS - for use in other luma.gl modules only
export { log } from "./utils/log.js";
export { getShaderLayoutBinding, normalizeBindingsByGroup, flattenBindingsByGroup } from "./adapter-utils/bind-groups.js";
export { assert, assertDefined } from "./utils/assert.js";
export { getScratchArray } from "./utils/array-utils-flat.js";
export { getAttributeInfosFromLayouts } from "./adapter-utils/get-attribute-from-layouts.js";
// TEST EXPORTS
export { getTextureFormatDefinition as _getTextureFormatDefinition, getTextureFormatTable as _getTextureFormatTable } from "./shadertypes/texture-types/texture-format-table.js";
