// @luma.gl/gltf/dist/pbr/pbr-environment.js
// luma.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { DynamicTexture } from '@luma.gl/engine';
import { loadImageTexture } from '@loaders.gl/textures';
/** Loads textures for PBR environment */
export function loadPBREnvironment(device, props) {
    const specularMipLevels = props.specularMipLevels ?? 1;
    const brdfLutTexture = new DynamicTexture(device, {
        id: 'brdfLUT',
        sampler: {
            addressModeU: 'clamp-to-edge',
            addressModeV: 'clamp-to-edge',
            minFilter: 'linear',
            magFilter: 'linear'
        },
        // Texture accepts a promise that returns an image as data (Async Textures)
        data: loadImageTexture(resolveTextureUrl(props.brdfLutUrl))
    });
    const diffuseEnvSampler = makeCube(device, {
        id: 'DiffuseEnvSampler',
        getTextureForFace: face => loadImageTexture(resolveTextureUrl(props.getTexUrl('diffuse', FACES.indexOf(face), 0))),
        sampler: {
            addressModeU: 'clamp-to-edge',
            addressModeV: 'clamp-to-edge',
            minFilter: 'linear',
            magFilter: 'linear'
        }
    });
    const specularEnvSampler = makeCube(device, {
        id: 'SpecularEnvSampler',
        getTextureForFace: (face) => {
            const imageArray = [];
            const direction = FACES.indexOf(face);
            for (let lod = 0; lod < specularMipLevels; lod++) {
                imageArray.push(loadImageTexture(resolveTextureUrl(props.getTexUrl('specular', direction, lod))));
            }
            return Promise.all(imageArray);
        },
        sampler: {
            addressModeU: 'clamp-to-edge',
            addressModeV: 'clamp-to-edge',
            minFilter: 'linear', // [GL.TEXTURE_MIN_FILTER]: GL.LINEAR_MIPMAP_LINEAR,
            magFilter: 'linear'
        }
    });
    return {
        brdfLutTexture,
        diffuseEnvSampler,
        specularEnvSampler
    };
}
// TODO put somewhere common
const FACES = ['+X', '-X', '+Y', '-Y', '+Z', '-Z'];
function resolveTextureUrl(url) {
    const baseUrl = globalThis.document?.baseURI ?? globalThis.location?.href;
    return baseUrl ? new URL(url, baseUrl).toString() : url;
}
/** Construction props for an asynchronously loaded cubemap. */
function makeCube(device, { id, getTextureForFace, sampler }) {
    const data = Promise.all(FACES.map(face => getTextureForFace(face))).then(faceDataArray => {
        const cubeData = {};
        FACES.forEach((face, index) => {
            cubeData[face] = faceDataArray[index];
        });
        return cubeData;
    });
    return new DynamicTexture(device, {
        id,
        dimension: 'cube',
        mipmaps: false,
        sampler,
        data
    });
}
