// @loaders.gl/textures/dist/dist.dev.js
(function webpackUniversalModuleDefinition(root, factory) {
  if (typeof exports === 'object' && typeof module === 'object')
    module.exports = factory();
  else if (typeof define === 'function' && define.amd) define([], factory);
        else if (typeof exports === 'object') exports['loaders'] = factory();
  else root['loaders'] = factory();})(globalThis, function () {
"use strict";
var __exports__ = (() => {
  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __commonJS = (cb, mod) => function __require() {
    return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
  };
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
  var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
    mod
  ));
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // external-global-plugin:@loaders.gl/core
  var require_core = __commonJS({
    "external-global-plugin:@loaders.gl/core"(exports, module) {
      module.exports = globalThis.loaders;
    }
  });

  // bundle.ts
  var bundle_exports = {};
  __export(bundle_exports, {
    BASIS_EXTERNAL_LIBRARIES: () => BASIS_EXTERNAL_LIBRARIES,
    BasisLoader: () => BasisLoader,
    BasisWorkerLoader: () => BasisWorkerLoader,
    CRUNCH_EXTERNAL_LIBRARIES: () => CRUNCH_EXTERNAL_LIBRARIES,
    CompressedTextureLoader: () => CompressedTextureLoader,
    CompressedTextureWorkerLoader: () => CompressedTextureWorkerLoader,
    CompressedTextureWriter: () => CompressedTextureWriter,
    CrunchLoader: () => CrunchLoader,
    CrunchWorkerLoader: () => CrunchLoader,
    GL_COMPRESSED_R11_EAC: () => GL_COMPRESSED_R11_EAC,
    GL_COMPRESSED_RED_GREEN_RGTC2_EXT: () => GL_COMPRESSED_RED_GREEN_RGTC2_EXT,
    GL_COMPRESSED_RED_RGTC1_EXT: () => GL_COMPRESSED_RED_RGTC1_EXT,
    GL_COMPRESSED_RG11_EAC: () => GL_COMPRESSED_RG11_EAC,
    GL_COMPRESSED_RGB8_ETC2: () => GL_COMPRESSED_RGB8_ETC2,
    GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2: () => GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2,
    GL_COMPRESSED_RGBA8_ETC2_EAC: () => GL_COMPRESSED_RGBA8_ETC2_EAC,
    GL_COMPRESSED_RGBA_ASTC_10x10_KHR: () => GL_COMPRESSED_RGBA_ASTC_10x10_KHR,
    GL_COMPRESSED_RGBA_ASTC_10x5_KHR: () => GL_COMPRESSED_RGBA_ASTC_10x5_KHR,
    GL_COMPRESSED_RGBA_ASTC_10x6_KHR: () => GL_COMPRESSED_RGBA_ASTC_10x6_KHR,
    GL_COMPRESSED_RGBA_ASTC_10x8_KHR: () => GL_COMPRESSED_RGBA_ASTC_10x8_KHR,
    GL_COMPRESSED_RGBA_ASTC_12x10_KHR: () => GL_COMPRESSED_RGBA_ASTC_12x10_KHR,
    GL_COMPRESSED_RGBA_ASTC_12x12_KHR: () => GL_COMPRESSED_RGBA_ASTC_12x12_KHR,
    GL_COMPRESSED_RGBA_ASTC_4x4_KHR: () => GL_COMPRESSED_RGBA_ASTC_4x4_KHR,
    GL_COMPRESSED_RGBA_ASTC_5x4_KHR: () => GL_COMPRESSED_RGBA_ASTC_5x4_KHR,
    GL_COMPRESSED_RGBA_ASTC_5x5_KHR: () => GL_COMPRESSED_RGBA_ASTC_5x5_KHR,
    GL_COMPRESSED_RGBA_ASTC_6x5_KHR: () => GL_COMPRESSED_RGBA_ASTC_6x5_KHR,
    GL_COMPRESSED_RGBA_ASTC_6x6_KHR: () => GL_COMPRESSED_RGBA_ASTC_6x6_KHR,
    GL_COMPRESSED_RGBA_ASTC_8x5_KHR: () => GL_COMPRESSED_RGBA_ASTC_8x5_KHR,
    GL_COMPRESSED_RGBA_ASTC_8x6_KHR: () => GL_COMPRESSED_RGBA_ASTC_8x6_KHR,
    GL_COMPRESSED_RGBA_ASTC_8x8_KHR: () => GL_COMPRESSED_RGBA_ASTC_8x8_KHR,
    GL_COMPRESSED_RGBA_ATC_EXPLICIT_ALPHA_WEBGL: () => GL_COMPRESSED_RGBA_ATC_EXPLICIT_ALPHA_WEBGL,
    GL_COMPRESSED_RGBA_ATC_INTERPOLATED_ALPHA_WEBGL: () => GL_COMPRESSED_RGBA_ATC_INTERPOLATED_ALPHA_WEBGL,
    GL_COMPRESSED_RGBA_BPTC_UNORM_EXT: () => GL_COMPRESSED_RGBA_BPTC_UNORM_EXT,
    GL_COMPRESSED_RGBA_PVRTC_2BPPV1_IMG: () => GL_COMPRESSED_RGBA_PVRTC_2BPPV1_IMG,
    GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG: () => GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG,
    GL_COMPRESSED_RGBA_S3TC_DXT1_EXT: () => GL_COMPRESSED_RGBA_S3TC_DXT1_EXT,
    GL_COMPRESSED_RGBA_S3TC_DXT3_EXT: () => GL_COMPRESSED_RGBA_S3TC_DXT3_EXT,
    GL_COMPRESSED_RGBA_S3TC_DXT5_EXT: () => GL_COMPRESSED_RGBA_S3TC_DXT5_EXT,
    GL_COMPRESSED_RGB_ATC_WEBGL: () => GL_COMPRESSED_RGB_ATC_WEBGL,
    GL_COMPRESSED_RGB_BPTC_SIGNED_FLOAT_EXT: () => GL_COMPRESSED_RGB_BPTC_SIGNED_FLOAT_EXT,
    GL_COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT_EXT: () => GL_COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT_EXT,
    GL_COMPRESSED_RGB_ETC1_WEBGL: () => GL_COMPRESSED_RGB_ETC1_WEBGL,
    GL_COMPRESSED_RGB_PVRTC_2BPPV1_IMG: () => GL_COMPRESSED_RGB_PVRTC_2BPPV1_IMG,
    GL_COMPRESSED_RGB_PVRTC_4BPPV1_IMG: () => GL_COMPRESSED_RGB_PVRTC_4BPPV1_IMG,
    GL_COMPRESSED_RGB_S3TC_DXT1_EXT: () => GL_COMPRESSED_RGB_S3TC_DXT1_EXT,
    GL_COMPRESSED_SIGNED_R11_EAC: () => GL_COMPRESSED_SIGNED_R11_EAC,
    GL_COMPRESSED_SIGNED_RED_GREEN_RGTC2_EXT: () => GL_COMPRESSED_SIGNED_RED_GREEN_RGTC2_EXT,
    GL_COMPRESSED_SIGNED_RED_RGTC1_EXT: () => GL_COMPRESSED_SIGNED_RED_RGTC1_EXT,
    GL_COMPRESSED_SIGNED_RG11_EAC: () => GL_COMPRESSED_SIGNED_RG11_EAC,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR: () => GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR,
    GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC: () => GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC,
    GL_COMPRESSED_SRGB8_ETC2: () => GL_COMPRESSED_SRGB8_ETC2,
    GL_COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2: () => GL_COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2,
    GL_COMPRESSED_SRGB_ALPHA_BPTC_UNORM_EXT: () => GL_COMPRESSED_SRGB_ALPHA_BPTC_UNORM_EXT,
    GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT: () => GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT,
    GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT: () => GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT,
    GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT: () => GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT,
    GL_COMPRESSED_SRGB_S3TC_DXT1_EXT: () => GL_COMPRESSED_SRGB_S3TC_DXT1_EXT,
    GL_EXTENSIONS_CONSTANTS: () => GL_EXTENSIONS_CONSTANTS,
    GL_RGB: () => GL_RGB,
    GL_RGB565: () => GL_RGB565,
    GL_RGB5_A1: () => GL_RGB5_A1,
    GL_RGBA: () => GL_RGBA,
    GL_RGBA32F: () => GL_RGBA32F,
    GL_RGBA4: () => GL_RGBA4,
    GL_RGBA8: () => GL_RGBA8,
    KTX2BasisWriter: () => KTX2BasisWriter,
    KTX2BasisWriterWorker: () => KTX2BasisWriterWorker,
    NPYLoader: () => NPYLoader,
    NPYWorkerLoader: () => NPYWorkerLoader,
    RadianceHDRLoader: () => RadianceHDRLoader,
    TextureArrayLoader: () => TextureArrayLoader,
    TextureCubeArrayLoader: () => TextureCubeArrayLoader,
    TextureCubeLoader: () => TextureCubeLoader,
    TextureLoader: () => TextureLoader,
    loadImageTexture: () => loadImageTexture,
    loadImageTextureArray: () => loadImageTextureArray,
    loadImageTextureCube: () => loadImageTextureCube
  });
  __reExport(bundle_exports, __toESM(require_core(), 1));

  // src/lib/utils/version.ts
  var VERSION = typeof __VERSION__ !== "undefined" ? __VERSION__ : "latest";

  // ../worker-utils/src/lib/npm-tag.ts
  var NPM_TAG = "latest";

  // ../worker-utils/src/lib/env-utils/version.ts
  var warningIssued = false;
  function getVersion() {
    if (!globalThis._loadersgl_?.version) {
      globalThis._loadersgl_ = globalThis._loadersgl_ || {};
      if (typeof __VERSION__ === "undefined" && !warningIssued) {
        console.warn(
          "loaders.gl: The __VERSION__ variable is not injected using babel plugin. Latest unstable workers would be fetched from the CDN."
        );
        globalThis._loadersgl_.version = NPM_TAG;
        warningIssued = true;
      } else {
        globalThis._loadersgl_.version = __VERSION__;
      }
    }
    return globalThis._loadersgl_.version;
  }
  var VERSION2 = getVersion();

  // ../worker-utils/src/lib/env-utils/assert.ts
  function assert(condition, message) {
    if (!condition) {
      throw new Error(message || "loaders.gl assertion failed.");
    }
  }

  // ../worker-utils/src/lib/env-utils/globals.ts
  var globals = {
    self: typeof self !== "undefined" && self,
    window: typeof window !== "undefined" && window,
    global: typeof global !== "undefined" && global,
    document: typeof document !== "undefined" && document
  };
  var self_ = globals.self || globals.window || globals.global || {};
  var window_ = globals.window || globals.self || globals.global || {};
  var global_ = globals.global || globals.self || globals.window || {};
  var document_ = globals.document || {};
  var isBrowser = (
    // @ts-ignore process.browser
    typeof process !== "object" || String(process) !== "[object process]" || process.browser
  );
  var isWorker = typeof importScripts === "function";
  var isMobile = typeof window !== "undefined" && typeof window.orientation !== "undefined";
  var matches = typeof process !== "undefined" && process.version && /v([0-9]*)/.exec(process.version);
  var nodeVersion = matches && parseFloat(matches[1]) || 0;

  // ../worker-utils/src/lib/library-utils/library-utils.ts
  var loadLibraryPromises = {};
  function extractLoadLibraryOptions(options = {}) {
    const useLocalLibraries = options.useLocalLibraries ?? options.core?.useLocalLibraries;
    const CDN = options.CDN ?? options.core?.CDN;
    const modules = options.modules;
    return {
      ...useLocalLibraries !== void 0 ? { useLocalLibraries } : {},
      ...CDN !== void 0 ? { CDN } : {},
      ...modules !== void 0 ? { modules } : {}
    };
  }
  async function loadLibrary(libraryUrl, moduleName = null, options = {}, libraryName = null) {
    if (moduleName) {
      libraryUrl = getLibraryUrl(libraryUrl, moduleName, options, libraryName);
    }
    loadLibraryPromises[libraryUrl] = // eslint-disable-next-line @typescript-eslint/no-misused-promises
    loadLibraryPromises[libraryUrl] || loadLibraryFromFile(libraryUrl);
    return await loadLibraryPromises[libraryUrl];
  }
  function getLibraryUrl(library, moduleName, options = {}, libraryName = null) {
    if (options?.core) {
      throw new Error("loadLibrary: options.core must be pre-normalized");
    }
    if (!options.useLocalLibraries && library.startsWith("http")) {
      return library;
    }
    libraryName = libraryName || library;
    const modules = options.modules || {};
    if (modules[libraryName]) {
      return modules[libraryName];
    }
    if (!isBrowser) {
      return `modules/${moduleName}/dist/libs/${libraryName}`;
    }
    if (options.CDN) {
      assert(options.CDN.startsWith("http"));
      return `${options.CDN}/${moduleName}@${VERSION2}/dist/libs/${libraryName}`;
    }
    if (isWorker) {
      return `../src/libs/${libraryName}`;
    }
    return `modules/${moduleName}/src/libs/${libraryName}`;
  }
  async function loadLibraryFromFile(libraryUrl) {
    if (libraryUrl.endsWith("wasm")) {
      return await loadAsArrayBuffer(libraryUrl);
    }
    if (!isBrowser) {
      const { requireFromFile } = globalThis.loaders || {};
      try {
        const result = await requireFromFile?.(libraryUrl);
        if (result || !libraryUrl.includes("/dist/libs/")) {
          return result;
        }
        return await requireFromFile?.(libraryUrl.replace("/dist/libs/", "/src/libs/"));
      } catch (error) {
        if (libraryUrl.includes("/dist/libs/")) {
          try {
            return await requireFromFile?.(libraryUrl.replace("/dist/libs/", "/src/libs/"));
          } catch {
          }
        }
        console.error(error);
        return null;
      }
    }
    if (isWorker) {
      return importScripts(libraryUrl);
    }
    const scriptSource = await loadAsText(libraryUrl);
    return loadLibraryFromString(scriptSource, libraryUrl);
  }
  function loadLibraryFromString(scriptSource, id) {
    if (!isBrowser) {
      const { requireFromString } = globalThis.loaders || {};
      return requireFromString?.(scriptSource, id);
    }
    if (isWorker) {
      eval.call(globalThis, scriptSource);
      return null;
    }
    const script = document.createElement("script");
    script.id = id;
    try {
      script.appendChild(document.createTextNode(scriptSource));
    } catch (e) {
      script.text = scriptSource;
    }
    document.body.appendChild(script);
    return null;
  }
  async function loadAsArrayBuffer(url) {
    const { readFileAsArrayBuffer } = globalThis.loaders || {};
    if (isBrowser || !readFileAsArrayBuffer || url.startsWith("http")) {
      const response = await fetch(url);
      return await response.arrayBuffer();
    }
    try {
      return await readFileAsArrayBuffer(url);
    } catch {
      if (url.includes("/dist/libs/")) {
        return await readFileAsArrayBuffer(url.replace("/dist/libs/", "/src/libs/"));
      }
      throw new Error(`Failed to load ArrayBuffer from ${url}`);
    }
  }
  async function loadAsText(url) {
    const { readFileAsText } = globalThis.loaders || {};
    if (isBrowser || !readFileAsText || url.startsWith("http")) {
      const response = await fetch(url);
      return await response.text();
    }
    try {
      return await readFileAsText(url);
    } catch {
      if (url.includes("/dist/libs/")) {
        return await readFileAsText(url.replace("/dist/libs/", "/src/libs/"));
      }
      throw new Error(`Failed to load text from ${url}`);
    }
  }

  // ../worker-utils/src/lib/process-utils/child-process-proxy.browser.ts
  var ChildProcessProxy = class {
    constructor() {
      throw new Error("ChildProcessProxy is not available in browser environments");
    }
    async start(props) {
      return await Promise.resolve({});
    }
    async stop() {
      return await Promise.resolve();
    }
    async exit() {
      return await Promise.resolve();
    }
  };

  // ../loader-utils/src/loader-types.ts
  async function parseFromContext(data, loaders, options, context) {
    return context._parse(data, loaders, options, context);
  }

  // ../loader-utils/src/lib/env-utils/assert.ts
  function assert2(condition, message) {
    if (!condition) {
      throw new Error(message || "loader assertion failed.");
    }
  }

  // ../loader-utils/src/lib/env-utils/globals.ts
  var globals2 = {
    self: typeof self !== "undefined" && self,
    window: typeof window !== "undefined" && window,
    global: typeof global !== "undefined" && global,
    document: typeof document !== "undefined" && document
  };
  var self_2 = globals2.self || globals2.window || globals2.global || {};
  var window_2 = globals2.window || globals2.self || globals2.global || {};
  var global_2 = globals2.global || globals2.self || globals2.window || {};
  var document_2 = globals2.document || {};
  var isBrowser2 = (
    // @ts-ignore process does not exist on browser
    Boolean(typeof process !== "object" || String(process) !== "[object process]" || process.browser)
  );
  var matches2 = typeof process !== "undefined" && process.version && /v([0-9]*)/.exec(process.version);
  var nodeVersion2 = matches2 && parseFloat(matches2[1]) || 0;

  // ../../node_modules/@probe.gl/env/dist/lib/globals.js
  var window_3 = globalThis;
  var document_3 = globalThis.document || {};
  var process_ = globalThis.process || {};
  var console_ = globalThis.console;
  var navigator_ = globalThis.navigator || {};

  // ../../node_modules/@probe.gl/env/dist/lib/is-electron.js
  function isElectron(mockUserAgent) {
    if (typeof window !== "undefined" && window.process?.type === "renderer") {
      return true;
    }
    if (typeof process !== "undefined" && Boolean(process.versions?.["electron"])) {
      return true;
    }
    const realUserAgent = typeof navigator !== "undefined" && navigator.userAgent;
    const userAgent = mockUserAgent || realUserAgent;
    return Boolean(userAgent && userAgent.indexOf("Electron") >= 0);
  }

  // ../../node_modules/@probe.gl/env/dist/lib/is-browser.js
  function isBrowser3() {
    const isNode = (
      // @ts-expect-error
      typeof process === "object" && String(process) === "[object process]" && !process?.browser
    );
    return !isNode || isElectron();
  }

  // ../../node_modules/@probe.gl/env/dist/index.js
  var VERSION3 = true ? "4.1.1" : "untranspiled source";

  // ../../node_modules/@probe.gl/log/dist/utils/assert.js
  function assert3(condition, message) {
    if (!condition) {
      throw new Error(message || "Assertion failed");
    }
  }

  // ../../node_modules/@probe.gl/log/dist/loggers/log-utils.js
  function normalizeLogLevel(logLevel) {
    if (!logLevel) {
      return 0;
    }
    let resolvedLevel;
    switch (typeof logLevel) {
      case "number":
        resolvedLevel = logLevel;
        break;
      case "object":
        resolvedLevel = logLevel.logLevel || logLevel.priority || 0;
        break;
      default:
        return 0;
    }
    assert3(Number.isFinite(resolvedLevel) && resolvedLevel >= 0);
    return resolvedLevel;
  }
  function normalizeArguments(opts) {
    const { logLevel, message } = opts;
    opts.logLevel = normalizeLogLevel(logLevel);
    const args = opts.args ? Array.from(opts.args) : [];
    while (args.length && args.shift() !== message) {
    }
    switch (typeof logLevel) {
      case "string":
      case "function":
        if (message !== void 0) {
          args.unshift(message);
        }
        opts.message = logLevel;
        break;
      case "object":
        Object.assign(opts, logLevel);
        break;
      default:
    }
    if (typeof opts.message === "function") {
      opts.message = opts.message();
    }
    const messageType = typeof opts.message;
    assert3(messageType === "string" || messageType === "object");
    return Object.assign(opts, { args }, opts.opts);
  }

  // ../../node_modules/@probe.gl/log/dist/loggers/base-log.js
  var noop = () => {
  };
  var BaseLog = class {
    constructor({ level = 0 } = {}) {
      this.userData = {};
      this._onceCache = /* @__PURE__ */ new Set();
      this._level = level;
    }
    set level(newLevel) {
      this.setLevel(newLevel);
    }
    get level() {
      return this.getLevel();
    }
    setLevel(level) {
      this._level = level;
      return this;
    }
    getLevel() {
      return this._level;
    }
    // Unconditional logging
    warn(message, ...args) {
      return this._log("warn", 0, message, args, { once: true });
    }
    error(message, ...args) {
      return this._log("error", 0, message, args);
    }
    // Conditional logging
    log(logLevel, message, ...args) {
      return this._log("log", logLevel, message, args);
    }
    info(logLevel, message, ...args) {
      return this._log("info", logLevel, message, args);
    }
    once(logLevel, message, ...args) {
      return this._log("once", logLevel, message, args, { once: true });
    }
    _log(type, logLevel, message, args, options = {}) {
      const normalized = normalizeArguments({
        logLevel,
        message,
        args: this._buildArgs(logLevel, message, args),
        opts: options
      });
      return this._createLogFunction(type, normalized, options);
    }
    _buildArgs(logLevel, message, args) {
      return [logLevel, message, ...args];
    }
    _createLogFunction(type, normalized, options) {
      if (!this._shouldLog(normalized.logLevel)) {
        return noop;
      }
      const tag = this._getOnceTag(options.tag ?? normalized.tag ?? normalized.message);
      if ((options.once || normalized.once) && tag !== void 0) {
        if (this._onceCache.has(tag)) {
          return noop;
        }
        this._onceCache.add(tag);
      }
      return this._emit(type, normalized);
    }
    _shouldLog(logLevel) {
      return this.getLevel() >= normalizeLogLevel(logLevel);
    }
    _getOnceTag(tag) {
      if (tag === void 0) {
        return void 0;
      }
      try {
        return typeof tag === "string" ? tag : String(tag);
      } catch {
        return void 0;
      }
    }
  };

  // ../../node_modules/@probe.gl/log/dist/utils/local-storage.js
  function getStorage(type) {
    try {
      const storage = window[type];
      const x = "__storage_test__";
      storage.setItem(x, x);
      storage.removeItem(x);
      return storage;
    } catch (e) {
      return null;
    }
  }
  var LocalStorage = class {
    constructor(id, defaultConfig, type = "sessionStorage") {
      this.storage = getStorage(type);
      this.id = id;
      this.config = defaultConfig;
      this._loadConfiguration();
    }
    getConfiguration() {
      return this.config;
    }
    setConfiguration(configuration) {
      Object.assign(this.config, configuration);
      if (this.storage) {
        const serialized = JSON.stringify(this.config);
        this.storage.setItem(this.id, serialized);
      }
    }
    // Get config from persistent store, if available
    _loadConfiguration() {
      let configuration = {};
      if (this.storage) {
        const serializedConfiguration = this.storage.getItem(this.id);
        configuration = serializedConfiguration ? JSON.parse(serializedConfiguration) : {};
      }
      Object.assign(this.config, configuration);
      return this;
    }
  };

  // ../../node_modules/@probe.gl/log/dist/utils/formatters.js
  function formatTime(ms) {
    let formatted;
    if (ms < 10) {
      formatted = `${ms.toFixed(2)}ms`;
    } else if (ms < 100) {
      formatted = `${ms.toFixed(1)}ms`;
    } else if (ms < 1e3) {
      formatted = `${ms.toFixed(0)}ms`;
    } else {
      formatted = `${(ms / 1e3).toFixed(2)}s`;
    }
    return formatted;
  }
  function leftPad(string, length = 8) {
    const padLength = Math.max(length - string.length, 0);
    return `${" ".repeat(padLength)}${string}`;
  }

  // ../../node_modules/@probe.gl/log/dist/utils/color.js
  var COLOR;
  (function(COLOR2) {
    COLOR2[COLOR2["BLACK"] = 30] = "BLACK";
    COLOR2[COLOR2["RED"] = 31] = "RED";
    COLOR2[COLOR2["GREEN"] = 32] = "GREEN";
    COLOR2[COLOR2["YELLOW"] = 33] = "YELLOW";
    COLOR2[COLOR2["BLUE"] = 34] = "BLUE";
    COLOR2[COLOR2["MAGENTA"] = 35] = "MAGENTA";
    COLOR2[COLOR2["CYAN"] = 36] = "CYAN";
    COLOR2[COLOR2["WHITE"] = 37] = "WHITE";
    COLOR2[COLOR2["BRIGHT_BLACK"] = 90] = "BRIGHT_BLACK";
    COLOR2[COLOR2["BRIGHT_RED"] = 91] = "BRIGHT_RED";
    COLOR2[COLOR2["BRIGHT_GREEN"] = 92] = "BRIGHT_GREEN";
    COLOR2[COLOR2["BRIGHT_YELLOW"] = 93] = "BRIGHT_YELLOW";
    COLOR2[COLOR2["BRIGHT_BLUE"] = 94] = "BRIGHT_BLUE";
    COLOR2[COLOR2["BRIGHT_MAGENTA"] = 95] = "BRIGHT_MAGENTA";
    COLOR2[COLOR2["BRIGHT_CYAN"] = 96] = "BRIGHT_CYAN";
    COLOR2[COLOR2["BRIGHT_WHITE"] = 97] = "BRIGHT_WHITE";
  })(COLOR || (COLOR = {}));
  var BACKGROUND_INCREMENT = 10;
  function getColor(color) {
    if (typeof color !== "string") {
      return color;
    }
    color = color.toUpperCase();
    return COLOR[color] || COLOR.WHITE;
  }
  function addColor(string, color, background) {
    if (!isBrowser3 && typeof string === "string") {
      if (color) {
        const colorCode = getColor(color);
        string = `\x1B[${colorCode}m${string}\x1B[39m`;
      }
      if (background) {
        const colorCode = getColor(background);
        string = `\x1B[${colorCode + BACKGROUND_INCREMENT}m${string}\x1B[49m`;
      }
    }
    return string;
  }

  // ../../node_modules/@probe.gl/log/dist/utils/autobind.js
  function autobind(obj, predefined = ["constructor"]) {
    const proto = Object.getPrototypeOf(obj);
    const propNames = Object.getOwnPropertyNames(proto);
    const object = obj;
    for (const key of propNames) {
      const value = object[key];
      if (typeof value === "function") {
        if (!predefined.find((name) => key === name)) {
          object[key] = value.bind(obj);
        }
      }
    }
  }

  // ../../node_modules/@probe.gl/log/dist/utils/hi-res-timestamp.js
  function getHiResTimestamp() {
    let timestamp;
    if (isBrowser3() && window_3.performance) {
      timestamp = window_3?.performance?.now?.();
    } else if ("hrtime" in process_) {
      const timeParts = process_?.hrtime?.();
      timestamp = timeParts[0] * 1e3 + timeParts[1] / 1e6;
    } else {
      timestamp = Date.now();
    }
    return timestamp;
  }

  // ../../node_modules/@probe.gl/log/dist/loggers/probe-log.js
  var originalConsole = {
    debug: isBrowser3() ? console.debug || console.log : console.log,
    log: console.log,
    info: console.info,
    warn: console.warn,
    error: console.error
  };
  var DEFAULT_LOG_CONFIGURATION = {
    enabled: true,
    level: 0
  };
  var ProbeLog = class extends BaseLog {
    constructor({ id } = { id: "" }) {
      super({ level: 0 });
      this.VERSION = VERSION3;
      this._startTs = getHiResTimestamp();
      this._deltaTs = getHiResTimestamp();
      this.userData = {};
      this.LOG_THROTTLE_TIMEOUT = 0;
      this.id = id;
      this.userData = {};
      this._storage = new LocalStorage(`__probe-${this.id}__`, { [this.id]: DEFAULT_LOG_CONFIGURATION });
      this.timeStamp(`${this.id} started`);
      autobind(this);
      Object.seal(this);
    }
    isEnabled() {
      return this._getConfiguration().enabled;
    }
    getLevel() {
      return this._getConfiguration().level;
    }
    /** @return milliseconds, with fractions */
    getTotal() {
      return Number((getHiResTimestamp() - this._startTs).toPrecision(10));
    }
    /** @return milliseconds, with fractions */
    getDelta() {
      return Number((getHiResTimestamp() - this._deltaTs).toPrecision(10));
    }
    /** @deprecated use logLevel */
    set priority(newPriority) {
      this.level = newPriority;
    }
    /** @deprecated use logLevel */
    get priority() {
      return this.level;
    }
    /** @deprecated use logLevel */
    getPriority() {
      return this.level;
    }
    // Configure
    enable(enabled = true) {
      this._updateConfiguration({ enabled });
      return this;
    }
    setLevel(level) {
      this._updateConfiguration({ level });
      return this;
    }
    /** return the current status of the setting */
    get(setting) {
      return this._getConfiguration()[setting];
    }
    // update the status of the setting
    set(setting, value) {
      this._updateConfiguration({ [setting]: value });
    }
    /** Logs the current settings as a table */
    settings() {
      if (console.table) {
        console.table(this._storage.config);
      } else {
        console.log(this._storage.config);
      }
    }
    // Unconditional logging
    assert(condition, message) {
      if (!condition) {
        throw new Error(message || "Assertion failed");
      }
    }
    warn(message, ...args) {
      return this._log("warn", 0, message, args, {
        method: originalConsole.warn,
        once: true
      });
    }
    error(message, ...args) {
      return this._log("error", 0, message, args, {
        method: originalConsole.error
      });
    }
    /** Print a deprecation warning */
    deprecated(oldUsage, newUsage) {
      return this.warn(`\`${oldUsage}\` is deprecated and will be removed in a later version. Use \`${newUsage}\` instead`);
    }
    /** Print a removal warning */
    removed(oldUsage, newUsage) {
      return this.error(`\`${oldUsage}\` has been removed. Use \`${newUsage}\` instead`);
    }
    probe(logLevel, message, ...args) {
      return this._log("log", logLevel, message, args, {
        method: originalConsole.log,
        time: true,
        once: true
      });
    }
    log(logLevel, message, ...args) {
      return this._log("log", logLevel, message, args, {
        method: originalConsole.debug
      });
    }
    info(logLevel, message, ...args) {
      return this._log("info", logLevel, message, args, { method: console.info });
    }
    once(logLevel, message, ...args) {
      return this._log("once", logLevel, message, args, {
        method: originalConsole.debug || originalConsole.info,
        once: true
      });
    }
    /** Logs an object as a table */
    table(logLevel, table, columns) {
      if (table) {
        return this._log("table", logLevel, table, columns && [columns] || [], {
          method: console.table || noop,
          tag: getTableHeader(table)
        });
      }
      return noop;
    }
    time(logLevel, message) {
      return this._log("time", logLevel, message, [], {
        method: console.time ? console.time : console.info
      });
    }
    timeEnd(logLevel, message) {
      return this._log("time", logLevel, message, [], {
        method: console.timeEnd ? console.timeEnd : console.info
      });
    }
    timeStamp(logLevel, message) {
      return this._log("time", logLevel, message, [], {
        method: console.timeStamp || noop
      });
    }
    group(logLevel, message, opts = { collapsed: false }) {
      const method = (opts.collapsed ? console.groupCollapsed : console.group) || console.info;
      return this._log("group", logLevel, message, [], { method });
    }
    groupCollapsed(logLevel, message, opts = {}) {
      return this.group(logLevel, message, Object.assign({}, opts, { collapsed: true }));
    }
    groupEnd(logLevel) {
      return this._log("groupEnd", logLevel, "", [], {
        method: console.groupEnd || noop
      });
    }
    // EXPERIMENTAL
    withGroup(logLevel, message, func) {
      this.group(logLevel, message)();
      try {
        func();
      } finally {
        this.groupEnd(logLevel)();
      }
    }
    trace() {
      if (console.trace) {
        console.trace();
      }
    }
    _shouldLog(logLevel) {
      return this.isEnabled() && super._shouldLog(logLevel);
    }
    _emit(_type, normalized) {
      const method = normalized.method;
      assert3(method);
      normalized.total = this.getTotal();
      normalized.delta = this.getDelta();
      this._deltaTs = getHiResTimestamp();
      const message = decorateMessage(this.id, normalized.message, normalized);
      return method.bind(console, message, ...normalized.args);
    }
    _getConfiguration() {
      if (!this._storage.config[this.id]) {
        this._updateConfiguration(DEFAULT_LOG_CONFIGURATION);
      }
      return this._storage.config[this.id];
    }
    _updateConfiguration(configuration) {
      const currentConfiguration = this._storage.config[this.id] || {
        ...DEFAULT_LOG_CONFIGURATION
      };
      this._storage.setConfiguration({
        [this.id]: { ...currentConfiguration, ...configuration }
      });
    }
  };
  ProbeLog.VERSION = VERSION3;
  function decorateMessage(id, message, opts) {
    if (typeof message === "string") {
      const time = opts.time ? leftPad(formatTime(opts.total)) : "";
      message = opts.time ? `${id}: ${time}  ${message}` : `${id}: ${message}`;
      message = addColor(message, opts.color, opts.background);
    }
    return message;
  }
  function getTableHeader(table) {
    for (const key in table) {
      for (const title in table[key]) {
        return title || "untitled";
      }
    }
    return "empty";
  }

  // ../../node_modules/@probe.gl/log/dist/init.js
  globalThis.probe = {};

  // ../../node_modules/@probe.gl/log/dist/index.js
  var dist_default = new ProbeLog({ id: "@probe.gl/log" });

  // ../loader-utils/src/lib/log-utils/log.ts
  var VERSION4 = typeof __VERSION__ !== "undefined" ? __VERSION__ : "latest";
  var version = VERSION4[0] >= "0" && VERSION4[0] <= "9" ? `v${VERSION4}` : "";
  function createLog() {
    const log2 = new ProbeLog({ id: "loaders.gl" });
    globalThis.loaders ||= {};
    globalThis.loaders.log = log2;
    globalThis.loaders.version = version;
    globalThis.probe ||= {};
    globalThis.probe.loaders = log2;
    return log2;
  }
  var log = createLog();

  // ../loader-utils/src/lib/module-utils/js-module-utils.ts
  function registerJSModules(modules) {
    globalThis.loaders ||= {};
    globalThis.loaders.modules ||= {};
    Object.assign(globalThis.loaders.modules, modules);
  }
  function getJSModuleOrNull(name) {
    const module = globalThis.loaders?.modules?.[name];
    return module || null;
  }

  // ../loader-utils/src/lib/path-utils/file-aliases.ts
  var pathPrefix = "";
  var fileAliases = {};
  function resolvePath(filename2) {
    for (const alias in fileAliases) {
      if (filename2.startsWith(alias)) {
        const replacement = fileAliases[alias];
        filename2 = filename2.replace(alias, replacement);
      }
    }
    if (!filename2.startsWith("http://") && !filename2.startsWith("https://")) {
      filename2 = `${pathPrefix}${filename2}`;
    }
    return filename2;
  }

  // ../loader-utils/src/lib/path-utils/path.ts
  var path_exports = {};
  __export(path_exports, {
    dirname: () => dirname,
    filename: () => filename,
    join: () => join,
    resolve: () => resolve
  });

  // ../loader-utils/src/lib/path-utils/get-cwd.ts
  function getCWD() {
    if (typeof process !== "undefined" && typeof process.cwd !== "undefined") {
      return process.cwd();
    }
    const pathname = window.location?.pathname;
    return pathname?.slice(0, pathname.lastIndexOf("/") + 1) || "";
  }

  // ../loader-utils/src/lib/path-utils/path.ts
  function filename(url) {
    const slashIndex = url ? url.lastIndexOf("/") : -1;
    return slashIndex >= 0 ? url.substr(slashIndex + 1) : url;
  }
  function dirname(url) {
    const slashIndex = url ? url.lastIndexOf("/") : -1;
    return slashIndex >= 0 ? url.substr(0, slashIndex) : "";
  }
  function join(...parts) {
    const separator = "/";
    parts = parts.map((part, index) => {
      if (index) {
        part = part.replace(new RegExp(`^${separator}`), "");
      }
      if (index !== parts.length - 1) {
        part = part.replace(new RegExp(`${separator}$`), "");
      }
      return part;
    });
    return parts.join(separator);
  }
  function resolve(...components) {
    const paths = [];
    for (let _i = 0; _i < components.length; _i++) {
      paths[_i] = components[_i];
    }
    let resolvedPath = "";
    let resolvedAbsolute = false;
    let cwd;
    for (let i = paths.length - 1; i >= -1 && !resolvedAbsolute; i--) {
      let path;
      if (i >= 0) {
        path = paths[i];
      } else {
        if (cwd === void 0) {
          cwd = getCWD();
        }
        path = cwd;
      }
      if (path.length === 0) {
        continue;
      }
      resolvedPath = `${path}/${resolvedPath}`;
      resolvedAbsolute = path.charCodeAt(0) === SLASH;
    }
    resolvedPath = normalizeStringPosix(resolvedPath, !resolvedAbsolute);
    if (resolvedAbsolute) {
      return `/${resolvedPath}`;
    } else if (resolvedPath.length > 0) {
      return resolvedPath;
    }
    return ".";
  }
  var SLASH = 47;
  var DOT = 46;
  function normalizeStringPosix(path, allowAboveRoot) {
    let res = "";
    let lastSlash = -1;
    let dots = 0;
    let code;
    let isAboveRoot = false;
    for (let i = 0; i <= path.length; ++i) {
      if (i < path.length) {
        code = path.charCodeAt(i);
      } else if (code === SLASH) {
        break;
      } else {
        code = SLASH;
      }
      if (code === SLASH) {
        if (lastSlash === i - 1 || dots === 1) {
        } else if (lastSlash !== i - 1 && dots === 2) {
          if (res.length < 2 || !isAboveRoot || res.charCodeAt(res.length - 1) !== DOT || res.charCodeAt(res.length - 2) !== DOT) {
            if (res.length > 2) {
              const start = res.length - 1;
              let j = start;
              for (; j >= 0; --j) {
                if (res.charCodeAt(j) === SLASH) {
                  break;
                }
              }
              if (j !== start) {
                res = j === -1 ? "" : res.slice(0, j);
                lastSlash = i;
                dots = 0;
                isAboveRoot = false;
                continue;
              }
            } else if (res.length === 2 || res.length === 1) {
              res = "";
              lastSlash = i;
              dots = 0;
              isAboveRoot = false;
              continue;
            }
          }
          if (allowAboveRoot) {
            if (res.length > 0) {
              res += "/..";
            } else {
              res = "..";
            }
            isAboveRoot = true;
          }
        } else {
          const slice = path.slice(lastSlash + 1, i);
          if (res.length > 0) {
            res += `/${slice}`;
          } else {
            res = slice;
          }
          isAboveRoot = false;
        }
        lastSlash = i;
        dots = 0;
      } else if (code === DOT && dots !== -1) {
        ++dots;
      } else {
        dots = -1;
      }
    }
    return res;
  }

  // src/lib/parsers/basis-module-loader.ts
  var BASIS_EXTERNAL_LIBRARIES = {
    /** Basis transcoder, javascript wrapper part */
    TRANSCODER: "basis_transcoder.js",
    /** Basis transcoder, compiled web assembly part */
    TRANSCODER_WASM: "basis_transcoder.wasm",
    /** Basis encoder, javascript wrapper part */
    ENCODER: "basis_encoder.js",
    /** Basis encoder, compiled web assembly part */
    ENCODER_WASM: "basis_encoder.wasm"
  };
  var loadBasisTranscoderPromise;
  async function loadBasisTranscoderModule(options) {
    registerJSModules(options.modules);
    const basis = getJSModuleOrNull("basis");
    if (basis) {
      return basis;
    }
    loadBasisTranscoderPromise ||= loadBasisTranscoder(options);
    return await loadBasisTranscoderPromise;
  }
  async function loadBasisTranscoder(options) {
    let BASIS = null;
    let wasmBinary = null;
    [BASIS, wasmBinary] = await Promise.all([
      await loadLibrary(BASIS_EXTERNAL_LIBRARIES.TRANSCODER, "textures", options),
      await loadLibrary(BASIS_EXTERNAL_LIBRARIES.TRANSCODER_WASM, "textures", options)
    ]);
    BASIS = BASIS || globalThis.BASIS;
    return await initializeBasisTranscoderModule(BASIS, wasmBinary);
  }
  function initializeBasisTranscoderModule(BasisModule, wasmBinary) {
    const options = {};
    if (wasmBinary) {
      options.wasmBinary = wasmBinary;
    }
    return new Promise((resolve2) => {
      BasisModule(options).then((module) => {
        const { BasisFile, initializeBasis } = module;
        initializeBasis();
        resolve2({ BasisFile });
      });
    });
  }
  var loadBasisEncoderPromise;
  async function loadBasisEncoderModule(options) {
    const modules = options.modules || {};
    if (modules.basisEncoder) {
      return modules.basisEncoder;
    }
    loadBasisEncoderPromise = loadBasisEncoderPromise || loadBasisEncoder(options);
    return await loadBasisEncoderPromise;
  }
  async function loadBasisEncoder(options) {
    let BASIS_ENCODER = null;
    let wasmBinary = null;
    [BASIS_ENCODER, wasmBinary] = await Promise.all([
      await loadLibrary(BASIS_EXTERNAL_LIBRARIES.ENCODER, "textures", options),
      await loadLibrary(BASIS_EXTERNAL_LIBRARIES.ENCODER_WASM, "textures", options)
    ]);
    BASIS_ENCODER = BASIS_ENCODER || globalThis.BASIS;
    return await initializeBasisEncoderModule(BASIS_ENCODER, wasmBinary);
  }
  function initializeBasisEncoderModule(BasisEncoderModule, wasmBinary) {
    const options = {};
    if (wasmBinary) {
      options.wasmBinary = wasmBinary;
    }
    return new Promise((resolve2) => {
      BasisEncoderModule(options).then((module) => {
        const { BasisFile, KTX2File, initializeBasis, BasisEncoder } = module;
        initializeBasis();
        resolve2({ BasisFile, KTX2File, BasisEncoder });
      });
    });
  }

  // src/lib/gl-extensions.ts
  var GL_RGB = 6407;
  var GL_RGBA = 6408;
  var GL_RGBA4 = 32854;
  var GL_RGB5_A1 = 32855;
  var GL_RGBA8 = 32856;
  var GL_RGB565 = 36194;
  var GL_RGBA32F = 34836;
  var GL_COMPRESSED_RGB_S3TC_DXT1_EXT = 33776;
  var GL_COMPRESSED_RGBA_S3TC_DXT1_EXT = 33777;
  var GL_COMPRESSED_RGBA_S3TC_DXT3_EXT = 33778;
  var GL_COMPRESSED_RGBA_S3TC_DXT5_EXT = 33779;
  var GL_COMPRESSED_R11_EAC = 37488;
  var GL_COMPRESSED_SIGNED_R11_EAC = 37489;
  var GL_COMPRESSED_RG11_EAC = 37490;
  var GL_COMPRESSED_SIGNED_RG11_EAC = 37491;
  var GL_COMPRESSED_RGB8_ETC2 = 37492;
  var GL_COMPRESSED_RGBA8_ETC2_EAC = 37493;
  var GL_COMPRESSED_SRGB8_ETC2 = 37494;
  var GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC = 37495;
  var GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2 = 37496;
  var GL_COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2 = 37497;
  var GL_COMPRESSED_RGB_PVRTC_4BPPV1_IMG = 35840;
  var GL_COMPRESSED_RGB_PVRTC_2BPPV1_IMG = 35841;
  var GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG = 35842;
  var GL_COMPRESSED_RGBA_PVRTC_2BPPV1_IMG = 35843;
  var GL_COMPRESSED_RGB_ETC1_WEBGL = 36196;
  var GL_COMPRESSED_RGB_ATC_WEBGL = 35986;
  var GL_COMPRESSED_RGBA_ATC_EXPLICIT_ALPHA_WEBGL = 35987;
  var GL_COMPRESSED_RGBA_ATC_INTERPOLATED_ALPHA_WEBGL = 34798;
  var GL_COMPRESSED_RGBA_ASTC_4x4_KHR = 37808;
  var GL_COMPRESSED_RGBA_ASTC_5x4_KHR = 37809;
  var GL_COMPRESSED_RGBA_ASTC_5x5_KHR = 37810;
  var GL_COMPRESSED_RGBA_ASTC_6x5_KHR = 37811;
  var GL_COMPRESSED_RGBA_ASTC_6x6_KHR = 37812;
  var GL_COMPRESSED_RGBA_ASTC_8x5_KHR = 37813;
  var GL_COMPRESSED_RGBA_ASTC_8x6_KHR = 37814;
  var GL_COMPRESSED_RGBA_ASTC_8x8_KHR = 37815;
  var GL_COMPRESSED_RGBA_ASTC_10x5_KHR = 37816;
  var GL_COMPRESSED_RGBA_ASTC_10x6_KHR = 37817;
  var GL_COMPRESSED_RGBA_ASTC_10x8_KHR = 37818;
  var GL_COMPRESSED_RGBA_ASTC_10x10_KHR = 37819;
  var GL_COMPRESSED_RGBA_ASTC_12x10_KHR = 37820;
  var GL_COMPRESSED_RGBA_ASTC_12x12_KHR = 37821;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR = 37840;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR = 37841;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR = 37842;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR = 37843;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR = 37844;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR = 37845;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR = 37846;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR = 37847;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR = 37848;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR = 37849;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR = 37850;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR = 37851;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR = 37852;
  var GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR = 37853;
  var GL_COMPRESSED_RED_RGTC1_EXT = 36283;
  var GL_COMPRESSED_SIGNED_RED_RGTC1_EXT = 36284;
  var GL_COMPRESSED_RED_GREEN_RGTC2_EXT = 36285;
  var GL_COMPRESSED_SIGNED_RED_GREEN_RGTC2_EXT = 36286;
  var GL_COMPRESSED_RGBA_BPTC_UNORM_EXT = 36492;
  var GL_COMPRESSED_SRGB_ALPHA_BPTC_UNORM_EXT = 36493;
  var GL_COMPRESSED_RGB_BPTC_SIGNED_FLOAT_EXT = 36494;
  var GL_COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT_EXT = 36495;
  var GL_COMPRESSED_SRGB_S3TC_DXT1_EXT = 35916;
  var GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT = 35917;
  var GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT = 35918;
  var GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT = 35919;
  var GL_EXTENSIONS_CONSTANTS = {
    RGB: GL_RGB,
    RGBA: GL_RGBA,
    RGBA4: GL_RGBA4,
    RGB5_A1: GL_RGB5_A1,
    RGBA8: GL_RGBA8,
    RGB565: GL_RGB565,
    RGBA32F: GL_RGBA32F,
    COMPRESSED_RGB_S3TC_DXT1_EXT: GL_COMPRESSED_RGB_S3TC_DXT1_EXT,
    COMPRESSED_RGBA_S3TC_DXT1_EXT: GL_COMPRESSED_RGBA_S3TC_DXT1_EXT,
    COMPRESSED_RGBA_S3TC_DXT3_EXT: GL_COMPRESSED_RGBA_S3TC_DXT3_EXT,
    COMPRESSED_RGBA_S3TC_DXT5_EXT: GL_COMPRESSED_RGBA_S3TC_DXT5_EXT,
    COMPRESSED_R11_EAC: GL_COMPRESSED_R11_EAC,
    COMPRESSED_SIGNED_R11_EAC: GL_COMPRESSED_SIGNED_R11_EAC,
    COMPRESSED_RG11_EAC: GL_COMPRESSED_RG11_EAC,
    COMPRESSED_SIGNED_RG11_EAC: GL_COMPRESSED_SIGNED_RG11_EAC,
    COMPRESSED_RGB8_ETC2: GL_COMPRESSED_RGB8_ETC2,
    COMPRESSED_RGBA8_ETC2_EAC: GL_COMPRESSED_RGBA8_ETC2_EAC,
    COMPRESSED_SRGB8_ETC2: GL_COMPRESSED_SRGB8_ETC2,
    COMPRESSED_SRGB8_ALPHA8_ETC2_EAC: GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC,
    COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2: GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2,
    COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2: GL_COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2,
    COMPRESSED_RGB_PVRTC_4BPPV1_IMG: GL_COMPRESSED_RGB_PVRTC_4BPPV1_IMG,
    COMPRESSED_RGB_PVRTC_2BPPV1_IMG: GL_COMPRESSED_RGB_PVRTC_2BPPV1_IMG,
    COMPRESSED_RGBA_PVRTC_4BPPV1_IMG: GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG,
    COMPRESSED_RGBA_PVRTC_2BPPV1_IMG: GL_COMPRESSED_RGBA_PVRTC_2BPPV1_IMG,
    COMPRESSED_RGB_ETC1_WEBGL: GL_COMPRESSED_RGB_ETC1_WEBGL,
    COMPRESSED_RGB_ATC_WEBGL: GL_COMPRESSED_RGB_ATC_WEBGL,
    COMPRESSED_RGBA_ATC_EXPLICIT_ALPHA_WEBGL: GL_COMPRESSED_RGBA_ATC_EXPLICIT_ALPHA_WEBGL,
    COMPRESSED_RGBA_ATC_INTERPOLATED_ALPHA_WEBGL: GL_COMPRESSED_RGBA_ATC_INTERPOLATED_ALPHA_WEBGL,
    COMPRESSED_RGBA_ASTC_4x4_KHR: GL_COMPRESSED_RGBA_ASTC_4x4_KHR,
    COMPRESSED_RGBA_ASTC_5x4_KHR: GL_COMPRESSED_RGBA_ASTC_5x4_KHR,
    COMPRESSED_RGBA_ASTC_5x5_KHR: GL_COMPRESSED_RGBA_ASTC_5x5_KHR,
    COMPRESSED_RGBA_ASTC_6x5_KHR: GL_COMPRESSED_RGBA_ASTC_6x5_KHR,
    COMPRESSED_RGBA_ASTC_6x6_KHR: GL_COMPRESSED_RGBA_ASTC_6x6_KHR,
    COMPRESSED_RGBA_ASTC_8x5_KHR: GL_COMPRESSED_RGBA_ASTC_8x5_KHR,
    COMPRESSED_RGBA_ASTC_8x6_KHR: GL_COMPRESSED_RGBA_ASTC_8x6_KHR,
    COMPRESSED_RGBA_ASTC_8x8_KHR: GL_COMPRESSED_RGBA_ASTC_8x8_KHR,
    COMPRESSED_RGBA_ASTC_10x5_KHR: GL_COMPRESSED_RGBA_ASTC_10x5_KHR,
    COMPRESSED_RGBA_ASTC_10x6_KHR: GL_COMPRESSED_RGBA_ASTC_10x6_KHR,
    COMPRESSED_RGBA_ASTC_10x8_KHR: GL_COMPRESSED_RGBA_ASTC_10x8_KHR,
    COMPRESSED_RGBA_ASTC_10x10_KHR: GL_COMPRESSED_RGBA_ASTC_10x10_KHR,
    COMPRESSED_RGBA_ASTC_12x10_KHR: GL_COMPRESSED_RGBA_ASTC_12x10_KHR,
    COMPRESSED_RGBA_ASTC_12x12_KHR: GL_COMPRESSED_RGBA_ASTC_12x12_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR,
    COMPRESSED_RED_RGTC1_EXT: GL_COMPRESSED_RED_RGTC1_EXT,
    COMPRESSED_SIGNED_RED_RGTC1_EXT: GL_COMPRESSED_SIGNED_RED_RGTC1_EXT,
    COMPRESSED_RED_GREEN_RGTC2_EXT: GL_COMPRESSED_RED_GREEN_RGTC2_EXT,
    COMPRESSED_SIGNED_RED_GREEN_RGTC2_EXT: GL_COMPRESSED_SIGNED_RED_GREEN_RGTC2_EXT,
    COMPRESSED_RGBA_BPTC_UNORM_EXT: GL_COMPRESSED_RGBA_BPTC_UNORM_EXT,
    COMPRESSED_SRGB_ALPHA_BPTC_UNORM_EXT: GL_COMPRESSED_SRGB_ALPHA_BPTC_UNORM_EXT,
    COMPRESSED_RGB_BPTC_SIGNED_FLOAT_EXT: GL_COMPRESSED_RGB_BPTC_SIGNED_FLOAT_EXT,
    COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT_EXT: GL_COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT_EXT,
    COMPRESSED_SRGB_S3TC_DXT1_EXT: GL_COMPRESSED_SRGB_S3TC_DXT1_EXT,
    COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT: GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT,
    COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT: GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT,
    COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT: GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT,
    // Deprecated legacy aliases with non-standard ASTC casing.
    COMPRESSED_RGBA_ASTC_4X4_KHR: GL_COMPRESSED_RGBA_ASTC_4x4_KHR,
    COMPRESSED_RGBA_ASTC_5X4_KHR: GL_COMPRESSED_RGBA_ASTC_5x4_KHR,
    COMPRESSED_RGBA_ASTC_5X5_KHR: GL_COMPRESSED_RGBA_ASTC_5x5_KHR,
    COMPRESSED_RGBA_ASTC_6X5_KHR: GL_COMPRESSED_RGBA_ASTC_6x5_KHR,
    COMPRESSED_RGBA_ASTC_6X6_KHR: GL_COMPRESSED_RGBA_ASTC_6x6_KHR,
    COMPRESSED_RGBA_ASTC_8X5_KHR: GL_COMPRESSED_RGBA_ASTC_8x5_KHR,
    COMPRESSED_RGBA_ASTC_8X6_KHR: GL_COMPRESSED_RGBA_ASTC_8x6_KHR,
    COMPRESSED_RGBA_ASTC_8X8_KHR: GL_COMPRESSED_RGBA_ASTC_8x8_KHR,
    COMPRESSED_RGBA_ASTC_10X5_KHR: GL_COMPRESSED_RGBA_ASTC_10x5_KHR,
    COMPRESSED_RGBA_ASTC_10X6_KHR: GL_COMPRESSED_RGBA_ASTC_10x6_KHR,
    COMPRESSED_RGBA_ASTC_10X8_KHR: GL_COMPRESSED_RGBA_ASTC_10x8_KHR,
    COMPRESSED_RGBA_ASTC_10X10_KHR: GL_COMPRESSED_RGBA_ASTC_10x10_KHR,
    COMPRESSED_RGBA_ASTC_12X10_KHR: GL_COMPRESSED_RGBA_ASTC_12x10_KHR,
    COMPRESSED_RGBA_ASTC_12X12_KHR: GL_COMPRESSED_RGBA_ASTC_12x12_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_4X4_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_5X4_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_5X5_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_6X5_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_6X6_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_8X5_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_8X6_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_8X8_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_10X5_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_10X6_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_10X8_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_10X10_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_12X10_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR,
    COMPRESSED_SRGB8_ALPHA8_ASTC_12X12_KHR: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR
  };

  // src/lib/utils/detect-supported-texture-formats.ts
  var BROWSER_PREFIXES = ["", "WEBKIT_", "MOZ_"];
  var WEBGL_TEXTURE_FORMATS = {
    /* eslint-disable camelcase */
    WEBGL_compressed_texture_s3tc: [
      "bc1-rgb-unorm-webgl",
      "bc1-rgba-unorm",
      "bc2-rgba-unorm",
      "bc3-rgba-unorm"
    ],
    WEBGL_compressed_texture_s3tc_srgb: [
      "bc1-rgb-unorm-srgb-webgl",
      "bc1-rgba-unorm-srgb",
      "bc2-rgba-unorm-srgb",
      "bc3-rgba-unorm-srgb"
    ],
    EXT_texture_compression_rgtc: ["bc4-r-unorm", "bc4-r-snorm", "bc5-rg-unorm", "bc5-rg-snorm"],
    EXT_texture_compression_bptc: [
      "bc6h-rgb-ufloat",
      "bc6h-rgb-float",
      "bc7-rgba-unorm",
      "bc7-rgba-unorm-srgb"
    ],
    WEBGL_compressed_texture_etc1: ["etc1-rgb-unorm-webgl"],
    WEBGL_compressed_texture_etc: [
      "etc2-rgb8unorm",
      "etc2-rgb8unorm-srgb",
      "etc2-rgb8a1unorm",
      "etc2-rgb8a1unorm-srgb",
      "etc2-rgba8unorm",
      "etc2-rgba8unorm-srgb",
      "eac-r11unorm",
      "eac-r11snorm",
      "eac-rg11unorm",
      "eac-rg11snorm"
    ],
    WEBGL_compressed_texture_pvrtc: [
      "pvrtc-rgb4unorm-webgl",
      "pvrtc-rgba4unorm-webgl",
      "pvrtc-rgb2unorm-webgl",
      "pvrtc-rgba2unorm-webgl"
    ],
    WEBGL_compressed_texture_atc: [
      "atc-rgb-unorm-webgl",
      "atc-rgba-unorm-webgl",
      "atc-rgbai-unorm-webgl"
    ],
    WEBGL_compressed_texture_astc: [
      "astc-4x4-unorm",
      "astc-4x4-unorm-srgb",
      "astc-5x4-unorm",
      "astc-5x4-unorm-srgb",
      "astc-5x5-unorm",
      "astc-5x5-unorm-srgb",
      "astc-6x5-unorm",
      "astc-6x5-unorm-srgb",
      "astc-6x6-unorm",
      "astc-6x6-unorm-srgb",
      "astc-8x5-unorm",
      "astc-8x5-unorm-srgb",
      "astc-8x6-unorm",
      "astc-8x6-unorm-srgb",
      "astc-8x8-unorm",
      "astc-8x8-unorm-srgb",
      "astc-10x5-unorm",
      "astc-10x5-unorm-srgb",
      "astc-10x6-unorm",
      "astc-10x6-unorm-srgb",
      "astc-10x8-unorm",
      "astc-10x8-unorm-srgb",
      "astc-10x10-unorm",
      "astc-10x10-unorm-srgb",
      "astc-12x10-unorm",
      "astc-12x10-unorm-srgb",
      "astc-12x12-unorm",
      "astc-12x12-unorm-srgb"
    ]
    /* eslint-enable camelcase */
  };
  var textureFormats = null;
  function detectSupportedTextureFormats(gl) {
    if (!textureFormats) {
      gl = gl || getWebGLContext() || void 0;
      textureFormats = /* @__PURE__ */ new Set();
      for (const prefix of BROWSER_PREFIXES) {
        for (const extension in WEBGL_TEXTURE_FORMATS) {
          if (gl && gl.getExtension(`${prefix}${extension}`)) {
            for (const textureFormat of WEBGL_TEXTURE_FORMATS[extension]) {
              textureFormats.add(textureFormat);
            }
          }
        }
      }
    }
    return textureFormats;
  }
  function getWebGLContext() {
    try {
      const canvas = document.createElement("canvas");
      return canvas.getContext("webgl");
    } catch (error) {
      return null;
    }
  }

  // ../../node_modules/ktx-parse/dist/ktx-parse.modern.js
  var KHR_SUPERCOMPRESSION_NONE = 0;
  var KHR_DF_KHR_DESCRIPTORTYPE_BASICFORMAT = 0;
  var KHR_DF_VENDORID_KHRONOS = 0;
  var KHR_DF_VERSION = 2;
  var KHR_DF_MODEL_UNSPECIFIED = 0;
  var KHR_DF_FLAG_ALPHA_STRAIGHT = 0;
  var KHR_DF_TRANSFER_SRGB = 2;
  var KHR_DF_PRIMARIES_BT709 = 1;
  var KHR_DF_SAMPLE_DATATYPE_SIGNED = 64;
  var VK_FORMAT_UNDEFINED = 0;
  var KTX2Container = class {
    constructor() {
      this.vkFormat = VK_FORMAT_UNDEFINED;
      this.typeSize = 1;
      this.pixelWidth = 0;
      this.pixelHeight = 0;
      this.pixelDepth = 0;
      this.layerCount = 0;
      this.faceCount = 1;
      this.supercompressionScheme = KHR_SUPERCOMPRESSION_NONE;
      this.levels = [];
      this.dataFormatDescriptor = [{
        vendorId: KHR_DF_VENDORID_KHRONOS,
        descriptorType: KHR_DF_KHR_DESCRIPTORTYPE_BASICFORMAT,
        descriptorBlockSize: 0,
        versionNumber: KHR_DF_VERSION,
        colorModel: KHR_DF_MODEL_UNSPECIFIED,
        colorPrimaries: KHR_DF_PRIMARIES_BT709,
        transferFunction: KHR_DF_TRANSFER_SRGB,
        flags: KHR_DF_FLAG_ALPHA_STRAIGHT,
        texelBlockDimension: [0, 0, 0, 0],
        bytesPlane: [0, 0, 0, 0, 0, 0, 0, 0],
        samples: []
      }];
      this.keyValue = {};
      this.globalData = null;
    }
  };
  var BufferReader = class {
    constructor(data, byteOffset, byteLength, littleEndian) {
      this._dataView = void 0;
      this._littleEndian = void 0;
      this._offset = void 0;
      this._dataView = new DataView(data.buffer, data.byteOffset + byteOffset, byteLength);
      this._littleEndian = littleEndian;
      this._offset = 0;
    }
    _nextUint8() {
      const value = this._dataView.getUint8(this._offset);
      this._offset += 1;
      return value;
    }
    _nextUint16() {
      const value = this._dataView.getUint16(this._offset, this._littleEndian);
      this._offset += 2;
      return value;
    }
    _nextUint32() {
      const value = this._dataView.getUint32(this._offset, this._littleEndian);
      this._offset += 4;
      return value;
    }
    _nextUint64() {
      const left = this._dataView.getUint32(this._offset, this._littleEndian);
      const right = this._dataView.getUint32(this._offset + 4, this._littleEndian);
      const value = left + 2 ** 32 * right;
      this._offset += 8;
      return value;
    }
    _nextInt32() {
      const value = this._dataView.getInt32(this._offset, this._littleEndian);
      this._offset += 4;
      return value;
    }
    _nextUint8Array(len) {
      const value = new Uint8Array(this._dataView.buffer, this._dataView.byteOffset + this._offset, len);
      this._offset += len;
      return value;
    }
    _skip(bytes) {
      this._offset += bytes;
      return this;
    }
    _scan(maxByteLength, term = 0) {
      const byteOffset = this._offset;
      let byteLength = 0;
      while (this._dataView.getUint8(this._offset) !== term && byteLength < maxByteLength) {
        byteLength++;
        this._offset++;
      }
      if (byteLength < maxByteLength)
        this._offset++;
      return new Uint8Array(this._dataView.buffer, this._dataView.byteOffset + byteOffset, byteLength);
    }
  };
  var NUL = new Uint8Array([0]);
  var KTX2_ID = [
    // '´', 'K', 'T', 'X', '2', '0', 'ª', '\r', '\n', '\x1A', '\n'
    171,
    75,
    84,
    88,
    32,
    50,
    48,
    187,
    13,
    10,
    26,
    10
  ];
  function decodeText(buffer) {
    return new TextDecoder().decode(buffer);
  }
  function read(data) {
    const id = new Uint8Array(data.buffer, data.byteOffset, KTX2_ID.length);
    if (id[0] !== KTX2_ID[0] || // '´'
    id[1] !== KTX2_ID[1] || // 'K'
    id[2] !== KTX2_ID[2] || // 'T'
    id[3] !== KTX2_ID[3] || // 'X'
    id[4] !== KTX2_ID[4] || // ' '
    id[5] !== KTX2_ID[5] || // '2'
    id[6] !== KTX2_ID[6] || // '0'
    id[7] !== KTX2_ID[7] || // 'ª'
    id[8] !== KTX2_ID[8] || // '\r'
    id[9] !== KTX2_ID[9] || // '\n'
    id[10] !== KTX2_ID[10] || // '\x1A'
    id[11] !== KTX2_ID[11]) {
      throw new Error("Missing KTX 2.0 identifier.");
    }
    const container = new KTX2Container();
    const headerByteLength = 17 * Uint32Array.BYTES_PER_ELEMENT;
    const headerReader = new BufferReader(data, KTX2_ID.length, headerByteLength, true);
    container.vkFormat = headerReader._nextUint32();
    container.typeSize = headerReader._nextUint32();
    container.pixelWidth = headerReader._nextUint32();
    container.pixelHeight = headerReader._nextUint32();
    container.pixelDepth = headerReader._nextUint32();
    container.layerCount = headerReader._nextUint32();
    container.faceCount = headerReader._nextUint32();
    const levelCount = headerReader._nextUint32();
    container.supercompressionScheme = headerReader._nextUint32();
    const dfdByteOffset = headerReader._nextUint32();
    const dfdByteLength = headerReader._nextUint32();
    const kvdByteOffset = headerReader._nextUint32();
    const kvdByteLength = headerReader._nextUint32();
    const sgdByteOffset = headerReader._nextUint64();
    const sgdByteLength = headerReader._nextUint64();
    const levelByteLength = levelCount * 3 * 8;
    const levelReader = new BufferReader(data, KTX2_ID.length + headerByteLength, levelByteLength, true);
    for (let i = 0; i < levelCount; i++) {
      container.levels.push({
        levelData: new Uint8Array(data.buffer, data.byteOffset + levelReader._nextUint64(), levelReader._nextUint64()),
        uncompressedByteLength: levelReader._nextUint64()
      });
    }
    const dfdReader = new BufferReader(data, dfdByteOffset, dfdByteLength, true);
    const dfd = {
      vendorId: dfdReader._skip(
        4
        /* totalSize */
      )._nextUint16(),
      descriptorType: dfdReader._nextUint16(),
      versionNumber: dfdReader._nextUint16(),
      descriptorBlockSize: dfdReader._nextUint16(),
      colorModel: dfdReader._nextUint8(),
      colorPrimaries: dfdReader._nextUint8(),
      transferFunction: dfdReader._nextUint8(),
      flags: dfdReader._nextUint8(),
      texelBlockDimension: [dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8()],
      bytesPlane: [dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8()],
      samples: []
    };
    const sampleStart = 6;
    const sampleWords = 4;
    const numSamples = (dfd.descriptorBlockSize / 4 - sampleStart) / sampleWords;
    for (let i = 0; i < numSamples; i++) {
      const sample = {
        bitOffset: dfdReader._nextUint16(),
        bitLength: dfdReader._nextUint8(),
        channelType: dfdReader._nextUint8(),
        samplePosition: [dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8(), dfdReader._nextUint8()],
        sampleLower: -Infinity,
        sampleUpper: Infinity
      };
      if (sample.channelType & KHR_DF_SAMPLE_DATATYPE_SIGNED) {
        sample.sampleLower = dfdReader._nextInt32();
        sample.sampleUpper = dfdReader._nextInt32();
      } else {
        sample.sampleLower = dfdReader._nextUint32();
        sample.sampleUpper = dfdReader._nextUint32();
      }
      dfd.samples[i] = sample;
    }
    container.dataFormatDescriptor.length = 0;
    container.dataFormatDescriptor.push(dfd);
    const kvdReader = new BufferReader(data, kvdByteOffset, kvdByteLength, true);
    while (kvdReader._offset < kvdByteLength) {
      const keyValueByteLength = kvdReader._nextUint32();
      const keyData = kvdReader._scan(keyValueByteLength);
      const key = decodeText(keyData);
      container.keyValue[key] = kvdReader._nextUint8Array(keyValueByteLength - keyData.byteLength - 1);
      if (key.match(/^ktx/i)) {
        const text = decodeText(container.keyValue[key]);
        container.keyValue[key] = text.substring(0, text.lastIndexOf("\0"));
      }
      const kvPadding = keyValueByteLength % 4 ? 4 - keyValueByteLength % 4 : 0;
      kvdReader._skip(kvPadding);
    }
    if (sgdByteLength <= 0)
      return container;
    const sgdReader = new BufferReader(data, sgdByteOffset, sgdByteLength, true);
    const endpointCount = sgdReader._nextUint16();
    const selectorCount = sgdReader._nextUint16();
    const endpointsByteLength = sgdReader._nextUint32();
    const selectorsByteLength = sgdReader._nextUint32();
    const tablesByteLength = sgdReader._nextUint32();
    const extendedByteLength = sgdReader._nextUint32();
    const imageDescs = [];
    for (let i = 0; i < levelCount; i++) {
      imageDescs.push({
        imageFlags: sgdReader._nextUint32(),
        rgbSliceByteOffset: sgdReader._nextUint32(),
        rgbSliceByteLength: sgdReader._nextUint32(),
        alphaSliceByteOffset: sgdReader._nextUint32(),
        alphaSliceByteLength: sgdReader._nextUint32()
      });
    }
    const endpointsByteOffset = sgdByteOffset + sgdReader._offset;
    const selectorsByteOffset = endpointsByteOffset + endpointsByteLength;
    const tablesByteOffset = selectorsByteOffset + selectorsByteLength;
    const extendedByteOffset = tablesByteOffset + tablesByteLength;
    const endpointsData = new Uint8Array(data.buffer, data.byteOffset + endpointsByteOffset, endpointsByteLength);
    const selectorsData = new Uint8Array(data.buffer, data.byteOffset + selectorsByteOffset, selectorsByteLength);
    const tablesData = new Uint8Array(data.buffer, data.byteOffset + tablesByteOffset, tablesByteLength);
    const extendedData = new Uint8Array(data.buffer, data.byteOffset + extendedByteOffset, extendedByteLength);
    container.globalData = {
      endpointCount,
      selectorCount,
      imageDescs,
      endpointsData,
      selectorsData,
      tablesData,
      extendedData
    };
    return container;
  }

  // src/lib/utils/texture-format-map.ts
  var WEBGL_TO_TEXTURE_FORMAT = {
    [GL_RGBA32F]: "rgba32float",
    [GL_COMPRESSED_RGB_S3TC_DXT1_EXT]: "bc1-rgb-unorm-webgl",
    [GL_COMPRESSED_SRGB_S3TC_DXT1_EXT]: "bc1-rgb-unorm-srgb-webgl",
    [GL_COMPRESSED_RGBA_S3TC_DXT1_EXT]: "bc1-rgba-unorm",
    [GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT]: "bc1-rgba-unorm-srgb",
    [GL_COMPRESSED_RGBA_S3TC_DXT3_EXT]: "bc2-rgba-unorm",
    [GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT]: "bc2-rgba-unorm-srgb",
    [GL_COMPRESSED_RGBA_S3TC_DXT5_EXT]: "bc3-rgba-unorm",
    [GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT]: "bc3-rgba-unorm-srgb",
    [GL_COMPRESSED_RED_RGTC1_EXT]: "bc4-r-unorm",
    [GL_COMPRESSED_SIGNED_RED_RGTC1_EXT]: "bc4-r-snorm",
    [GL_COMPRESSED_RED_GREEN_RGTC2_EXT]: "bc5-rg-unorm",
    [GL_COMPRESSED_SIGNED_RED_GREEN_RGTC2_EXT]: "bc5-rg-snorm",
    [GL_COMPRESSED_RGB8_ETC2]: "etc2-rgb8unorm",
    [GL_COMPRESSED_SRGB8_ETC2]: "etc2-rgb8unorm-srgb",
    [GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2]: "etc2-rgb8a1unorm",
    [GL_COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2]: "etc2-rgb8a1unorm-srgb",
    [GL_COMPRESSED_RGBA8_ETC2_EAC]: "etc2-rgba8unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC]: "etc2-rgba8unorm-srgb",
    [GL_COMPRESSED_R11_EAC]: "eac-r11unorm",
    [GL_COMPRESSED_SIGNED_R11_EAC]: "eac-r11snorm",
    [GL_COMPRESSED_RG11_EAC]: "eac-rg11unorm",
    [GL_COMPRESSED_SIGNED_RG11_EAC]: "eac-rg11snorm",
    [GL_COMPRESSED_RGBA_ASTC_4x4_KHR]: "astc-4x4-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR]: "astc-4x4-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_5x4_KHR]: "astc-5x4-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR]: "astc-5x4-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_5x5_KHR]: "astc-5x5-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR]: "astc-5x5-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_6x5_KHR]: "astc-6x5-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR]: "astc-6x5-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_6x6_KHR]: "astc-6x6-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR]: "astc-6x6-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_8x5_KHR]: "astc-8x5-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR]: "astc-8x5-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_8x6_KHR]: "astc-8x6-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR]: "astc-8x6-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_8x8_KHR]: "astc-8x8-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR]: "astc-8x8-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_10x5_KHR]: "astc-10x5-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR]: "astc-10x5-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_10x6_KHR]: "astc-10x6-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR]: "astc-10x6-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_10x8_KHR]: "astc-10x8-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR]: "astc-10x8-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_10x10_KHR]: "astc-10x10-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR]: "astc-10x10-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_12x10_KHR]: "astc-12x10-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR]: "astc-12x10-unorm-srgb",
    [GL_COMPRESSED_RGBA_ASTC_12x12_KHR]: "astc-12x12-unorm",
    [GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR]: "astc-12x12-unorm-srgb",
    [GL_COMPRESSED_RGB_PVRTC_4BPPV1_IMG]: "pvrtc-rgb4unorm-webgl",
    [GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG]: "pvrtc-rgba4unorm-webgl",
    [GL_COMPRESSED_RGB_PVRTC_2BPPV1_IMG]: "pvrtc-rgb2unorm-webgl",
    [GL_COMPRESSED_RGBA_PVRTC_2BPPV1_IMG]: "pvrtc-rgba2unorm-webgl",
    [GL_COMPRESSED_RGB_ETC1_WEBGL]: "etc1-rgb-unorm-webgl",
    [GL_COMPRESSED_RGB_ATC_WEBGL]: "atc-rgb-unorm-webgl",
    [GL_COMPRESSED_RGBA_ATC_EXPLICIT_ALPHA_WEBGL]: "atc-rgba-unorm-webgl",
    [GL_COMPRESSED_RGBA_ATC_INTERPOLATED_ALPHA_WEBGL]: "atc-rgbai-unorm-webgl"
  };
  var TEXTURE_FORMAT_TO_WEBGL = Object.fromEntries(
    Object.entries(WEBGL_TO_TEXTURE_FORMAT).map(([format, textureFormat]) => [
      textureFormat,
      Number(format)
    ])
  );
  function getTextureFormatFromWebGLFormat(format) {
    if (format === void 0) {
      return void 0;
    }
    return WEBGL_TO_TEXTURE_FORMAT[format];
  }
  function getWebGLFormatFromTextureFormat(textureFormat) {
    if (textureFormat === void 0) {
      return void 0;
    }
    return TEXTURE_FORMAT_TO_WEBGL[textureFormat];
  }

  // src/lib/utils/extract-mipmap-images.ts
  function extractMipmapImages(data, options) {
    const images = new Array(options.mipMapLevels);
    const textureFormat = options.textureFormat || getTextureFormatFromWebGLFormat(options.internalFormat);
    const format = options.internalFormat || getWebGLFormatFromTextureFormat(options.textureFormat);
    let levelWidth = options.width;
    let levelHeight = options.height;
    let offset = 0;
    for (let i = 0; i < options.mipMapLevels; ++i) {
      const levelSize = getLevelSize(options, levelWidth, levelHeight, data, i);
      const levelData = getLevelData(data, i, offset, levelSize);
      const image = {
        shape: "texture-level",
        compressed: true,
        data: levelData,
        width: levelWidth,
        height: levelHeight,
        levelSize
      };
      if (format !== void 0) {
        image.format = format;
      }
      if (textureFormat) {
        image.textureFormat = textureFormat;
      }
      images[i] = image;
      levelWidth = Math.max(1, levelWidth >> 1);
      levelHeight = Math.max(1, levelHeight >> 1);
      offset += levelSize;
    }
    return images;
  }
  function getLevelData(data, index, offset, levelSize) {
    if (!Array.isArray(data)) {
      return new Uint8Array(data.buffer, data.byteOffset + offset, levelSize);
    }
    return data[index].levelData;
  }
  function getLevelSize(options, levelWidth, levelHeight, data, index) {
    if (!Array.isArray(data)) {
      return options.sizeFunction(levelWidth, levelHeight);
    }
    return options.sizeFunction(data[index]);
  }

  // src/lib/utils/ktx-format-helper.ts
  var VULKAN_TO_WEBGL_FORMAT_MAP = {
    131: GL_COMPRESSED_RGB_S3TC_DXT1_EXT,
    132: GL_COMPRESSED_SRGB_S3TC_DXT1_EXT,
    133: GL_COMPRESSED_RGBA_S3TC_DXT1_EXT,
    134: GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT,
    135: GL_COMPRESSED_RGBA_S3TC_DXT3_EXT,
    136: GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT,
    137: GL_COMPRESSED_RGBA_S3TC_DXT5_EXT,
    138: GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT,
    139: GL_COMPRESSED_RED_RGTC1_EXT,
    140: GL_COMPRESSED_SIGNED_RED_RGTC1_EXT,
    141: GL_COMPRESSED_RED_GREEN_RGTC2_EXT,
    142: GL_COMPRESSED_SIGNED_RED_GREEN_RGTC2_EXT,
    147: GL_COMPRESSED_RGB8_ETC2,
    148: GL_COMPRESSED_SRGB8_ETC2,
    149: GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2,
    150: GL_COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2,
    151: GL_COMPRESSED_RGBA8_ETC2_EAC,
    152: GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC,
    153: GL_COMPRESSED_R11_EAC,
    154: GL_COMPRESSED_SIGNED_R11_EAC,
    155: GL_COMPRESSED_RG11_EAC,
    156: GL_COMPRESSED_SIGNED_RG11_EAC,
    157: GL_COMPRESSED_RGBA_ASTC_4x4_KHR,
    158: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR,
    159: GL_COMPRESSED_RGBA_ASTC_5x4_KHR,
    160: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR,
    161: GL_COMPRESSED_RGBA_ASTC_5x5_KHR,
    162: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR,
    163: GL_COMPRESSED_RGBA_ASTC_6x5_KHR,
    164: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR,
    165: GL_COMPRESSED_RGBA_ASTC_6x6_KHR,
    166: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR,
    167: GL_COMPRESSED_RGBA_ASTC_8x5_KHR,
    168: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR,
    169: GL_COMPRESSED_RGBA_ASTC_8x6_KHR,
    170: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR,
    171: GL_COMPRESSED_RGBA_ASTC_8x8_KHR,
    172: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR,
    173: GL_COMPRESSED_RGBA_ASTC_10x5_KHR,
    174: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR,
    175: GL_COMPRESSED_RGBA_ASTC_10x6_KHR,
    176: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR,
    177: GL_COMPRESSED_RGBA_ASTC_10x8_KHR,
    178: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR,
    179: GL_COMPRESSED_RGBA_ASTC_10x10_KHR,
    180: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR,
    181: GL_COMPRESSED_RGBA_ASTC_12x10_KHR,
    182: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR,
    183: GL_COMPRESSED_RGBA_ASTC_12x12_KHR,
    184: GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR,
    1000054e3: GL_COMPRESSED_RGBA_PVRTC_2BPPV1_IMG,
    1000054001: GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG,
    1000066e3: GL_COMPRESSED_RGBA_ASTC_4x4_KHR,
    1000066001: GL_COMPRESSED_RGBA_ASTC_5x4_KHR,
    1000066002: GL_COMPRESSED_RGBA_ASTC_5x5_KHR,
    1000066003: GL_COMPRESSED_RGBA_ASTC_6x5_KHR,
    1000066004: GL_COMPRESSED_RGBA_ASTC_6x6_KHR,
    1000066005: GL_COMPRESSED_RGBA_ASTC_8x5_KHR,
    1000066006: GL_COMPRESSED_RGBA_ASTC_8x6_KHR,
    1000066007: GL_COMPRESSED_RGBA_ASTC_8x8_KHR,
    1000066008: GL_COMPRESSED_RGBA_ASTC_10x5_KHR,
    1000066009: GL_COMPRESSED_RGBA_ASTC_10x6_KHR,
    1000066010: GL_COMPRESSED_RGBA_ASTC_10x8_KHR,
    1000066011: GL_COMPRESSED_RGBA_ASTC_10x10_KHR,
    1000066012: GL_COMPRESSED_RGBA_ASTC_12x10_KHR,
    1000066013: GL_COMPRESSED_RGBA_ASTC_12x12_KHR
  };
  function mapVkFormatToWebGL(vkFormat) {
    return VULKAN_TO_WEBGL_FORMAT_MAP[vkFormat];
  }
  function mapVkFormatToTextureFormat(vkFormat) {
    return getTextureFormatFromWebGLFormat(mapVkFormatToWebGL(vkFormat));
  }

  // src/lib/parsers/parse-ktx.ts
  var KTX2_ID2 = [
    // '´', 'K', 'T', 'X', '2', '0', 'ª', '\r', '\n', '\x1A', '\n'
    171,
    75,
    84,
    88,
    32,
    50,
    48,
    187,
    13,
    10,
    26,
    10
  ];
  function isKTX(data) {
    const id = new Uint8Array(data);
    const notKTX = id.byteLength < KTX2_ID2.length || id[0] !== KTX2_ID2[0] || // '´'
    id[1] !== KTX2_ID2[1] || // 'K'
    id[2] !== KTX2_ID2[2] || // 'T'
    id[3] !== KTX2_ID2[3] || // 'X'
    id[4] !== KTX2_ID2[4] || // ' '
    id[5] !== KTX2_ID2[5] || // '2'
    id[6] !== KTX2_ID2[6] || // '0'
    id[7] !== KTX2_ID2[7] || // 'ª'
    id[8] !== KTX2_ID2[8] || // '\r'
    id[9] !== KTX2_ID2[9] || // '\n'
    id[10] !== KTX2_ID2[10] || // '\x1A'
    id[11] !== KTX2_ID2[11];
    return !notKTX;
  }
  function parseKTX(arrayBuffer) {
    const uint8Array = new Uint8Array(arrayBuffer);
    const ktx = read(uint8Array);
    const mipMapLevels = Math.max(1, ktx.levels.length);
    const width = ktx.pixelWidth;
    const height = ktx.pixelHeight;
    const textureFormat = mapVkFormatToTextureFormat(ktx.vkFormat);
    if (textureFormat === void 0) {
      log.warn(
        `KTX2 container vkFormat ${ktx.vkFormat} does not map to a known texture format; returning texture levels without format metadata.`
      )();
    }
    return extractMipmapImages(ktx.levels, {
      mipMapLevels,
      width,
      height,
      sizeFunction: (level) => level.uncompressedByteLength,
      textureFormat
    });
  }

  // src/lib/parsers/parse-basis.ts
  var basisTranscodingLock = Promise.resolve();
  var BASIS_FORMAT_TO_OUTPUT_OPTIONS = {
    etc1: {
      basisFormat: 0,
      compressed: true,
      format: GL_COMPRESSED_RGB_ETC1_WEBGL,
      textureFormat: "etc1-rgb-unorm-webgl"
    },
    etc2: {
      basisFormat: 1,
      compressed: true,
      format: GL_COMPRESSED_RGBA8_ETC2_EAC,
      textureFormat: "etc2-rgba8unorm"
    },
    bc1: {
      basisFormat: 2,
      compressed: true,
      format: GL_COMPRESSED_RGB_S3TC_DXT1_EXT,
      textureFormat: "bc1-rgb-unorm-webgl"
    },
    bc3: {
      basisFormat: 3,
      compressed: true,
      format: GL_COMPRESSED_RGBA_S3TC_DXT5_EXT,
      textureFormat: "bc3-rgba-unorm"
    },
    bc4: {
      basisFormat: 4,
      compressed: true,
      format: GL_COMPRESSED_RED_RGTC1_EXT,
      textureFormat: "bc4-r-unorm"
    },
    bc5: {
      basisFormat: 5,
      compressed: true,
      format: GL_COMPRESSED_RED_GREEN_RGTC2_EXT,
      textureFormat: "bc5-rg-unorm"
    },
    "bc7-m6-opaque-only": {
      basisFormat: 6,
      compressed: true,
      format: GL_COMPRESSED_RGBA_BPTC_UNORM_EXT,
      textureFormat: "bc7-rgba-unorm"
    },
    "bc7-m5": {
      basisFormat: 7,
      compressed: true,
      format: GL_COMPRESSED_RGBA_BPTC_UNORM_EXT,
      textureFormat: "bc7-rgba-unorm"
    },
    "pvrtc1-4-rgb": {
      basisFormat: 8,
      compressed: true,
      format: GL_COMPRESSED_RGB_PVRTC_4BPPV1_IMG,
      textureFormat: "pvrtc-rgb4unorm-webgl"
    },
    "pvrtc1-4-rgba": {
      basisFormat: 9,
      compressed: true,
      format: GL_COMPRESSED_RGBA_PVRTC_4BPPV1_IMG,
      textureFormat: "pvrtc-rgba4unorm-webgl"
    },
    "astc-4x4": {
      basisFormat: 10,
      compressed: true,
      format: GL_COMPRESSED_RGBA_ASTC_4x4_KHR,
      textureFormat: "astc-4x4-unorm"
    },
    "atc-rgb": {
      basisFormat: 11,
      compressed: true,
      format: GL_COMPRESSED_RGB_ATC_WEBGL,
      textureFormat: "atc-rgb-unorm-webgl"
    },
    "atc-rgba-interpolated-alpha": {
      basisFormat: 12,
      compressed: true,
      format: GL_COMPRESSED_RGBA_ATC_INTERPOLATED_ALPHA_WEBGL,
      textureFormat: "atc-rgbai-unorm-webgl"
    },
    rgba32: {
      basisFormat: 13,
      compressed: false,
      format: GL_RGBA8,
      textureFormat: "rgba8unorm"
    },
    rgb565: {
      basisFormat: 14,
      compressed: false,
      format: GL_RGB565,
      textureFormat: "rgb565unorm-webgl"
    },
    bgr565: {
      basisFormat: 15,
      compressed: false,
      format: GL_RGB565,
      textureFormat: "rgb565unorm-webgl"
    },
    rgba4444: {
      basisFormat: 16,
      compressed: false,
      format: GL_RGBA4,
      textureFormat: "rgba4unorm-webgl"
    }
  };
  var BASIS_FORMATS = Object.freeze(
    Object.keys(BASIS_FORMAT_TO_OUTPUT_OPTIONS)
  );
  async function withBasisTranscodingLock(transcode) {
    const previousLock = basisTranscodingLock;
    let releaseLock;
    basisTranscodingLock = new Promise((resolve2) => {
      releaseLock = resolve2;
    });
    await previousLock;
    try {
      return await transcode();
    } finally {
      releaseLock();
    }
  }
  async function parseBasis(data, options = {}) {
    const loadLibraryOptions = extractLoadLibraryOptions(options);
    return await withBasisTranscodingLock(async () => {
      if (!options.basis?.containerFormat || options.basis.containerFormat === "auto") {
        if (isKTX(data)) {
          const fileConstructors = await loadBasisEncoderModule(loadLibraryOptions);
          return parseKTX2File(fileConstructors.KTX2File, data, options);
        }
        const { BasisFile } = await loadBasisTranscoderModule(loadLibraryOptions);
        return parseBasisFile(BasisFile, data, options);
      }
      switch (options.basis.module) {
        case "encoder":
          const fileConstructors = await loadBasisEncoderModule(loadLibraryOptions);
          switch (options.basis.containerFormat) {
            case "ktx2":
              return parseKTX2File(fileConstructors.KTX2File, data, options);
            case "basis":
            default:
              return parseBasisFile(fileConstructors.BasisFile, data, options);
          }
        case "transcoder":
        default:
          const { BasisFile } = await loadBasisTranscoderModule(loadLibraryOptions);
          return parseBasisFile(BasisFile, data, options);
      }
    });
  }
  function parseBasisFile(BasisFile, data, options) {
    const basisFile = new BasisFile(new Uint8Array(data));
    try {
      if (!basisFile.startTranscoding()) {
        throw new Error("Failed to start basis transcoding");
      }
      const imageCount = basisFile.getNumImages();
      const images = [];
      for (let imageIndex = 0; imageIndex < imageCount; imageIndex++) {
        const levelsCount = basisFile.getNumLevels(imageIndex);
        const levels = [];
        for (let levelIndex = 0; levelIndex < levelsCount; levelIndex++) {
          levels.push(transcodeImage(basisFile, imageIndex, levelIndex, options));
        }
        images.push(levels);
      }
      return images;
    } finally {
      basisFile.close();
      basisFile.delete();
    }
  }
  function transcodeImage(basisFile, imageIndex, levelIndex, options) {
    const width = basisFile.getImageWidth(imageIndex, levelIndex);
    const height = basisFile.getImageHeight(imageIndex, levelIndex);
    const hasAlpha = basisFile.getHasAlpha(
      /* imageIndex, levelIndex */
    );
    const { compressed, format, basisFormat, textureFormat } = getBasisOptions(options, hasAlpha);
    const decodedSize = basisFile.getImageTranscodedSizeInBytes(imageIndex, levelIndex, basisFormat);
    const decodedData = new Uint8Array(decodedSize);
    if (!basisFile.transcodeImage(decodedData, imageIndex, levelIndex, basisFormat, 0, 0)) {
      throw new Error("failed to start Basis transcoding");
    }
    return {
      // standard loaders.gl image category payload
      shape: "texture-level",
      width,
      height,
      data: decodedData,
      compressed,
      ...format !== void 0 ? { format } : {},
      ...textureFormat !== void 0 ? { textureFormat } : {},
      // Additional fields
      // Add levelSize field.
      hasAlpha
    };
  }
  function parseKTX2File(KTX2File, data, options) {
    const ktx2File = new KTX2File(new Uint8Array(data));
    try {
      if (!ktx2File.startTranscoding()) {
        throw new Error("failed to start KTX2 transcoding");
      }
      const levelsCount = ktx2File.getLevels();
      const levels = [];
      for (let levelIndex = 0; levelIndex < levelsCount; levelIndex++) {
        levels.push(transcodeKTX2Image(ktx2File, levelIndex, options));
      }
      return [levels];
    } finally {
      ktx2File.close();
      ktx2File.delete();
    }
  }
  function transcodeKTX2Image(ktx2File, levelIndex, options) {
    const { alphaFlag, height, width } = ktx2File.getImageLevelInfo(levelIndex, 0, 0);
    const { compressed, format, basisFormat, textureFormat } = getBasisOptions(options, alphaFlag);
    const decodedSize = ktx2File.getImageTranscodedSizeInBytes(
      levelIndex,
      0,
      0,
      basisFormat
    );
    const decodedData = new Uint8Array(decodedSize);
    if (!ktx2File.transcodeImage(
      decodedData,
      levelIndex,
      0,
      0,
      basisFormat,
      0,
      -1,
      -1
      /* channel1 */
    )) {
      throw new Error("Failed to transcode KTX2 image");
    }
    return {
      // standard loaders.gl image category payload
      shape: "texture-level",
      width,
      height,
      data: decodedData,
      compressed,
      ...format !== void 0 ? { format } : {},
      ...textureFormat !== void 0 ? { textureFormat } : {},
      // Additional fields
      levelSize: decodedSize,
      hasAlpha: alphaFlag
    };
  }
  function getBasisOptions(options, hasAlpha) {
    let format = options.basis?.format || "auto";
    if (format === "auto") {
      format = options.basis?.supportedTextureFormats ? selectSupportedBasisFormat(options.basis.supportedTextureFormats) : selectSupportedBasisFormat();
    }
    if (typeof format === "object") {
      format = hasAlpha ? format.alpha : format.noAlpha;
    }
    const normalizedFormat = format.toLowerCase();
    const basisOutputOptions = BASIS_FORMAT_TO_OUTPUT_OPTIONS[normalizedFormat];
    if (!basisOutputOptions) {
      throw new Error(`Unknown Basis format ${format}`);
    }
    return basisOutputOptions;
  }
  function selectSupportedBasisFormat(supportedTextureFormats = detectSupportedTextureFormats()) {
    const textureFormats2 = new Set(supportedTextureFormats);
    if (hasSupportedTextureFormat(textureFormats2, ["astc-4x4-unorm", "astc-4x4-unorm-srgb"])) {
      return "astc-4x4";
    } else if (hasSupportedTextureFormat(textureFormats2, ["bc7-rgba-unorm", "bc7-rgba-unorm-srgb"])) {
      return {
        alpha: "bc7-m5",
        noAlpha: "bc7-m6-opaque-only"
      };
    } else if (hasSupportedTextureFormat(textureFormats2, [
      "bc1-rgb-unorm-webgl",
      "bc1-rgb-unorm-srgb-webgl",
      "bc1-rgba-unorm",
      "bc1-rgba-unorm-srgb",
      "bc2-rgba-unorm",
      "bc2-rgba-unorm-srgb",
      "bc3-rgba-unorm",
      "bc3-rgba-unorm-srgb"
    ])) {
      return {
        alpha: "bc3",
        noAlpha: "bc1"
      };
    } else if (hasSupportedTextureFormat(textureFormats2, [
      "pvrtc-rgb4unorm-webgl",
      "pvrtc-rgba4unorm-webgl",
      "pvrtc-rgb2unorm-webgl",
      "pvrtc-rgba2unorm-webgl"
    ])) {
      return {
        alpha: "pvrtc1-4-rgba",
        noAlpha: "pvrtc1-4-rgb"
      };
    } else if (hasSupportedTextureFormat(textureFormats2, [
      "etc2-rgb8unorm",
      "etc2-rgb8unorm-srgb",
      "etc2-rgb8a1unorm",
      "etc2-rgb8a1unorm-srgb",
      "etc2-rgba8unorm",
      "etc2-rgba8unorm-srgb",
      "eac-r11unorm",
      "eac-r11snorm",
      "eac-rg11unorm",
      "eac-rg11snorm"
    ])) {
      return "etc2";
    } else if (textureFormats2.has("etc1-rgb-unorm-webgl")) {
      return "etc1";
    } else if (hasSupportedTextureFormat(textureFormats2, [
      "atc-rgb-unorm-webgl",
      "atc-rgba-unorm-webgl",
      "atc-rgbai-unorm-webgl"
    ])) {
      return {
        alpha: "atc-rgba-interpolated-alpha",
        noAlpha: "atc-rgb"
      };
    }
    return "rgb565";
  }
  function hasSupportedTextureFormat(supportedTextureFormats, candidateTextureFormats) {
    return candidateTextureFormats.some(
      (textureFormat) => supportedTextureFormats.has(textureFormat)
    );
  }

  // src/basis-loader.ts
  var BasisWorkerLoader = {
    dataType: null,
    batchType: null,
    name: "Basis",
    id: "basis",
    module: "textures",
    version: VERSION,
    worker: true,
    extensions: ["basis", "ktx2"],
    mimeTypes: ["application/octet-stream", "image/ktx2"],
    tests: ["sB"],
    binary: true,
    options: {
      basis: {
        format: "auto",
        containerFormat: "auto",
        module: "transcoder"
      }
    }
  };
  var BasisLoader = {
    ...BasisWorkerLoader,
    parse: parseBasis
  };

  // src/lib/parsers/parse-dds.ts
  var DDS_CONSTANTS = {
    MAGIC_NUMBER: 542327876,
    HEADER_LENGTH: 31,
    MAGIC_NUMBER_INDEX: 0,
    HEADER_SIZE_INDEX: 1,
    HEADER_FLAGS_INDEX: 2,
    HEADER_HEIGHT_INDEX: 3,
    HEADER_WIDTH_INDEX: 4,
    MIPMAPCOUNT_INDEX: 7,
    HEADER_PF_FLAGS_INDEX: 20,
    HEADER_PF_FOURCC_INDEX: 21,
    DDSD_MIPMAPCOUNT: 131072,
    DDPF_FOURCC: 4
  };
  var DDS_TEXTURE_FORMATS = {
    DXT1: "bc1-rgb-unorm-webgl",
    DXT3: "bc2-rgba-unorm",
    DXT5: "bc3-rgba-unorm",
    "ATC ": "atc-rgb-unorm-webgl",
    ATCA: "atc-rgba-unorm-webgl",
    ATCI: "atc-rgbai-unorm-webgl"
  };
  var getATCLevelSize = getDxt1LevelSize;
  var getATCALevelSize = getDxtXLevelSize;
  var getATCILevelSize = getDxtXLevelSize;
  var DDS_SIZE_FUNCTIONS = {
    DXT1: getDxt1LevelSize,
    DXT3: getDxtXLevelSize,
    DXT5: getDxtXLevelSize,
    "ATC ": getATCLevelSize,
    ATCA: getATCALevelSize,
    ATCI: getATCILevelSize
  };
  function isDDS(data) {
    const header = new Uint32Array(data, 0, DDS_CONSTANTS.HEADER_LENGTH);
    const magic = header[DDS_CONSTANTS.MAGIC_NUMBER_INDEX];
    return magic === DDS_CONSTANTS.MAGIC_NUMBER;
  }
  function parseDDS(data) {
    const header = new Int32Array(data, 0, DDS_CONSTANTS.HEADER_LENGTH);
    const pixelFormatNumber = header[DDS_CONSTANTS.HEADER_PF_FOURCC_INDEX];
    assert2(
      Boolean(header[DDS_CONSTANTS.HEADER_PF_FLAGS_INDEX] & DDS_CONSTANTS.DDPF_FOURCC),
      "DDS: Unsupported format, must contain a FourCC code"
    );
    const fourCC = int32ToFourCC(pixelFormatNumber);
    const textureFormat = DDS_TEXTURE_FORMATS[fourCC];
    const sizeFunction = DDS_SIZE_FUNCTIONS[fourCC];
    assert2(textureFormat && sizeFunction, `DDS: Unknown pixel format ${pixelFormatNumber}`);
    let mipMapLevels = 1;
    if (header[DDS_CONSTANTS.HEADER_FLAGS_INDEX] & DDS_CONSTANTS.DDSD_MIPMAPCOUNT) {
      mipMapLevels = Math.max(1, header[DDS_CONSTANTS.MIPMAPCOUNT_INDEX]);
    }
    const width = header[DDS_CONSTANTS.HEADER_WIDTH_INDEX];
    const height = header[DDS_CONSTANTS.HEADER_HEIGHT_INDEX];
    const dataOffset = header[DDS_CONSTANTS.HEADER_SIZE_INDEX] + 4;
    const image = new Uint8Array(data, dataOffset);
    return extractMipmapImages(image, {
      mipMapLevels,
      width,
      height,
      sizeFunction,
      textureFormat
    });
  }
  function getDxt1LevelSize(width, height) {
    return (width + 3 >> 2) * (height + 3 >> 2) * 8;
  }
  function getDxtXLevelSize(width, height) {
    return (width + 3 >> 2) * (height + 3 >> 2) * 16;
  }
  function int32ToFourCC(value) {
    return String.fromCharCode(
      value & 255,
      value >> 8 & 255,
      value >> 16 & 255,
      value >> 24 & 255
    );
  }

  // src/lib/parsers/parse-pvr.ts
  var PVR_CONSTANTS = {
    MAGIC_NUMBER: 55727696,
    MAGIC_NUMBER_EXTRA: 1347834371,
    HEADER_LENGTH: 13,
    HEADER_SIZE: 52,
    MAGIC_NUMBER_INDEX: 0,
    PIXEL_FORMAT_INDEX: 2,
    COLOUR_SPACE_INDEX: 4,
    HEIGHT_INDEX: 6,
    WIDTH_INDEX: 7,
    MIPMAPCOUNT_INDEX: 11,
    METADATA_SIZE_INDEX: 12
  };
  var PVR_TEXTURE_FORMATS = {
    0: ["pvrtc-rgb2unorm-webgl"],
    1: ["pvrtc-rgba2unorm-webgl"],
    2: ["pvrtc-rgb4unorm-webgl"],
    3: ["pvrtc-rgba4unorm-webgl"],
    6: ["etc1-rgb-unorm-webgl"],
    7: ["bc1-rgb-unorm-webgl"],
    9: ["bc2-rgba-unorm"],
    11: ["bc3-rgba-unorm"],
    22: ["etc2-rgb8unorm"],
    23: ["etc2-rgba8unorm"],
    24: ["etc2-rgb8a1unorm"],
    25: ["eac-r11unorm"],
    26: ["eac-rg11unorm"],
    27: ["astc-4x4-unorm", "astc-4x4-unorm-srgb"],
    28: ["astc-5x4-unorm", "astc-5x4-unorm-srgb"],
    29: ["astc-5x5-unorm", "astc-5x5-unorm-srgb"],
    30: ["astc-6x5-unorm", "astc-6x5-unorm-srgb"],
    31: ["astc-6x6-unorm", "astc-6x6-unorm-srgb"],
    32: ["astc-8x5-unorm", "astc-8x5-unorm-srgb"],
    33: ["astc-8x6-unorm", "astc-8x6-unorm-srgb"],
    34: ["astc-8x8-unorm", "astc-8x8-unorm-srgb"],
    35: ["astc-10x5-unorm", "astc-10x5-unorm-srgb"],
    36: ["astc-10x6-unorm", "astc-10x6-unorm-srgb"],
    37: ["astc-10x8-unorm", "astc-10x8-unorm-srgb"],
    38: ["astc-10x10-unorm", "astc-10x10-unorm-srgb"],
    39: ["astc-12x10-unorm", "astc-12x10-unorm-srgb"],
    40: ["astc-12x12-unorm", "astc-12x12-unorm-srgb"]
  };
  var PVR_SIZE_FUNCTIONS = {
    0: pvrtc2bppSize,
    1: pvrtc2bppSize,
    2: pvrtc4bppSize,
    3: pvrtc4bppSize,
    6: dxtEtcSmallSize,
    7: dxtEtcSmallSize,
    9: dxtEtcAstcBigSize,
    11: dxtEtcAstcBigSize,
    22: dxtEtcSmallSize,
    23: dxtEtcAstcBigSize,
    24: dxtEtcSmallSize,
    25: dxtEtcSmallSize,
    26: dxtEtcAstcBigSize,
    27: dxtEtcAstcBigSize,
    28: atc5x4Size,
    29: atc5x5Size,
    30: atc6x5Size,
    31: atc6x6Size,
    32: atc8x5Size,
    33: atc8x6Size,
    34: atc8x8Size,
    35: atc10x5Size,
    36: atc10x6Size,
    37: atc10x8Size,
    38: atc10x10Size,
    39: atc12x10Size,
    40: atc12x12Size
  };
  function isPVR(data) {
    const header = new Uint32Array(data, 0, PVR_CONSTANTS.HEADER_LENGTH);
    const version2 = header[PVR_CONSTANTS.MAGIC_NUMBER_INDEX];
    return version2 === PVR_CONSTANTS.MAGIC_NUMBER || version2 === PVR_CONSTANTS.MAGIC_NUMBER_EXTRA;
  }
  function parsePVR(data) {
    const header = new Uint32Array(data, 0, PVR_CONSTANTS.HEADER_LENGTH);
    const pvrFormat = header[PVR_CONSTANTS.PIXEL_FORMAT_INDEX];
    const colourSpace = header[PVR_CONSTANTS.COLOUR_SPACE_INDEX];
    const textureFormats2 = PVR_TEXTURE_FORMATS[pvrFormat] || [];
    const textureFormat = textureFormats2.length > 1 && colourSpace ? textureFormats2[1] : textureFormats2[0];
    const sizeFunction = PVR_SIZE_FUNCTIONS[pvrFormat];
    const mipMapLevels = header[PVR_CONSTANTS.MIPMAPCOUNT_INDEX];
    const width = header[PVR_CONSTANTS.WIDTH_INDEX];
    const height = header[PVR_CONSTANTS.HEIGHT_INDEX];
    const dataOffset = PVR_CONSTANTS.HEADER_SIZE + header[PVR_CONSTANTS.METADATA_SIZE_INDEX];
    const image = new Uint8Array(data, dataOffset);
    return extractMipmapImages(image, {
      mipMapLevels,
      width,
      height,
      sizeFunction,
      textureFormat
    });
  }
  function pvrtc2bppSize(width, height) {
    width = Math.max(width, 16);
    height = Math.max(height, 8);
    return width * height / 4;
  }
  function pvrtc4bppSize(width, height) {
    width = Math.max(width, 8);
    height = Math.max(height, 8);
    return width * height / 2;
  }
  function dxtEtcSmallSize(width, height) {
    return Math.floor((width + 3) / 4) * Math.floor((height + 3) / 4) * 8;
  }
  function dxtEtcAstcBigSize(width, height) {
    return Math.floor((width + 3) / 4) * Math.floor((height + 3) / 4) * 16;
  }
  function atc5x4Size(width, height) {
    return Math.floor((width + 4) / 5) * Math.floor((height + 3) / 4) * 16;
  }
  function atc5x5Size(width, height) {
    return Math.floor((width + 4) / 5) * Math.floor((height + 4) / 5) * 16;
  }
  function atc6x5Size(width, height) {
    return Math.floor((width + 5) / 6) * Math.floor((height + 4) / 5) * 16;
  }
  function atc6x6Size(width, height) {
    return Math.floor((width + 5) / 6) * Math.floor((height + 5) / 6) * 16;
  }
  function atc8x5Size(width, height) {
    return Math.floor((width + 7) / 8) * Math.floor((height + 4) / 5) * 16;
  }
  function atc8x6Size(width, height) {
    return Math.floor((width + 7) / 8) * Math.floor((height + 5) / 6) * 16;
  }
  function atc8x8Size(width, height) {
    return Math.floor((width + 7) / 8) * Math.floor((height + 7) / 8) * 16;
  }
  function atc10x5Size(width, height) {
    return Math.floor((width + 9) / 10) * Math.floor((height + 4) / 5) * 16;
  }
  function atc10x6Size(width, height) {
    return Math.floor((width + 9) / 10) * Math.floor((height + 5) / 6) * 16;
  }
  function atc10x8Size(width, height) {
    return Math.floor((width + 9) / 10) * Math.floor((height + 7) / 8) * 16;
  }
  function atc10x10Size(width, height) {
    return Math.floor((width + 9) / 10) * Math.floor((height + 9) / 10) * 16;
  }
  function atc12x10Size(width, height) {
    return Math.floor((width + 11) / 12) * Math.floor((height + 9) / 10) * 16;
  }
  function atc12x12Size(width, height) {
    return Math.floor((width + 11) / 12) * Math.floor((height + 11) / 12) * 16;
  }

  // src/lib/parsers/parse-compressed-texture.ts
  function parseCompressedTexture(data) {
    if (isKTX(data)) {
      return parseKTX(data);
    }
    if (isDDS(data)) {
      return parseDDS(data);
    }
    if (isPVR(data)) {
      return parsePVR(data);
    }
    throw new Error("Texture container format not recognized");
  }

  // src/compressed-texture-loader.ts
  var CompressedTextureWorkerLoader = {
    dataType: null,
    batchType: null,
    name: "Texture Containers",
    id: "compressed-texture",
    module: "textures",
    version: VERSION,
    worker: true,
    extensions: [
      "ktx",
      "ktx2",
      "dds",
      // WEBGL_compressed_texture_s3tc, WEBGL_compressed_texture_atc
      "pvr"
      // WEBGL_compressed_texture_pvrtc
    ],
    mimeTypes: [
      "image/ktx2",
      "image/ktx",
      "image/vnd-ms.dds",
      "image/x-dds",
      "application/octet-stream"
    ],
    binary: true,
    options: {
      "compressed-texture": {
        useBasis: false
      }
    }
  };
  var CompressedTextureLoader = {
    ...CompressedTextureWorkerLoader,
    parse: async (arrayBuffer, options) => {
      options = { ...options };
      if (options?.["compressed-texture"]?.useBasis) {
        options.basis = {
          format: {
            alpha: "BC3",
            noAlpha: "BC1"
          },
          ...options.basis,
          containerFormat: "ktx2",
          module: "encoder"
        };
        const result = await parseBasis(arrayBuffer, options);
        return result[0];
      }
      return parseCompressedTexture(arrayBuffer);
    }
  };

  // src/crunch-loader.ts
  var CrunchLoader = {
    dataType: null,
    batchType: null,
    id: "crunch",
    name: "Crunch",
    module: "textures",
    version: VERSION,
    worker: true,
    extensions: ["crn"],
    mimeTypes: ["image/crn", "image/x-crn", "application/octet-stream"],
    binary: true,
    options: {
      crunch: {}
    }
  };

  // src/lib/parsers/parse-hdr.ts
  var HDR_MAGIC_HEADERS = ["#?RADIANCE", "#?RGBE"];
  var HDR_FORMAT = "32-bit_rle_rgbe";
  function isHDR(arrayBuffer) {
    const state = {
      data: new Uint8Array(arrayBuffer),
      offset: 0
    };
    const firstLine = readLine(state);
    return firstLine ? HDR_MAGIC_HEADERS.includes(firstLine) : false;
  }
  function parseHDR(arrayBuffer) {
    const state = {
      data: new Uint8Array(arrayBuffer),
      offset: 0
    };
    const header = readHeader(state);
    const { width, height } = header;
    const rgbeData = readPixels(state, header);
    const data = convertRGBEToFloat(rgbeData);
    const level = {
      shape: "texture-level",
      compressed: false,
      width,
      height,
      data,
      levelSize: data.byteLength,
      format: GL_RGBA32F,
      textureFormat: "rgba32float"
    };
    return {
      shape: "texture",
      type: "2d",
      format: "rgba32float",
      ...header.metadata ? { metadata: header.metadata } : {},
      data: [level]
    };
  }
  function readHeader(state) {
    const magicHeader = readLine(state);
    if (!magicHeader || !HDR_MAGIC_HEADERS.includes(magicHeader)) {
      throw new Error("RadianceHDRLoader: bad initial token");
    }
    let hasFormat = false;
    const metadata = {};
    while (state.offset < state.data.length) {
      const line = readLine(state);
      if (line === null) {
        break;
      }
      if (!line || line.startsWith("#")) {
        continue;
      }
      if (line.startsWith("FORMAT=")) {
        hasFormat = line.slice("FORMAT=".length) === HDR_FORMAT;
        if (!hasFormat) {
          throw new Error("RadianceHDRLoader: unsupported format specifier");
        }
        continue;
      }
      parseMetadataLine(metadata, line);
      const dimensions = parseDimensions(line);
      if (dimensions) {
        if (!hasFormat) {
          throw new Error("RadianceHDRLoader: missing format specifier");
        }
        return {
          ...dimensions,
          ...hasMetadata(metadata) ? { metadata } : {}
        };
      }
    }
    if (!hasFormat) {
      throw new Error("RadianceHDRLoader: missing format specifier");
    }
    throw new Error("RadianceHDRLoader: missing image size specifier");
  }
  function parseDimensions(line) {
    const match = line.match(/^([+-])([YX])\s+(\d+)\s+([+-])([YX])\s+(\d+)$/);
    if (!match) {
      return null;
    }
    const majorSign = match[1] === "+" ? 1 : -1;
    const majorAxis = match[2];
    const majorLength = Number(match[3]);
    const minorSign = match[4] === "+" ? 1 : -1;
    const minorAxis = match[5];
    const minorLength = Number(match[6]);
    if (majorAxis === minorAxis) {
      throw new Error("RadianceHDRLoader: invalid image dimensions");
    }
    const width = majorAxis === "X" ? majorLength : minorLength;
    const height = majorAxis === "Y" ? majorLength : minorLength;
    if (!Number.isFinite(width) || !Number.isFinite(height) || width <= 0 || height <= 0) {
      throw new Error("RadianceHDRLoader: invalid image dimensions");
    }
    return { width, height, majorAxis, majorSign, minorAxis, minorSign };
  }
  function readPixels(state, header) {
    const { width, height } = header;
    const pixelCount = width * height;
    const flatByteLength = pixelCount * 4;
    const scanlineLength = header.minorAxis === "X" ? width : height;
    const scanlineCount = header.majorAxis === "Y" ? height : width;
    if (scanlineLength < 8 || scanlineLength > 32767) {
      return reorderPixels(readFlatPixels(state, flatByteLength), header);
    }
    if (state.offset + 4 > state.data.length) {
      throw new Error("RadianceHDRLoader: unexpected end of file");
    }
    const data = state.data;
    const isRunLengthEncoded = data[state.offset] === 2 && data[state.offset + 1] === 2 && !(data[state.offset + 2] & 128);
    if (!isRunLengthEncoded) {
      return reorderPixels(readFlatPixels(state, flatByteLength), header);
    }
    const scanlineWidth = data[state.offset + 2] << 8 | data[state.offset + 3];
    if (scanlineWidth !== scanlineLength) {
      return reorderPixels(readFlatPixels(state, flatByteLength), header);
    }
    const pixels = new Uint8Array(flatByteLength);
    const scanlineBuffer = new Uint8Array(scanlineLength * 4);
    for (let scanlineIndex = 0; scanlineIndex < scanlineCount; scanlineIndex++) {
      if (state.offset + 4 > data.length) {
        throw new Error("RadianceHDRLoader: unexpected end of file");
      }
      const red = data[state.offset++];
      const green = data[state.offset++];
      const blue = data[state.offset++];
      const exponent = data[state.offset++];
      if (red !== 2 || green !== 2 || (blue << 8 | exponent) !== scanlineLength) {
        throw new Error("RadianceHDRLoader: bad rgbe scanline format");
      }
      for (let channelIndex = 0; channelIndex < 4; channelIndex++) {
        const channelOffset = channelIndex * scanlineLength;
        const channelEnd = channelOffset + scanlineLength;
        let pixelOffset = channelOffset;
        while (pixelOffset < channelEnd) {
          if (state.offset >= data.length) {
            throw new Error("RadianceHDRLoader: unexpected end of file");
          }
          let count = data[state.offset++];
          if (count > 128) {
            count -= 128;
            if (count === 0 || pixelOffset + count > channelEnd || state.offset >= data.length) {
              throw new Error("RadianceHDRLoader: bad scanline data");
            }
            const value = data[state.offset++];
            scanlineBuffer.fill(value, pixelOffset, pixelOffset + count);
            pixelOffset += count;
            continue;
          }
          if (count === 0 || pixelOffset + count > channelEnd || state.offset + count > data.length) {
            throw new Error("RadianceHDRLoader: bad scanline data");
          }
          scanlineBuffer.set(data.subarray(state.offset, state.offset + count), pixelOffset);
          pixelOffset += count;
          state.offset += count;
        }
      }
      for (let pixelIndex = 0; pixelIndex < scanlineLength; pixelIndex++) {
        const outputOffset = getOutputOffset(header, scanlineIndex, pixelIndex);
        pixels[outputOffset] = scanlineBuffer[pixelIndex];
        pixels[outputOffset + 1] = scanlineBuffer[pixelIndex + scanlineLength];
        pixels[outputOffset + 2] = scanlineBuffer[pixelIndex + scanlineLength * 2];
        pixels[outputOffset + 3] = scanlineBuffer[pixelIndex + scanlineLength * 3];
      }
    }
    return pixels;
  }
  function reorderPixels(data, header) {
    const pixels = new Uint8Array(data.length);
    const scanlineLength = header.minorAxis === "X" ? header.width : header.height;
    const scanlineCount = header.majorAxis === "Y" ? header.height : header.width;
    for (let scanlineIndex = 0; scanlineIndex < scanlineCount; scanlineIndex++) {
      for (let pixelIndex = 0; pixelIndex < scanlineLength; pixelIndex++) {
        const sourceOffset = (scanlineIndex * scanlineLength + pixelIndex) * 4;
        const outputOffset = getOutputOffset(header, scanlineIndex, pixelIndex);
        pixels[outputOffset] = data[sourceOffset];
        pixels[outputOffset + 1] = data[sourceOffset + 1];
        pixels[outputOffset + 2] = data[sourceOffset + 2];
        pixels[outputOffset + 3] = data[sourceOffset + 3];
      }
    }
    return pixels;
  }
  function getOutputOffset(header, scanlineIndex, pixelIndex) {
    const majorCoordinate = getCoordinate(
      header.majorAxis === "X" ? header.width : header.height,
      header.majorSign,
      scanlineIndex
    );
    const minorCoordinate = getCoordinate(
      header.minorAxis === "X" ? header.width : header.height,
      header.minorSign,
      pixelIndex
    );
    const x = header.majorAxis === "X" ? majorCoordinate : minorCoordinate;
    const y = header.majorAxis === "Y" ? majorCoordinate : minorCoordinate;
    return ((header.height - 1 - y) * header.width + x) * 4;
  }
  function getCoordinate(length, sign, index) {
    return sign === 1 ? index : length - 1 - index;
  }
  function readFlatPixels(state, byteLength) {
    if (state.offset + byteLength > state.data.length) {
      throw new Error("RadianceHDRLoader: unexpected end of file");
    }
    const pixels = state.data.slice(state.offset, state.offset + byteLength);
    state.offset += byteLength;
    return pixels;
  }
  function convertRGBEToFloat(data) {
    const floatData = new Float32Array(data.length);
    for (let sourceOffset = 0; sourceOffset < data.length; sourceOffset += 4) {
      const exponent = data[sourceOffset + 3];
      const destinationOffset = sourceOffset;
      if (exponent > 0) {
        const scale = Math.pow(2, exponent - 128) / 255;
        floatData[destinationOffset] = data[sourceOffset] * scale;
        floatData[destinationOffset + 1] = data[sourceOffset + 1] * scale;
        floatData[destinationOffset + 2] = data[sourceOffset + 2] * scale;
      }
      floatData[destinationOffset + 3] = 1;
    }
    return floatData;
  }
  function readLine(state) {
    if (state.offset >= state.data.length) {
      return null;
    }
    const lineStart = state.offset;
    while (state.offset < state.data.length) {
      const byte = state.data[state.offset++];
      if (byte === 10) {
        const line2 = decodeASCII(state.data.subarray(lineStart, state.offset - 1));
        return line2.endsWith("\r") ? line2.slice(0, -1) : line2;
      }
    }
    const line = decodeASCII(state.data.subarray(lineStart, state.offset));
    return line.endsWith("\r") ? line.slice(0, -1) : line;
  }
  function decodeASCII(data) {
    let line = "";
    for (const byte of data) {
      line += String.fromCharCode(byte);
    }
    return line;
  }
  function parseMetadataLine(metadata, line) {
    if (line.startsWith("COLORCORR=")) {
      const values = parseNumberList(line.slice("COLORCORR=".length), 3);
      if (values) {
        metadata.colorCorrection = values;
      }
      return;
    }
    if (line.startsWith("EXPOSURE=")) {
      const value = parseNumber(line.slice("EXPOSURE=".length));
      if (value !== null) {
        metadata.exposure = value;
      }
      return;
    }
    if (line.startsWith("GAMMA=")) {
      const value = parseNumber(line.slice("GAMMA=".length));
      if (value !== null) {
        metadata.gamma = value;
      }
      return;
    }
    if (line.startsWith("PIXASPECT=")) {
      const value = parseNumber(line.slice("PIXASPECT=".length));
      if (value !== null) {
        metadata.pixelAspectRatio = value;
      }
      return;
    }
    if (line.startsWith("PRIMARIES=")) {
      const values = parseNumberList(line.slice("PRIMARIES=".length), 8);
      if (values) {
        metadata.primaries = values;
      }
      return;
    }
    if (line.startsWith("SOFTWARE=")) {
      metadata.software = line.slice("SOFTWARE=".length).trim();
      return;
    }
    if (line.startsWith("VIEW=")) {
      metadata.view = line.slice("VIEW=".length).trim();
    }
  }
  function parseNumber(text) {
    const value = Number(text.trim());
    return Number.isFinite(value) ? value : null;
  }
  function parseNumberList(text, count) {
    const values = text.trim().split(/\s+/).map((value) => Number(value));
    if (values.length !== count || values.some((value) => !Number.isFinite(value))) {
      return null;
    }
    return values;
  }
  function hasMetadata(metadata) {
    return Object.keys(metadata).length > 0;
  }

  // src/radiance-hdr-loader.ts
  var RadianceHDRLoader = {
    dataType: null,
    batchType: null,
    name: "Radiance HDR",
    id: "hdr",
    module: "textures",
    version: VERSION,
    extensions: ["hdr"],
    mimeTypes: ["image/vnd.radiance", "image/x-hdr", "application/octet-stream"],
    binary: true,
    tests: [isHDR],
    options: {
      hdr: {}
    },
    parseSync: parseHDR,
    parse: async (arrayBuffer) => parseHDR(arrayBuffer)
  };

  // src/lib/parsers/parse-npy.ts
  var a = new Uint32Array([305419896]);
  var b = new Uint8Array(a.buffer, a.byteOffset, a.byteLength);
  var isLittleEndian = !(b[0] === 18);
  var LITTLE_ENDIAN_OS = isLittleEndian;
  var DTYPES = {
    u1: Uint8Array,
    i1: Int8Array,
    u2: Uint16Array,
    i2: Int16Array,
    u4: Uint32Array,
    i4: Int32Array,
    f4: Float32Array,
    f8: Float64Array
  };
  function parseNPY(arrayBuffer, options) {
    const view = new DataView(arrayBuffer);
    const { header, headerEndOffset } = parseHeader(view);
    const numpyType = header.descr;
    const ArrayType = DTYPES[numpyType.slice(1, 3)];
    if (!ArrayType) {
      throw new Error(`Unimplemented type ${numpyType}`);
    }
    const nArrayElements = header.shape?.reduce((a2, b2) => a2 * b2);
    const arrayByteLength = nArrayElements * ArrayType.BYTES_PER_ELEMENT;
    if (arrayBuffer.byteLength < headerEndOffset + arrayByteLength) {
      throw new Error("Buffer overflow");
    }
    const data = new ArrayType(arrayBuffer.slice(headerEndOffset, headerEndOffset + arrayByteLength));
    if (numpyType[0] === ">" && LITTLE_ENDIAN_OS || numpyType[0] === "<" && !LITTLE_ENDIAN_OS) {
      throw new Error("Incorrect endianness");
    }
    return {
      data,
      header
    };
  }
  function parseHeader(view) {
    const majorVersion = view.getUint8(6);
    let offset = 8;
    let headerLength;
    if (majorVersion >= 2) {
      headerLength = view.getUint32(offset, true);
      offset += 4;
    } else {
      headerLength = view.getUint16(offset, true);
      offset += 2;
    }
    const encoding = majorVersion <= 2 ? "latin1" : "utf-8";
    const decoder = new TextDecoder(encoding);
    const headerArray = new Uint8Array(view.buffer, offset, headerLength);
    const headerText = decoder.decode(headerArray);
    offset += headerLength;
    const header = JSON.parse(
      headerText.replace(/'/g, '"').replace("False", "false").replace("(", "[").replace(/,*\),*/g, "]")
    );
    return { header, headerEndOffset: offset };
  }

  // src/npy-loader.ts
  var NPY_MAGIC_NUMBER = new Uint8Array([147, 78, 85, 77, 80, 89]);
  var NPYWorkerLoader = {
    dataType: null,
    batchType: null,
    name: "NPY",
    id: "npy",
    module: "textures",
    version: VERSION,
    worker: true,
    extensions: ["npy"],
    mimeTypes: [],
    tests: [NPY_MAGIC_NUMBER.buffer],
    options: {
      npy: {}
    }
  };
  var NPYLoader = {
    ...NPYWorkerLoader,
    parseSync: parseNPY,
    parse: async (arrayBuffer, options) => parseNPY(arrayBuffer, options)
  };

  // ../images/src/lib/utils/version.ts
  var VERSION5 = typeof __VERSION__ !== "undefined" ? __VERSION__ : "latest";

  // ../images/src/lib/category-api/image-type.ts
  var parseImageNode = globalThis.loaders?.parseImageNode;
  var IMAGE_SUPPORTED = typeof Image !== "undefined";
  var IMAGE_BITMAP_SUPPORTED = typeof ImageBitmap !== "undefined";
  var NODE_IMAGE_SUPPORTED = Boolean(parseImageNode);
  var DATA_SUPPORTED = isBrowser2 ? true : NODE_IMAGE_SUPPORTED;
  function isImageTypeSupported(type) {
    switch (type) {
      case "auto":
        return IMAGE_BITMAP_SUPPORTED || IMAGE_SUPPORTED || DATA_SUPPORTED;
      case "imagebitmap":
        return IMAGE_BITMAP_SUPPORTED;
      case "image":
        return IMAGE_SUPPORTED;
      case "data":
        return DATA_SUPPORTED;
      default:
        throw new Error(`@loaders.gl/images: image ${type} not supported in this environment`);
    }
  }
  function getDefaultImageType() {
    if (IMAGE_BITMAP_SUPPORTED) {
      return "imagebitmap";
    }
    if (IMAGE_SUPPORTED) {
      return "image";
    }
    if (DATA_SUPPORTED) {
      return "data";
    }
    throw new Error("Install '@loaders.gl/polyfills' to parse images under Node.js");
  }

  // ../images/src/lib/category-api/parsed-image-api.ts
  function isImage(image) {
    return Boolean(getImageTypeOrNull(image));
  }
  function getImageType(image) {
    const format = getImageTypeOrNull(image);
    if (!format) {
      throw new Error("Not an image");
    }
    return format;
  }
  function getImageSize(image) {
    return getImageData(image);
  }
  function getImageData(image) {
    switch (getImageType(image)) {
      case "data":
        return image;
      case "image":
      case "imagebitmap":
        const canvas = document.createElement("canvas");
        const context = canvas.getContext("2d");
        if (!context) {
          throw new Error("getImageData");
        }
        canvas.width = image.width;
        canvas.height = image.height;
        context.drawImage(image, 0, 0);
        return context.getImageData(0, 0, image.width, image.height);
      default:
        throw new Error("getImageData");
    }
  }
  function getImageTypeOrNull(image) {
    if (typeof ImageBitmap !== "undefined" && image instanceof ImageBitmap) {
      return "imagebitmap";
    }
    if (typeof Image !== "undefined" && image instanceof Image) {
      return "image";
    }
    if (image && typeof image === "object" && image.data && image.width && image.height) {
      return "data";
    }
    return null;
  }

  // ../images/src/lib/parsers/svg-utils.ts
  var SVG_DATA_URL_PATTERN = /^data:image\/svg\+xml/;
  var SVG_URL_PATTERN = /\.svg((\?|#).*)?$/;
  function isSVG(url) {
    return url && (SVG_DATA_URL_PATTERN.test(url) || SVG_URL_PATTERN.test(url));
  }
  function getBlobOrSVGDataUrl(arrayBuffer, url) {
    if (isSVG(url)) {
      const textDecoder = new TextDecoder();
      let xmlText = textDecoder.decode(arrayBuffer);
      try {
        if (typeof unescape === "function" && typeof encodeURIComponent === "function") {
          xmlText = unescape(encodeURIComponent(xmlText));
        }
      } catch (error) {
        throw new Error(error.message);
      }
      const src = `data:image/svg+xml;base64,${btoa(xmlText)}`;
      return src;
    }
    return getBlob(arrayBuffer, url);
  }
  function getBlob(arrayBuffer, url) {
    if (isSVG(url)) {
      throw new Error("SVG cannot be parsed directly to imagebitmap");
    }
    return new Blob([new Uint8Array(arrayBuffer)]);
  }

  // ../images/src/lib/parsers/parse-to-image.ts
  async function parseToImage(arrayBuffer, options, url) {
    const blobOrDataUrl = getBlobOrSVGDataUrl(arrayBuffer, url);
    const URL2 = self.URL || self.webkitURL;
    const objectUrl = typeof blobOrDataUrl !== "string" && URL2.createObjectURL(blobOrDataUrl);
    try {
      return await loadToImage(objectUrl || blobOrDataUrl, options);
    } finally {
      if (objectUrl) {
        URL2.revokeObjectURL(objectUrl);
      }
    }
  }
  async function loadToImage(url, options) {
    const image = new Image();
    image.src = url;
    if (options.image && options.image.decode && image.decode) {
      await image.decode();
      return image;
    }
    return await new Promise((resolve2, reject) => {
      try {
        image.onload = () => resolve2(image);
        image.onerror = (error) => {
          const message = error instanceof Error ? error.message : "error";
          reject(new Error(message));
        };
      } catch (error) {
        reject(error);
      }
    });
  }

  // ../images/src/lib/parsers/parse-to-image-bitmap.ts
  var imagebitmapOptionsSupported = true;
  async function parseToImageBitmap(arrayBuffer, options, url) {
    let blob;
    if (isSVG(url)) {
      const image = await parseToImage(arrayBuffer, options, url);
      blob = image;
    } else {
      blob = getBlob(arrayBuffer, url);
    }
    const imagebitmapOptions = options && options.imagebitmap;
    return await safeCreateImageBitmap(blob, imagebitmapOptions);
  }
  async function safeCreateImageBitmap(blob, imagebitmapOptions = null) {
    if (isEmptyObject(imagebitmapOptions) || !imagebitmapOptionsSupported) {
      imagebitmapOptions = null;
    }
    if (imagebitmapOptions) {
      try {
        return await createImageBitmap(blob, imagebitmapOptions);
      } catch (error) {
        console.warn(error);
        imagebitmapOptionsSupported = false;
      }
    }
    return await createImageBitmap(blob);
  }
  function isEmptyObject(object) {
    if (!object) {
      return true;
    }
    for (const key in object) {
      if (Object.prototype.hasOwnProperty.call(object, key)) {
        return false;
      }
    }
    return true;
  }

  // ../images/src/lib/category-api/parse-isobmff-binary.ts
  function getISOBMFFMediaType(buffer) {
    if (!checkString(buffer, "ftyp", 4)) {
      return null;
    }
    if ((buffer[8] & 96) === 0) {
      return null;
    }
    return decodeMajorBrand(buffer);
  }
  function decodeMajorBrand(buffer) {
    const brandMajor = getUTF8String(buffer, 8, 12).replace("\0", " ").trim();
    switch (brandMajor) {
      case "avif":
      case "avis":
        return { extension: "avif", mimeType: "image/avif" };
      default:
        return null;
    }
  }
  function getUTF8String(array, start, end) {
    return String.fromCharCode(...array.slice(start, end));
  }
  function stringToBytes(string) {
    return [...string].map((character) => character.charCodeAt(0));
  }
  function checkString(buffer, header, offset = 0) {
    const headerBytes = stringToBytes(header);
    for (let i = 0; i < headerBytes.length; ++i) {
      if (headerBytes[i] !== buffer[i + offset]) {
        return false;
      }
    }
    return true;
  }

  // ../images/src/lib/category-api/binary-image-api.ts
  var BIG_ENDIAN = false;
  var LITTLE_ENDIAN = true;
  function getBinaryImageMetadata(binaryData) {
    const dataView = toDataView(binaryData);
    return getPngMetadata(dataView) || getJpegMetadata(dataView) || getGifMetadata(dataView) || getBmpMetadata(dataView) || getISOBMFFMetadata(dataView);
  }
  function getISOBMFFMetadata(binaryData) {
    const buffer = new Uint8Array(binaryData instanceof DataView ? binaryData.buffer : binaryData);
    const mediaType = getISOBMFFMediaType(buffer);
    if (!mediaType) {
      return null;
    }
    return {
      mimeType: mediaType.mimeType,
      // TODO - decode width and height
      width: 0,
      height: 0
    };
  }
  function getPngMetadata(binaryData) {
    const dataView = toDataView(binaryData);
    const isPng = dataView.byteLength >= 24 && dataView.getUint32(0, BIG_ENDIAN) === 2303741511;
    if (!isPng) {
      return null;
    }
    return {
      mimeType: "image/png",
      width: dataView.getUint32(16, BIG_ENDIAN),
      height: dataView.getUint32(20, BIG_ENDIAN)
    };
  }
  function getGifMetadata(binaryData) {
    const dataView = toDataView(binaryData);
    const isGif = dataView.byteLength >= 10 && dataView.getUint32(0, BIG_ENDIAN) === 1195984440;
    if (!isGif) {
      return null;
    }
    return {
      mimeType: "image/gif",
      width: dataView.getUint16(6, LITTLE_ENDIAN),
      height: dataView.getUint16(8, LITTLE_ENDIAN)
    };
  }
  function getBmpMetadata(binaryData) {
    const dataView = toDataView(binaryData);
    const isBmp = dataView.byteLength >= 14 && dataView.getUint16(0, BIG_ENDIAN) === 16973 && dataView.getUint32(2, LITTLE_ENDIAN) === dataView.byteLength;
    if (!isBmp) {
      return null;
    }
    return {
      mimeType: "image/bmp",
      width: dataView.getUint32(18, LITTLE_ENDIAN),
      height: dataView.getUint32(22, LITTLE_ENDIAN)
    };
  }
  function getJpegMetadata(binaryData) {
    const dataView = toDataView(binaryData);
    const isJpeg = dataView.byteLength >= 3 && dataView.getUint16(0, BIG_ENDIAN) === 65496 && dataView.getUint8(2) === 255;
    if (!isJpeg) {
      return null;
    }
    const { tableMarkers, sofMarkers } = getJpegMarkers();
    let i = 2;
    while (i + 9 < dataView.byteLength) {
      const marker = dataView.getUint16(i, BIG_ENDIAN);
      if (sofMarkers.has(marker)) {
        return {
          mimeType: "image/jpeg",
          height: dataView.getUint16(i + 5, BIG_ENDIAN),
          // Number of lines
          width: dataView.getUint16(i + 7, BIG_ENDIAN)
          // Number of pixels per line
        };
      }
      if (!tableMarkers.has(marker)) {
        return null;
      }
      i += 2;
      i += dataView.getUint16(i, BIG_ENDIAN);
    }
    return null;
  }
  function getJpegMarkers() {
    const tableMarkers = /* @__PURE__ */ new Set([65499, 65476, 65484, 65501, 65534]);
    for (let i = 65504; i < 65520; ++i) {
      tableMarkers.add(i);
    }
    const sofMarkers = /* @__PURE__ */ new Set([
      65472,
      65473,
      65474,
      65475,
      65477,
      65478,
      65479,
      65481,
      65482,
      65483,
      65485,
      65486,
      65487,
      65502
    ]);
    return { tableMarkers, sofMarkers };
  }
  function toDataView(data) {
    if (data instanceof DataView) {
      return data;
    }
    if (ArrayBuffer.isView(data)) {
      return new DataView(data.buffer);
    }
    if (data instanceof ArrayBuffer) {
      return new DataView(data);
    }
    throw new Error("toDataView");
  }

  // ../images/src/lib/parsers/parse-to-node-image.ts
  async function parseToNodeImage(arrayBuffer, options) {
    const { mimeType } = getBinaryImageMetadata(arrayBuffer) || {};
    const parseImageNode2 = globalThis.loaders?.parseImageNode;
    assert2(parseImageNode2);
    return await parseImageNode2(arrayBuffer, mimeType);
  }

  // ../images/src/lib/parsers/parse-image.ts
  async function parseImage(arrayBuffer, options, context) {
    options = options || {};
    const imageOptions = options.image || {};
    const imageType = imageOptions.type || "auto";
    const { url } = context || {};
    const loadType = getLoadableImageType(imageType);
    let image;
    switch (loadType) {
      case "imagebitmap":
        image = await parseToImageBitmap(arrayBuffer, options, url);
        break;
      case "image":
        image = await parseToImage(arrayBuffer, options, url);
        break;
      case "data":
        image = await parseToNodeImage(arrayBuffer, options);
        break;
      default:
        assert2(false);
    }
    if (imageType === "data") {
      image = getImageData(image);
    }
    return image;
  }
  function getLoadableImageType(type) {
    switch (type) {
      case "auto":
      case "data":
        return getDefaultImageType();
      default:
        isImageTypeSupported(type);
        return type;
    }
  }

  // ../images/src/image-loader.ts
  var EXTENSIONS = ["png", "jpg", "jpeg", "gif", "webp", "bmp", "ico", "svg", "avif"];
  var MIME_TYPES = [
    "image/png",
    "image/jpeg",
    "image/gif",
    "image/webp",
    "image/avif",
    "image/bmp",
    "image/vnd.microsoft.icon",
    "image/svg+xml"
  ];
  var DEFAULT_IMAGE_LOADER_OPTIONS = {
    image: {
      type: "auto",
      decode: true
      // if format is HTML
    }
    // imagebitmap: {} - passes (platform dependent) parameters to ImageBitmap constructor
  };
  var ImageLoader = {
    dataType: null,
    batchType: null,
    id: "image",
    module: "images",
    name: "Images",
    version: VERSION5,
    mimeTypes: MIME_TYPES,
    extensions: EXTENSIONS,
    parse: parseImage,
    // TODO: byteOffset, byteLength;
    tests: [(arrayBuffer) => Boolean(getBinaryImageMetadata(new DataView(arrayBuffer)))],
    options: DEFAULT_IMAGE_LOADER_OPTIONS
  };

  // src/lib/texture-api/async-deep-map.ts
  var isObject = (value) => value && typeof value === "object";
  async function asyncDeepMap(tree, func, options = {}) {
    return await mapSubtree(tree, func, options);
  }
  async function mapSubtree(object, func, options) {
    if (Array.isArray(object)) {
      return await mapArray(object, func, options);
    }
    if (isObject(object)) {
      return await mapObject(object, func, options);
    }
    const url = object;
    return await func(url, options);
  }
  async function mapObject(object, func, options) {
    const promises = [];
    const values = {};
    for (const key in object) {
      const url = object[key];
      const promise = mapSubtree(url, func, options).then((value) => {
        values[key] = value;
      });
      promises.push(promise);
    }
    await Promise.all(promises);
    return values;
  }
  async function mapArray(urlArray, func, options = {}) {
    const promises = urlArray.map((url) => mapSubtree(url, func, options));
    return await Promise.all(promises);
  }

  // src/lib/composite-image/image-texture-cube.ts
  var GL_TEXTURE_CUBE_MAP_POSITIVE_X = 34069;
  var GL_TEXTURE_CUBE_MAP_NEGATIVE_X = 34070;
  var GL_TEXTURE_CUBE_MAP_POSITIVE_Y = 34071;
  var GL_TEXTURE_CUBE_MAP_NEGATIVE_Y = 34072;
  var GL_TEXTURE_CUBE_MAP_POSITIVE_Z = 34073;
  var GL_TEXTURE_CUBE_MAP_NEGATIVE_Z = 34074;
  var IMAGE_TEXTURE_CUBE_FACES = [
    {
      face: GL_TEXTURE_CUBE_MAP_POSITIVE_X,
      name: "+X",
      direction: "right",
      axis: "x",
      sign: "positive"
    },
    {
      face: GL_TEXTURE_CUBE_MAP_NEGATIVE_X,
      name: "-X",
      direction: "left",
      axis: "x",
      sign: "negative"
    },
    { face: GL_TEXTURE_CUBE_MAP_POSITIVE_Y, name: "+Y", direction: "top", axis: "y", sign: "positive" },
    {
      face: GL_TEXTURE_CUBE_MAP_NEGATIVE_Y,
      name: "-Y",
      direction: "bottom",
      axis: "y",
      sign: "negative"
    },
    {
      face: GL_TEXTURE_CUBE_MAP_POSITIVE_Z,
      name: "+Z",
      direction: "front",
      axis: "z",
      sign: "positive"
    },
    { face: GL_TEXTURE_CUBE_MAP_NEGATIVE_Z, name: "-Z", direction: "back", axis: "z", sign: "negative" }
  ];

  // src/lib/composite-image/parse-composite-image.ts
  async function parseCompositeImageManifest(text, expectedShape, options = {}, context) {
    const manifest = parseCompositeImageManifestJSON(text);
    if (manifest.shape !== expectedShape) {
      throw new Error(`Expected ${expectedShape} manifest, got ${manifest.shape}`);
    }
    return await loadCompositeImageManifest(manifest, options, context);
  }
  function testCompositeImageManifestShape(text, shape) {
    try {
      return parseCompositeImageManifestJSON(text).shape === shape;
    } catch {
      return false;
    }
  }
  async function loadCompositeImageManifest(manifest, options = {}, context) {
    const normalizedOptions = normalizeCompositeImageManifestOptions(options);
    const urlTree = await getCompositeImageUrlTree(manifest, normalizedOptions, context);
    const imageData = await loadCompositeImageUrlTree(urlTree, normalizedOptions, context);
    return convertCompositeImageToTexture(manifest.shape, imageData);
  }
  async function loadCompositeImageUrlTree(urlTree, options = {}, context) {
    const normalizedOptions = normalizeCompositeImageOptions(options);
    return await asyncDeepMap(
      urlTree,
      async (url) => await loadCompositeImageMember(url, normalizedOptions, context)
    );
  }
  async function loadCompositeImageMember(url, options = {}, context) {
    const resolvedUrl = resolveCompositeImageUrl(url, options, context);
    const fetch2 = getCompositeImageFetch(options, context);
    const response = await fetch2(resolvedUrl);
    const subloaderOptions = getCompositeImageSubloaderOptions(options);
    if (context) {
      const childContext = getCompositeImageMemberContext(resolvedUrl, response, context);
      return await parseFromContext(
        response,
        [ImageLoader],
        subloaderOptions,
        childContext
      );
    }
    const arrayBuffer = await response.arrayBuffer();
    return await ImageLoader.parse(arrayBuffer, subloaderOptions);
  }
  async function getCompositeImageUrlTree(manifest, options = {}, context) {
    switch (manifest.shape) {
      case "image-texture":
        return await getImageTextureSource(manifest, options, context);
      case "image-texture-array":
        if (!Array.isArray(manifest.layers) || manifest.layers.length === 0) {
          throw new Error("image-texture-array manifest must define one or more layers");
        }
        return await Promise.all(
          manifest.layers.map(
            async (layer, index) => await getNormalizedImageTextureSource(layer, options, context, { index })
          )
        );
      case "image-texture-cube":
        return await getImageTextureCubeUrls(manifest, options, context);
      case "image-texture-cube-array":
        if (!Array.isArray(manifest.layers) || manifest.layers.length === 0) {
          throw new Error("image-texture-cube-array manifest must define one or more layers");
        }
        return await Promise.all(
          manifest.layers.map(
            async (layer, index) => await getImageTextureCubeUrls(layer, options, context, { index })
          )
        );
      default:
        throw new Error("Unsupported composite image manifest");
    }
  }
  function normalizeCompositeImageOptions(options = {}) {
    if (options.core?.baseUrl) {
      return options;
    }
    const fallbackBaseUrl = options.baseUrl;
    if (!fallbackBaseUrl) {
      return options;
    }
    return {
      ...options,
      core: {
        ...options.core,
        baseUrl: fallbackBaseUrl
      }
    };
  }
  function resolveCompositeImageUrl(url, options = {}, context) {
    const resolvedUrl = resolvePath(url);
    if (isAbsoluteCompositeImageUrl(url)) {
      return resolvedUrl;
    }
    const baseUrl = getCompositeImageBaseUrl(options, context);
    if (!baseUrl) {
      if (resolvedUrl !== url || url.startsWith("@")) {
        return resolvedUrl;
      }
      throw new Error(`Unable to resolve relative image URL ${url} without a base URL`);
    }
    return resolvePath(joinCompositeImageUrl(baseUrl, url));
  }
  function parseCompositeImageManifestJSON(text) {
    const manifest = JSON.parse(text);
    if (!manifest?.shape) {
      throw new Error("Composite image manifest must contain a shape field");
    }
    return manifest;
  }
  async function getImageTextureSource(manifest, options, context) {
    if ((manifest.image || manifest.mipmaps) && manifest.template) {
      throw new Error("image-texture manifest must define image, mipmaps, or template source");
    }
    if (manifest.image && manifest.mipmaps) {
      throw new Error("image-texture manifest must define image, mipmaps, or template source");
    }
    if (manifest.image) {
      return manifest.image;
    }
    if (manifest.mipmaps?.length) {
      return manifest.mipmaps;
    }
    if (manifest.template) {
      return await expandImageTextureSource(
        { mipLevels: manifest.mipLevels ?? "auto", template: manifest.template },
        options,
        context,
        {}
      );
    }
    throw new Error("image-texture manifest must define image, mipmaps, or template source");
  }
  async function getImageTextureCubeUrls(manifest, options, context, templateOptions = {}) {
    const urls = {};
    for (const { face, name, direction, axis, sign } of IMAGE_TEXTURE_CUBE_FACES) {
      const source = manifest.faces?.[name] || manifest.faces?.[direction];
      if (!source) {
        throw new Error(`image-texture-cube manifest is missing ${name} face`);
      }
      urls[face] = await getNormalizedImageTextureSource(source, options, context, {
        ...templateOptions,
        face: name,
        direction,
        axis,
        sign
      });
    }
    return urls;
  }
  async function getNormalizedImageTextureSource(source, options, context, templateOptions) {
    if (typeof source === "string") {
      return source;
    }
    if (Array.isArray(source) && source.length > 0) {
      return source;
    }
    if (isImageTextureTemplateSource(source)) {
      return await expandImageTextureSource(source, options, context, templateOptions);
    }
    throw new Error("Composite image source entries must be strings or non-empty mip arrays");
  }
  async function expandImageTextureSource(source, options, context, templateOptions) {
    const mipLevels = source.mipLevels === "auto" ? await getAutoMipLevels(source.template, options, context, templateOptions) : source.mipLevels;
    if (!Number.isFinite(mipLevels) || mipLevels <= 0) {
      throw new Error(`Invalid mipLevels value ${source.mipLevels}`);
    }
    const urls = [];
    for (let lod = 0; lod < mipLevels; lod++) {
      urls.push(expandTemplate(source.template, { ...templateOptions, lod }));
    }
    return urls;
  }
  async function getAutoMipLevels(template, options, context, templateOptions) {
    if (!template.includes("{lod}")) {
      throw new Error("Template sources with mipLevels: auto must include a {lod} placeholder");
    }
    const level0Url = expandTemplate(template, { ...templateOptions, lod: 0 });
    const image = await loadCompositeImageMember(
      level0Url,
      normalizeCompositeImageOptions(options),
      context
    );
    const { width, height } = getImageSize(image);
    return 1 + Math.floor(Math.log2(Math.max(width, height)));
  }
  function expandTemplate(template, templateOptions) {
    let expanded = "";
    for (let index = 0; index < template.length; index++) {
      const character = template[index];
      if (character === "\\") {
        const nextCharacter = template[index + 1];
        if (nextCharacter === "{" || nextCharacter === "}" || nextCharacter === "\\") {
          expanded += nextCharacter;
          index++;
          continue;
        }
        throw new Error(`Invalid escape sequence \\${nextCharacter || ""} in template ${template}`);
      }
      if (character === "}") {
        throw new Error(`Unexpected } in template ${template}`);
      }
      if (character !== "{") {
        expanded += character;
        continue;
      }
      const closingBraceIndex = findClosingBraceIndex(template, index + 1);
      if (closingBraceIndex < 0) {
        throw new Error(`Unterminated placeholder in template ${template}`);
      }
      const placeholder = template.slice(index + 1, closingBraceIndex);
      if (!/^[a-z][a-zA-Z0-9]*$/.test(placeholder)) {
        throw new Error(`Invalid placeholder {${placeholder}} in template ${template}`);
      }
      const value = getTemplateValue(placeholder, templateOptions);
      if (value === void 0) {
        throw new Error(
          `Template ${template} uses unsupported placeholder {${placeholder}} for this source`
        );
      }
      expanded += String(value);
      index = closingBraceIndex;
    }
    return expanded;
  }
  function findClosingBraceIndex(template, startIndex) {
    for (let index = startIndex; index < template.length; index++) {
      const character = template[index];
      if (character === "\\") {
        index++;
        continue;
      }
      if (character === "{") {
        throw new Error(`Nested placeholders are not supported in template ${template}`);
      }
      if (character === "}") {
        return index;
      }
    }
    return -1;
  }
  function getTemplateValue(placeholder, templateOptions) {
    switch (placeholder) {
      case "lod":
        return templateOptions.lod;
      case "index":
        return templateOptions.index;
      case "face":
        return templateOptions.face;
      case "direction":
        return templateOptions.direction;
      case "axis":
        return templateOptions.axis;
      case "sign":
        return templateOptions.sign;
      default:
        return void 0;
    }
  }
  function isImageTextureTemplateSource(source) {
    return typeof source === "object" && source !== null && !Array.isArray(source);
  }
  function getCompositeImageBaseUrl(options, context) {
    if (context?.baseUrl) {
      return context.baseUrl;
    }
    if (options.baseUrl) {
      return stripTrailingSlash(options.baseUrl);
    }
    if (options.core?.baseUrl) {
      return getSourceUrlDirectory(options.core.baseUrl);
    }
    return null;
  }
  function stripTrailingSlash(baseUrl) {
    if (baseUrl.endsWith("/")) {
      return baseUrl.slice(0, -1);
    }
    return baseUrl;
  }
  function getSourceUrlDirectory(baseUrl) {
    return stripTrailingSlash(path_exports.dirname(baseUrl));
  }
  function joinCompositeImageUrl(baseUrl, url) {
    if (isRequestLikeUrl(baseUrl)) {
      return new URL(url, `${stripTrailingSlash(baseUrl)}/`).toString();
    }
    const normalizedBaseUrl = baseUrl.startsWith("/") ? baseUrl : `/${baseUrl}`;
    const normalizedUrl = path_exports.resolve(normalizedBaseUrl, url);
    return baseUrl.startsWith("/") ? normalizedUrl : normalizedUrl.slice(1);
  }
  function isRequestLikeUrl(url) {
    return url.startsWith("http:") || url.startsWith("https:") || url.startsWith("file:") || url.startsWith("blob:");
  }
  function getCompositeImageFetch(options, context) {
    const fetchOption = options.fetch ?? options.core?.fetch;
    if (context?.fetch) {
      return context.fetch;
    }
    if (typeof fetchOption === "function") {
      return fetchOption;
    }
    if (fetchOption && typeof fetchOption === "object") {
      return (url) => fetch(url, fetchOption);
    }
    return fetch;
  }
  function getCompositeImageSubloaderOptions(options) {
    const core = options.core;
    const rest = { ...options };
    delete rest.baseUrl;
    if (!core?.baseUrl) {
      return rest;
    }
    const restCore = { ...core };
    delete restCore.baseUrl;
    return {
      ...rest,
      core: restCore
    };
  }
  function normalizeCompositeImageManifestOptions(options) {
    if (options.image?.type || typeof ImageBitmap === "undefined") {
      return options;
    }
    return {
      ...options,
      image: {
        ...options.image,
        type: "imagebitmap"
      }
    };
  }
  function getCompositeImageMemberContext(resolvedUrl, response, context) {
    const url = response.url || resolvedUrl;
    const [urlWithoutQueryString, queryString = ""] = url.split("?");
    return {
      ...context,
      url,
      response,
      filename: path_exports.filename(urlWithoutQueryString),
      baseUrl: path_exports.dirname(urlWithoutQueryString),
      queryString
    };
  }
  function convertCompositeImageToTexture(shape, imageData) {
    switch (shape) {
      case "image-texture": {
        const data = normalizeCompositeImageMember(imageData);
        return {
          shape: "texture",
          type: "2d",
          format: getCompositeTextureFormat(data),
          data
        };
      }
      case "image-texture-array": {
        const data = imageData.map((layer) => normalizeCompositeImageMember(layer));
        return {
          shape: "texture",
          type: "2d-array",
          format: getCompositeTextureFormat(data[0]),
          data
        };
      }
      case "image-texture-cube": {
        const data = IMAGE_TEXTURE_CUBE_FACES.map(
          ({ face }) => normalizeCompositeImageMember(imageData[face])
        );
        return {
          shape: "texture",
          type: "cube",
          format: getCompositeTextureFormat(data[0]),
          data
        };
      }
      case "image-texture-cube-array": {
        const data = imageData.map(
          (layer) => IMAGE_TEXTURE_CUBE_FACES.map(({ face }) => normalizeCompositeImageMember(layer[face]))
        );
        return {
          shape: "texture",
          type: "cube-array",
          format: getCompositeTextureFormat(data[0][0]),
          data
        };
      }
      default:
        throw new Error(`Unsupported composite image shape ${shape}`);
    }
  }
  function normalizeCompositeImageMember(imageData) {
    if (Array.isArray(imageData)) {
      if (imageData.length === 0) {
        throw new Error("Composite image members must not be empty");
      }
      if (imageData.every(isTextureLevel)) {
        return imageData;
      }
      if (imageData.every(isImage)) {
        return imageData.map((image) => getTextureLevelFromImage(image));
      }
      if (imageData.every((entry) => Array.isArray(entry) && entry.every(isTextureLevel))) {
        if (imageData.length !== 1) {
          throw new Error("Composite image members must resolve to a single image or mip chain");
        }
        return imageData[0];
      }
    }
    if (isTexture(imageData)) {
      if (imageData.type !== "2d") {
        throw new Error(`Composite image members must resolve to 2d textures, got ${imageData.type}`);
      }
      return imageData.data;
    }
    if (isTextureLevel(imageData)) {
      return [imageData];
    }
    if (isImage(imageData)) {
      return [getTextureLevelFromImage(imageData)];
    }
    throw new Error("Composite image members must resolve to an image, mip chain, or texture");
  }
  function getTextureLevelFromImage(image) {
    const { width, height } = getImageSize(image);
    return {
      shape: "texture-level",
      compressed: false,
      width,
      height,
      imageBitmap: typeof ImageBitmap !== "undefined" && image instanceof ImageBitmap ? image : void 0,
      data: new Uint8Array(0),
      textureFormat: "rgba8unorm"
    };
  }
  function getCompositeTextureFormat(textureLevels) {
    return textureLevels[0]?.textureFormat || "rgba8unorm";
  }
  function isTextureLevel(textureLevel) {
    return Boolean(
      textureLevel && typeof textureLevel === "object" && "shape" in textureLevel && textureLevel.shape === "texture-level"
    );
  }
  function isTexture(texture) {
    return Boolean(
      texture && typeof texture === "object" && "shape" in texture && texture.shape === "texture"
    );
  }
  function isAbsoluteCompositeImageUrl(url) {
    return url.startsWith("data:") || url.startsWith("blob:") || url.startsWith("file:") || url.startsWith("http:") || url.startsWith("https:") || url.startsWith("/");
  }

  // src/texture-loader.ts
  var TextureLoader = {
    dataType: null,
    batchType: null,
    id: "texture",
    name: "Texture",
    module: "textures",
    version: VERSION,
    extensions: [],
    mimeTypes: [],
    text: true,
    worker: false,
    testText: (text) => testCompositeImageManifestShape(text, "image-texture"),
    options: {
      image: {}
    },
    parse: async (arrayBuffer, options, context) => await parseCompositeImageManifest(
      new TextDecoder().decode(arrayBuffer),
      "image-texture",
      options,
      context
    ),
    parseText: async (text, options, context) => await parseCompositeImageManifest(text, "image-texture", options, context)
  };

  // src/texture-array-loader.ts
  var TextureArrayLoader = {
    dataType: null,
    batchType: null,
    id: "texture-array",
    name: "Texture Array",
    module: "textures",
    version: VERSION,
    extensions: [],
    mimeTypes: [],
    text: true,
    worker: false,
    testText: (text) => testCompositeImageManifestShape(text, "image-texture-array"),
    options: {
      image: {}
    },
    parse: async (arrayBuffer, options, context) => await parseCompositeImageManifest(
      new TextDecoder().decode(arrayBuffer),
      "image-texture-array",
      options,
      context
    ),
    parseText: async (text, options, context) => await parseCompositeImageManifest(text, "image-texture-array", options, context)
  };

  // src/texture-cube-loader.ts
  var TextureCubeLoader = {
    dataType: null,
    batchType: null,
    id: "texture-cube",
    name: "Texture Cube",
    module: "textures",
    version: VERSION,
    extensions: [],
    mimeTypes: [],
    text: true,
    worker: false,
    testText: (text) => testCompositeImageManifestShape(text, "image-texture-cube"),
    options: {
      image: {}
    },
    parse: async (arrayBuffer, options, context) => await parseCompositeImageManifest(
      new TextDecoder().decode(arrayBuffer),
      "image-texture-cube",
      options,
      context
    ),
    parseText: async (text, options, context) => await parseCompositeImageManifest(text, "image-texture-cube", options, context)
  };

  // src/texture-cube-array-loader.ts
  var TextureCubeArrayLoader = {
    dataType: null,
    batchType: null,
    id: "texture-cube-array",
    name: "Texture Cube Array",
    module: "textures",
    version: VERSION,
    extensions: [],
    mimeTypes: [],
    text: true,
    worker: false,
    testText: (text) => testCompositeImageManifestShape(text, "image-texture-cube-array"),
    options: {
      image: {}
    },
    parse: async (arrayBuffer, options, context) => await parseCompositeImageManifest(
      new TextDecoder().decode(arrayBuffer),
      "image-texture-cube-array",
      options,
      context
    ),
    parseText: async (text, options, context) => await parseCompositeImageManifest(text, "image-texture-cube-array", options, context)
  };

  // src/lib/parsers/crunch-module-loader.ts
  var CRUNCH_EXTERNAL_LIBRARIES = {
    /** Crunch decoder library. It is used as dynamically imported script */
    DECODER: "crunch.js"
  };

  // src/lib/encoders/encode-texture.ts
  async function encodeImageURLToCompressedTextureURL(inputUrl, outputUrl, options) {
    const args = [
      // Note: our actual executable is `npx`, so `texture-compressor` is an argument
      "texture-compressor",
      "--type",
      "s3tc",
      "--compression",
      "DXT1",
      "--quality",
      "normal",
      "--input",
      inputUrl,
      "--output",
      outputUrl
    ];
    const childProcess = new ChildProcessProxy();
    await childProcess.start({
      command: "npx",
      arguments: args,
      spawn: options
    });
    return outputUrl;
  }

  // src/compressed-texture-writer.ts
  var CompressedTextureWriter = {
    name: "DDS Texture Container",
    id: "dds",
    module: "textures",
    version: VERSION,
    extensions: ["dds"],
    mimeTypes: ["image/vnd-ms.dds", "image/x-dds", "application/octet-stream"],
    options: {
      texture: {
        format: "auto",
        compression: "auto",
        quality: "auto",
        mipmap: false,
        flipY: false,
        toolFlags: ""
      }
    },
    encodeURLtoURL: encodeImageURLToCompressedTextureURL,
    encode() {
      throw new Error("Not implemented");
    }
  };

  // src/lib/encoders/encode-ktx2-basis-texture.ts
  async function encodeKTX2BasisTexture(image, options = {}) {
    const {
      useSRGB = false,
      qualityLevel = 10,
      encodeUASTC = false,
      mipmaps = false
    } = options?.["ktx2-basis-writer"] || {};
    const { BasisEncoder } = await loadBasisEncoderModule(extractLoadLibraryOptions(options));
    const basisEncoder = new BasisEncoder();
    try {
      const basisFileData = new Uint8Array(image.width * image.height * 4);
      basisEncoder.setCreateKTX2File(true);
      basisEncoder.setKTX2UASTCSupercompression(true);
      basisEncoder.setKTX2SRGBTransferFunc(true);
      basisEncoder.setSliceSourceImage(0, image.data, image.width, image.height, false);
      basisEncoder.setPerceptual(useSRGB);
      basisEncoder.setMipSRGB(useSRGB);
      basisEncoder.setQualityLevel(qualityLevel);
      basisEncoder.setUASTC(encodeUASTC);
      basisEncoder.setMipGen(mipmaps);
      const numOutputBytes = basisEncoder.encode(basisFileData);
      const actualKTX2FileData = basisFileData.subarray(0, numOutputBytes).buffer;
      return actualKTX2FileData;
    } catch (error) {
      console.error("Basis Universal Supercompressed GPU Texture encoder Error: ", error);
      throw error;
    } finally {
      basisEncoder.delete();
    }
  }

  // src/ktx2-basis-writer.ts
  var KTX2BasisWriter = {
    name: "Basis Universal Supercompressed GPU Texture",
    id: "ktx2-basis-writer",
    module: "textures",
    version: VERSION,
    extensions: ["ktx2"],
    mimeTypes: ["image/ktx2"],
    options: {
      ["ktx2-basis-writer"]: {
        useSRGB: false,
        qualityLevel: 10,
        encodeUASTC: false,
        mipmaps: false
      }
    },
    encode: encodeKTX2BasisTexture
  };

  // src/lib/texture-api/generate-url.ts
  function generateUrl(getUrl, options, urlOptions) {
    return typeof getUrl === "function" ? getUrl({ ...options, ...urlOptions }) : getUrl;
  }

  // src/lib/texture-api/load-image.ts
  async function loadImageTexture(getUrl, options = {}) {
    const imageUrls = await getImageUrls(getUrl, options);
    return await loadCompositeImageUrlTree(imageUrls, normalizeCompositeImageOptions(options));
  }
  async function getImageUrls(getUrl, options, urlOptions = {}) {
    const mipLevels = options && options.image && options.image.mipLevels || 0;
    return mipLevels !== 0 ? await getMipmappedImageUrls(getUrl, mipLevels, options, urlOptions) : generateUrl(getUrl, options, urlOptions);
  }
  async function getMipmappedImageUrls(getUrl, mipLevels, options, urlOptions) {
    const urls = [];
    const normalizedOptions = normalizeCompositeImageOptions(options);
    if (mipLevels === "auto") {
      const url = generateUrl(getUrl, options, { ...urlOptions, lod: 0 });
      const image = await loadCompositeImageMember(url, normalizedOptions);
      const { width, height } = getImageSize(image);
      mipLevels = getMipLevels({ width, height });
      urls.push(url);
    }
    assert2(mipLevels > 0);
    for (let mipLevel = urls.length; mipLevel < mipLevels; ++mipLevel) {
      const url = generateUrl(getUrl, options, { ...urlOptions, lod: mipLevel });
      urls.push(url);
    }
    return urls;
  }
  function getMipLevels(size) {
    return 1 + Math.floor(Math.log2(Math.max(size.width, size.height)));
  }

  // src/lib/texture-api/load-image-array.ts
  async function loadImageTextureArray(count, getUrl, options = {}) {
    const imageUrls = await getImageArrayUrls(count, getUrl, options);
    return await loadCompositeImageUrlTree(imageUrls, normalizeCompositeImageOptions(options));
  }
  async function getImageArrayUrls(count, getUrl, options = {}) {
    const promises = [];
    for (let index = 0; index < count; index++) {
      const promise = getImageUrls(getUrl, options, { index });
      promises.push(promise);
    }
    return await Promise.all(promises);
  }

  // src/lib/texture-api/load-image-cube.ts
  async function getImageCubeUrls(getUrl, options) {
    const urls = {};
    const promises = [];
    let index = 0;
    for (let i = 0; i < IMAGE_TEXTURE_CUBE_FACES.length; ++i) {
      const face = IMAGE_TEXTURE_CUBE_FACES[index];
      const promise = getImageUrls(getUrl, options, { ...face, index: index++ }).then((url) => {
        urls[face.face] = url;
      });
      promises.push(promise);
    }
    await Promise.all(promises);
    return urls;
  }
  async function loadImageTextureCube(getUrl, options = {}) {
    const urls = await getImageCubeUrls(getUrl, options);
    return await loadCompositeImageUrlTree(
      urls,
      normalizeCompositeImageOptions(options)
    );
  }

  // src/index.ts
  var KTX2BasisWriterWorker = {
    name: "Basis Universal Supercompressed GPU Texture",
    id: "ktx2-basis-writer",
    module: "textures",
    version: VERSION,
    extensions: ["ktx2"],
    worker: true,
    options: {
      useSRGB: false,
      qualityLevel: 10,
      encodeUASTC: false,
      mipmaps: false
    }
  };
  return __toCommonJS(bundle_exports);
})();
      return __exports__;
      });
