// @loaders.gl/gis/dist/lib/geometry-converters/wkb/helpers/wkb-types.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
/**
 * Integer codes for geometry types in WKB and related formats
 * @see: https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry#Well-known_binary
 */
export var WKBGeometryType;
(function (WKBGeometryType) {
    WKBGeometryType[WKBGeometryType["Point"] = 1] = "Point";
    WKBGeometryType[WKBGeometryType["LineString"] = 2] = "LineString";
    WKBGeometryType[WKBGeometryType["Polygon"] = 3] = "Polygon";
    WKBGeometryType[WKBGeometryType["MultiPoint"] = 4] = "MultiPoint";
    WKBGeometryType[WKBGeometryType["MultiLineString"] = 5] = "MultiLineString";
    WKBGeometryType[WKBGeometryType["MultiPolygon"] = 6] = "MultiPolygon";
    WKBGeometryType[WKBGeometryType["GeometryCollection"] = 7] = "GeometryCollection";
})(WKBGeometryType || (WKBGeometryType = {}));
/** WKB Geometry has a z coordinate */
export const EWKB_FLAG_Z = 0x80000000;
/** WKB Geometry has an m coordinate */
export const EWKB_FLAG_M = 0x40000000;
/** WKB Geometry has a spatial coordinate reference system id */
export const EWKB_FLAG_SRID = 0x20000000;
/** @todo: Assume no more than 10K SRIDs are defined */
export const MAX_SRID = 10000;
/**
 * @see: https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry#Well-known_binary
 * @note Order matches WKBGeometryType enum (need to add 1)
 * @note: We only support this "geojson" subset of the OGC simple features standard
 */
export const WKT_MAGIC_STRINGS = [
    'POINT(',
    'LINESTRING(',
    'POLYGON(',
    'MULTIPOINT(',
    'MULTILINESTRING(',
    'MULTIPOLYGON(',
    'GEOMETRYCOLLECTION('
];
const textEncoder = new TextEncoder();
/**
 * @see: https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry#Well-known_binary
 * @note Order matches WKBGeometryType enum (need to add 1)
 * @note: We only support this "geojson" subset of the OGC simple features standard
 */
export const WKT_MAGIC_BYTES = WKT_MAGIC_STRINGS.map((string) => textEncoder.encode(string));
