// @loaders.gl/gis/dist/lib/geometry-converters/wkb/helpers/wkb-utils.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { WKBGeometryType } from "./wkb-types.js";
const WKBGeometryTypeNames = {
    Point: WKBGeometryType.Point,
    LineString: WKBGeometryType.LineString,
    Polygon: WKBGeometryType.Polygon,
    MultiPoint: WKBGeometryType.MultiPoint,
    MultiLineString: WKBGeometryType.MultiLineString,
    MultiPolygon: WKBGeometryType.MultiPolygon,
    GeometryCollection: WKBGeometryType.GeometryCollection
};
/** Convert a WKB geometry type integer to GeoJSON style geometry type string*/
export function getGeometryTypeFromWKBType(wkbGeometryType) {
    for (const key in WKBGeometryTypeNames) {
        if (WKBGeometryTypeNames[key] === wkbGeometryType) {
            return key;
        }
    }
    throw new Error(String(wkbGeometryType));
}
/** Convert a GeoJSON style geometry type string to a WKB geometry type integer*/
export function getWKBTypeFromGeometryType(geometryType) {
    return WKBGeometryTypeNames[geometryType];
}
/** Sanity: Adjust Z/M dimensions with actual point size */
export function matchWKBOptionsToPointSize(pointSize, options) {
    const wkbOptions = { hasZ: false, hasM: false, srid: undefined, ...options };
    // Align options with actual point size
    switch (pointSize) {
        case 4:
            wkbOptions.hasZ = true;
            wkbOptions.hasM = true;
            break;
        case 3:
            // Prefer Z over M
            if (wkbOptions.hasZ && wkbOptions.hasM) {
                wkbOptions.hasM = false;
            }
            if (!wkbOptions.hasZ && !wkbOptions.hasM) {
                wkbOptions.hasZ = true;
            }
            break;
        case 2:
            wkbOptions.hasZ = false;
            wkbOptions.hasM = false;
            break;
        default:
        // ignore - broken input data?
    }
    return wkbOptions;
}
/** Get coordinate size given Z/M dimensions */
export function getCoordinateByteSize(options) {
    let coordinateByteSize = 16;
    if (options?.hasZ) {
        coordinateByteSize += 8;
    }
    if (options?.hasM) {
        coordinateByteSize += 8;
    }
    return coordinateByteSize;
}
