// @luma.gl/gltf/dist/webgl-to-webgpu/convert-webgl-sampler.js
// luma.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { GLEnum } from "./gltf-webgl-constants.js";
/** Converts a glTF sampler into luma.gl sampler props. */
export function convertSampler(gltfSampler) {
    return {
        addressModeU: convertSamplerWrapMode(gltfSampler.wrapS),
        addressModeV: convertSamplerWrapMode(gltfSampler.wrapT),
        magFilter: convertSamplerMagFilter(gltfSampler.magFilter),
        ...convertSamplerMinFilter(gltfSampler.minFilter)
    };
}
/** Converts a glTF wrap enum into a luma.gl address mode. */
function convertSamplerWrapMode(mode) {
    switch (mode) {
        case GLEnum.CLAMP_TO_EDGE:
            return 'clamp-to-edge';
        case GLEnum.REPEAT:
            return 'repeat';
        case GLEnum.MIRRORED_REPEAT:
            return 'mirror-repeat';
        default:
            return undefined;
    }
}
/** Converts a glTF mag filter enum into a luma.gl mag filter. */
function convertSamplerMagFilter(mode) {
    switch (mode) {
        case GLEnum.NEAREST:
            return 'nearest';
        case GLEnum.LINEAR:
            return 'linear';
        default:
            return undefined;
    }
}
/** Converts a glTF min filter enum into luma.gl minification and mipmap filters. */
function convertSamplerMinFilter(mode) {
    switch (mode) {
        case GLEnum.NEAREST:
            return { minFilter: 'nearest' };
        case GLEnum.LINEAR:
            return { minFilter: 'linear' };
        case GLEnum.NEAREST_MIPMAP_NEAREST:
            return { minFilter: 'nearest', mipmapFilter: 'nearest' };
        case GLEnum.LINEAR_MIPMAP_NEAREST:
            return { minFilter: 'linear', mipmapFilter: 'nearest' };
        case GLEnum.NEAREST_MIPMAP_LINEAR:
            return { minFilter: 'nearest', mipmapFilter: 'linear' };
        case GLEnum.LINEAR_MIPMAP_LINEAR:
            return { minFilter: 'linear', mipmapFilter: 'linear' };
        default:
            return {};
    }
}
