// @luma.gl/gltf/dist/webgl-to-webgpu/gltf-webgl-constants.js
// luma.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
// NOTE: `@luma.gl/gltf` intentionally keeps this as a local enum subset so it
// does not need to depend on `@luma.gl/webgl` for a handful of stable WebGL values.
// eslint-disable-next-line no-shadow
export var GLEnum;
(function (GLEnum) {
    GLEnum[GLEnum["POINTS"] = 0] = "POINTS";
    GLEnum[GLEnum["LINES"] = 1] = "LINES";
    GLEnum[GLEnum["LINE_LOOP"] = 2] = "LINE_LOOP";
    GLEnum[GLEnum["LINE_STRIP"] = 3] = "LINE_STRIP";
    GLEnum[GLEnum["TRIANGLES"] = 4] = "TRIANGLES";
    GLEnum[GLEnum["TRIANGLE_STRIP"] = 5] = "TRIANGLE_STRIP";
    GLEnum[GLEnum["TRIANGLE_FAN"] = 6] = "TRIANGLE_FAN";
    GLEnum[GLEnum["ONE"] = 1] = "ONE";
    GLEnum[GLEnum["SRC_ALPHA"] = 770] = "SRC_ALPHA";
    GLEnum[GLEnum["ONE_MINUS_SRC_ALPHA"] = 771] = "ONE_MINUS_SRC_ALPHA";
    GLEnum[GLEnum["FUNC_ADD"] = 32774] = "FUNC_ADD";
    GLEnum[GLEnum["LINEAR"] = 9729] = "LINEAR";
    GLEnum[GLEnum["NEAREST"] = 9728] = "NEAREST";
    GLEnum[GLEnum["NEAREST_MIPMAP_NEAREST"] = 9984] = "NEAREST_MIPMAP_NEAREST";
    GLEnum[GLEnum["LINEAR_MIPMAP_NEAREST"] = 9985] = "LINEAR_MIPMAP_NEAREST";
    GLEnum[GLEnum["NEAREST_MIPMAP_LINEAR"] = 9986] = "NEAREST_MIPMAP_LINEAR";
    GLEnum[GLEnum["LINEAR_MIPMAP_LINEAR"] = 9987] = "LINEAR_MIPMAP_LINEAR";
    GLEnum[GLEnum["TEXTURE_MIN_FILTER"] = 10241] = "TEXTURE_MIN_FILTER";
    GLEnum[GLEnum["TEXTURE_WRAP_S"] = 10242] = "TEXTURE_WRAP_S";
    GLEnum[GLEnum["TEXTURE_WRAP_T"] = 10243] = "TEXTURE_WRAP_T";
    GLEnum[GLEnum["REPEAT"] = 10497] = "REPEAT";
    GLEnum[GLEnum["CLAMP_TO_EDGE"] = 33071] = "CLAMP_TO_EDGE";
    GLEnum[GLEnum["MIRRORED_REPEAT"] = 33648] = "MIRRORED_REPEAT";
    GLEnum[GLEnum["UNPACK_FLIP_Y_WEBGL"] = 37440] = "UNPACK_FLIP_Y_WEBGL";
})(GLEnum || (GLEnum = {}));
