// @loaders.gl/schema-utils/dist/lib/table/batch-builder/columnar-table-batch-aggregator.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { isTypedArray } from '@math.gl/types';
import { getArrayTypeFromDataType } from "../../schema/data-type.js";
const DEFAULT_ROW_COUNT = 100;
export class ColumnarTableBatchAggregator {
    schema;
    length = 0;
    allocated = 0;
    columns = {};
    constructor(schema, options) {
        this.schema = schema;
        this._reallocateColumns();
    }
    rowCount() {
        return this.length;
    }
    addArrayRow(row) {
        // If user keeps pushing rows beyond batch size, reallocate
        this._reallocateColumns();
        let i = 0;
        // TODO what if no csv header, columns not populated?
        for (const fieldName in this.columns) {
            this.columns[fieldName][this.length] = row[i++];
        }
        this.length++;
    }
    addObjectRow(row) {
        // If user keeps pushing rows beyond batch size, reallocate
        this._reallocateColumns();
        for (const fieldName in row) {
            this.columns[fieldName][this.length] = row[fieldName];
        }
        this.length++;
    }
    getBatch() {
        this._pruneColumns();
        const batch = {
            shape: 'columnar-table',
            batchType: 'data',
            data: this.columns,
            schema: this.schema,
            length: this.length
        };
        return batch;
    }
    // HELPERS
    _reallocateColumns() {
        if (this.length < this.allocated) {
            return;
        }
        // @ts-ignore TODO
        this.allocated = this.allocated > 0 ? (this.allocated *= 2) : DEFAULT_ROW_COUNT;
        this.columns = {};
        for (const field of this.schema.fields) {
            const ArrayType = getArrayTypeFromDataType(field.type, field.nullable);
            const oldColumn = this.columns[field.name];
            if (!oldColumn) {
                // Create new
                this.columns[field.name] = new ArrayType(this.allocated);
            }
            else if (Array.isArray(oldColumn)) {
                // Plain array, just increase its size
                oldColumn.length = this.allocated;
            }
            else if (isTypedArray(oldColumn)) {
                const typedArray = new ArrayType(this.allocated);
                // Copy the old data to the new array
                typedArray.set(oldColumn);
                this.columns[field.name] = typedArray;
            }
            else {
                throw new Error('Invalid column type');
            }
        }
    }
    _pruneColumns() {
        for (const [columnName, column] of Object.entries(this.columns)) {
            this.columns[columnName] = column.slice(0, this.length);
        }
    }
}
