// @loaders.gl/mvt/dist/mvt-loader.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright vis.gl contributors
// import type {MVTOptions} from './lib/types';
import { parseMVT } from "./lib/parse-mvt.js";
import { MVTFormat } from "./mvt-format.js";
// __VERSION__ is injected by babel-plugin-version-inline
// @ts-ignore TS2304: Cannot find name '__VERSION__'.
const VERSION = typeof "4.4.1" !== 'undefined' ? "4.4.1" : 'latest';
/**
 * Worker loader for the Mapbox Vector Tile format
 */
export const MVTWorkerLoader = {
    ...MVTFormat,
    dataType: null,
    batchType: null,
    version: VERSION,
    worker: true,
    options: {
        mvt: {
            shape: 'geojson',
            coordinates: 'local',
            layerProperty: 'layerName',
            layers: undefined,
            tileIndex: undefined
        }
    }
};
/**
 * Loader for the Mapbox Vector Tile format
 */
export const MVTLoader = {
    ...MVTWorkerLoader,
    parse: async (arrayBuffer, options) => parseMVT(arrayBuffer, options),
    parseSync: parseMVT,
    binary: true
};
