// @luma.gl/gltf/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// dist/index.js
var dist_exports = {};
__export(dist_exports, {
  GLTFAnimator: () => GLTFAnimator,
  createScenegraphsFromGLTF: () => createScenegraphsFromGLTF,
  getGLTFExtensionSupport: () => getGLTFExtensionSupport,
  loadPBREnvironment: () => loadPBREnvironment,
  parseGLTFLights: () => parseGLTFLights,
  parsePBRMaterial: () => parsePBRMaterial
});
module.exports = __toCommonJS(dist_exports);

// dist/pbr/pbr-environment.js
var import_engine = require("@luma.gl/engine");
var import_textures = require("@loaders.gl/textures");
function loadPBREnvironment(device, props) {
  const specularMipLevels = props.specularMipLevels ?? 1;
  const brdfLutTexture = new import_engine.DynamicTexture(device, {
    id: "brdfLUT",
    sampler: {
      addressModeU: "clamp-to-edge",
      addressModeV: "clamp-to-edge",
      minFilter: "linear",
      magFilter: "linear"
    },
    // Texture accepts a promise that returns an image as data (Async Textures)
    data: (0, import_textures.loadImageTexture)(resolveTextureUrl(props.brdfLutUrl))
  });
  const diffuseEnvSampler = makeCube(device, {
    id: "DiffuseEnvSampler",
    getTextureForFace: (face) => (0, import_textures.loadImageTexture)(resolveTextureUrl(props.getTexUrl("diffuse", FACES.indexOf(face), 0))),
    sampler: {
      addressModeU: "clamp-to-edge",
      addressModeV: "clamp-to-edge",
      minFilter: "linear",
      magFilter: "linear"
    }
  });
  const specularEnvSampler = makeCube(device, {
    id: "SpecularEnvSampler",
    getTextureForFace: (face) => {
      const imageArray = [];
      const direction = FACES.indexOf(face);
      for (let lod = 0; lod < specularMipLevels; lod++) {
        imageArray.push((0, import_textures.loadImageTexture)(resolveTextureUrl(props.getTexUrl("specular", direction, lod))));
      }
      return Promise.all(imageArray);
    },
    sampler: {
      addressModeU: "clamp-to-edge",
      addressModeV: "clamp-to-edge",
      minFilter: "linear",
      // [GL.TEXTURE_MIN_FILTER]: GL.LINEAR_MIPMAP_LINEAR,
      magFilter: "linear"
    }
  });
  return {
    brdfLutTexture,
    diffuseEnvSampler,
    specularEnvSampler
  };
}
var FACES = ["+X", "-X", "+Y", "-Y", "+Z", "-Z"];
function resolveTextureUrl(url) {
  var _a, _b;
  const baseUrl = ((_a = globalThis.document) == null ? void 0 : _a.baseURI) ?? ((_b = globalThis.location) == null ? void 0 : _b.href);
  return baseUrl ? new URL(url, baseUrl).toString() : url;
}
function makeCube(device, { id, getTextureForFace, sampler }) {
  const data = Promise.all(FACES.map((face) => getTextureForFace(face))).then((faceDataArray) => {
    const cubeData = {};
    FACES.forEach((face, index) => {
      cubeData[face] = faceDataArray[index];
    });
    return cubeData;
  });
  return new import_engine.DynamicTexture(device, {
    id,
    dimension: "cube",
    mipmaps: false,
    sampler,
    data
  });
}

// dist/parsers/parse-pbr-material.js
var import_core2 = require("@luma.gl/core");

// dist/webgl-to-webgpu/gltf-webgl-constants.js
var GLEnum;
(function(GLEnum2) {
  GLEnum2[GLEnum2["POINTS"] = 0] = "POINTS";
  GLEnum2[GLEnum2["LINES"] = 1] = "LINES";
  GLEnum2[GLEnum2["LINE_LOOP"] = 2] = "LINE_LOOP";
  GLEnum2[GLEnum2["LINE_STRIP"] = 3] = "LINE_STRIP";
  GLEnum2[GLEnum2["TRIANGLES"] = 4] = "TRIANGLES";
  GLEnum2[GLEnum2["TRIANGLE_STRIP"] = 5] = "TRIANGLE_STRIP";
  GLEnum2[GLEnum2["TRIANGLE_FAN"] = 6] = "TRIANGLE_FAN";
  GLEnum2[GLEnum2["ONE"] = 1] = "ONE";
  GLEnum2[GLEnum2["SRC_ALPHA"] = 770] = "SRC_ALPHA";
  GLEnum2[GLEnum2["ONE_MINUS_SRC_ALPHA"] = 771] = "ONE_MINUS_SRC_ALPHA";
  GLEnum2[GLEnum2["FUNC_ADD"] = 32774] = "FUNC_ADD";
  GLEnum2[GLEnum2["LINEAR"] = 9729] = "LINEAR";
  GLEnum2[GLEnum2["NEAREST"] = 9728] = "NEAREST";
  GLEnum2[GLEnum2["NEAREST_MIPMAP_NEAREST"] = 9984] = "NEAREST_MIPMAP_NEAREST";
  GLEnum2[GLEnum2["LINEAR_MIPMAP_NEAREST"] = 9985] = "LINEAR_MIPMAP_NEAREST";
  GLEnum2[GLEnum2["NEAREST_MIPMAP_LINEAR"] = 9986] = "NEAREST_MIPMAP_LINEAR";
  GLEnum2[GLEnum2["LINEAR_MIPMAP_LINEAR"] = 9987] = "LINEAR_MIPMAP_LINEAR";
  GLEnum2[GLEnum2["TEXTURE_MIN_FILTER"] = 10241] = "TEXTURE_MIN_FILTER";
  GLEnum2[GLEnum2["TEXTURE_WRAP_S"] = 10242] = "TEXTURE_WRAP_S";
  GLEnum2[GLEnum2["TEXTURE_WRAP_T"] = 10243] = "TEXTURE_WRAP_T";
  GLEnum2[GLEnum2["REPEAT"] = 10497] = "REPEAT";
  GLEnum2[GLEnum2["CLAMP_TO_EDGE"] = 33071] = "CLAMP_TO_EDGE";
  GLEnum2[GLEnum2["MIRRORED_REPEAT"] = 33648] = "MIRRORED_REPEAT";
  GLEnum2[GLEnum2["UNPACK_FLIP_Y_WEBGL"] = 37440] = "UNPACK_FLIP_Y_WEBGL";
})(GLEnum || (GLEnum = {}));

// dist/webgl-to-webgpu/convert-webgl-sampler.js
function convertSampler(gltfSampler) {
  return {
    addressModeU: convertSamplerWrapMode(gltfSampler.wrapS),
    addressModeV: convertSamplerWrapMode(gltfSampler.wrapT),
    magFilter: convertSamplerMagFilter(gltfSampler.magFilter),
    ...convertSamplerMinFilter(gltfSampler.minFilter)
  };
}
function convertSamplerWrapMode(mode) {
  switch (mode) {
    case GLEnum.CLAMP_TO_EDGE:
      return "clamp-to-edge";
    case GLEnum.REPEAT:
      return "repeat";
    case GLEnum.MIRRORED_REPEAT:
      return "mirror-repeat";
    default:
      return void 0;
  }
}
function convertSamplerMagFilter(mode) {
  switch (mode) {
    case GLEnum.NEAREST:
      return "nearest";
    case GLEnum.LINEAR:
      return "linear";
    default:
      return void 0;
  }
}
function convertSamplerMinFilter(mode) {
  switch (mode) {
    case GLEnum.NEAREST:
      return { minFilter: "nearest" };
    case GLEnum.LINEAR:
      return { minFilter: "linear" };
    case GLEnum.NEAREST_MIPMAP_NEAREST:
      return { minFilter: "nearest", mipmapFilter: "nearest" };
    case GLEnum.LINEAR_MIPMAP_NEAREST:
      return { minFilter: "linear", mipmapFilter: "nearest" };
    case GLEnum.NEAREST_MIPMAP_LINEAR:
      return { minFilter: "nearest", mipmapFilter: "linear" };
    case GLEnum.LINEAR_MIPMAP_LINEAR:
      return { minFilter: "linear", mipmapFilter: "linear" };
    default:
      return {};
  }
}

// dist/pbr/texture-transform.js
var import_core = require("@math.gl/core");
var TEXTURE_TRANSFORM_SLOT_DEFINITIONS = [
  createTextureTransformSlotDefinition("baseColor", "pbr_baseColorSampler", "baseColorTexture", [
    "pbrMetallicRoughness",
    "baseColorTexture"
  ]),
  createTextureTransformSlotDefinition("metallicRoughness", "pbr_metallicRoughnessSampler", "metallicRoughnessTexture", ["pbrMetallicRoughness", "metallicRoughnessTexture"]),
  createTextureTransformSlotDefinition("normal", "pbr_normalSampler", "normalTexture", [
    "normalTexture"
  ]),
  createTextureTransformSlotDefinition("occlusion", "pbr_occlusionSampler", "occlusionTexture", [
    "occlusionTexture"
  ]),
  createTextureTransformSlotDefinition("emissive", "pbr_emissiveSampler", "emissiveTexture", [
    "emissiveTexture"
  ]),
  createTextureTransformSlotDefinition("specularColor", "pbr_specularColorSampler", "KHR_materials_specular.specularColorTexture", ["extensions", "KHR_materials_specular", "specularColorTexture"]),
  createTextureTransformSlotDefinition("specularIntensity", "pbr_specularIntensitySampler", "KHR_materials_specular.specularTexture", ["extensions", "KHR_materials_specular", "specularTexture"]),
  createTextureTransformSlotDefinition("transmission", "pbr_transmissionSampler", "KHR_materials_transmission.transmissionTexture", ["extensions", "KHR_materials_transmission", "transmissionTexture"]),
  createTextureTransformSlotDefinition("thickness", "pbr_thicknessSampler", "KHR_materials_volume.thicknessTexture", ["extensions", "KHR_materials_volume", "thicknessTexture"]),
  createTextureTransformSlotDefinition("clearcoat", "pbr_clearcoatSampler", "KHR_materials_clearcoat.clearcoatTexture", ["extensions", "KHR_materials_clearcoat", "clearcoatTexture"]),
  createTextureTransformSlotDefinition("clearcoatRoughness", "pbr_clearcoatRoughnessSampler", "KHR_materials_clearcoat.clearcoatRoughnessTexture", ["extensions", "KHR_materials_clearcoat", "clearcoatRoughnessTexture"]),
  createTextureTransformSlotDefinition("clearcoatNormal", "pbr_clearcoatNormalSampler", "KHR_materials_clearcoat.clearcoatNormalTexture", ["extensions", "KHR_materials_clearcoat", "clearcoatNormalTexture"]),
  createTextureTransformSlotDefinition("sheenColor", "pbr_sheenColorSampler", "KHR_materials_sheen.sheenColorTexture", ["extensions", "KHR_materials_sheen", "sheenColorTexture"]),
  createTextureTransformSlotDefinition("sheenRoughness", "pbr_sheenRoughnessSampler", "KHR_materials_sheen.sheenRoughnessTexture", ["extensions", "KHR_materials_sheen", "sheenRoughnessTexture"]),
  createTextureTransformSlotDefinition("iridescence", "pbr_iridescenceSampler", "KHR_materials_iridescence.iridescenceTexture", ["extensions", "KHR_materials_iridescence", "iridescenceTexture"]),
  createTextureTransformSlotDefinition("iridescenceThickness", "pbr_iridescenceThicknessSampler", "KHR_materials_iridescence.iridescenceThicknessTexture", ["extensions", "KHR_materials_iridescence", "iridescenceThicknessTexture"]),
  createTextureTransformSlotDefinition("anisotropy", "pbr_anisotropySampler", "KHR_materials_anisotropy.anisotropyTexture", ["extensions", "KHR_materials_anisotropy", "anisotropyTexture"])
];
var TEXTURE_TRANSFORM_SLOT_DEFINITION_MAP = new Map(TEXTURE_TRANSFORM_SLOT_DEFINITIONS.map((definition) => [definition.slot, definition]));
function createTextureTransformSlotDefinition(slot, binding, displayName, pathSegments) {
  return {
    slot,
    binding,
    displayName,
    pathSegments,
    uvSetUniform: `${slot}UVSet`,
    uvTransformUniform: `${slot}UVTransform`
  };
}
function getTextureTransformSlotDefinitions() {
  return TEXTURE_TRANSFORM_SLOT_DEFINITIONS;
}
function getTextureTransformSlotDefinition(slot) {
  const definition = TEXTURE_TRANSFORM_SLOT_DEFINITION_MAP.get(slot);
  if (!definition) {
    throw new Error(`Unknown PBR texture transform slot ${slot}`);
  }
  return definition;
}
function resolveTextureTransform(textureInfo) {
  var _a;
  const extensionTextureTransform = (_a = textureInfo == null ? void 0 : textureInfo["extensions"]) == null ? void 0 : _a["KHR_texture_transform"];
  return {
    offset: (extensionTextureTransform == null ? void 0 : extensionTextureTransform.offset) ? [extensionTextureTransform.offset[0], extensionTextureTransform.offset[1]] : [0, 0],
    rotation: (extensionTextureTransform == null ? void 0 : extensionTextureTransform.rotation) ?? 0,
    scale: (extensionTextureTransform == null ? void 0 : extensionTextureTransform.scale) ? [extensionTextureTransform.scale[0], extensionTextureTransform.scale[1]] : [1, 1]
  };
}
function resolveTextureCoordinateSet(textureInfo) {
  var _a;
  const extensionTextureTransform = (_a = textureInfo == null ? void 0 : textureInfo["extensions"]) == null ? void 0 : _a["KHR_texture_transform"];
  return (extensionTextureTransform == null ? void 0 : extensionTextureTransform["texCoord"]) ?? (textureInfo == null ? void 0 : textureInfo["texCoord"]) ?? 0;
}
function resolveTextureTransformSlot(pointerSegments) {
  return TEXTURE_TRANSFORM_SLOT_DEFINITIONS.find((definition) => definition.pathSegments.length === pointerSegments.length && definition.pathSegments.every((segment, index) => pointerSegments[index] === segment)) || null;
}
function getTextureTransformMatrix(transform) {
  const translationMatrix = new import_core.Matrix3().set(1, 0, 0, 0, 1, 0, transform.offset[0], transform.offset[1], 1);
  const rotationMatrix = new import_core.Matrix3().set(Math.cos(transform.rotation), Math.sin(transform.rotation), 0, -Math.sin(transform.rotation), Math.cos(transform.rotation), 0, 0, 0, 1);
  const scaleMatrix = new import_core.Matrix3().set(transform.scale[0], 0, 0, 0, transform.scale[1], 0, 0, 0, 1);
  return Array.from(translationMatrix.multiplyRight(rotationMatrix).multiplyRight(scaleMatrix));
}
function getTextureTransformDeltaMatrix(baseTransform, currentTransform) {
  const baseMatrix = new import_core.Matrix3(getTextureTransformMatrix(baseTransform));
  const currentMatrix = new import_core.Matrix3(getTextureTransformMatrix(currentTransform));
  const inverseBaseMatrix = new import_core.Matrix3(baseMatrix).invert();
  return Array.from(currentMatrix.multiplyRight(inverseBaseMatrix));
}

// dist/parsers/parse-pbr-material.js
function parsePBRMaterial(device, material, attributes, options) {
  const parsedMaterial = {
    defines: {
      // TODO: Use EXT_sRGB if available (Standard in WebGL 2.0)
      MANUAL_SRGB: true,
      SRGB_FAST_APPROXIMATION: true
    },
    bindings: {},
    uniforms: {
      // TODO: find better values?
      camera: [0, 0, 0],
      // Model should override
      metallicRoughnessValues: [1, 1]
      // Default is 1 and 1
    },
    parameters: {},
    glParameters: {},
    generatedTextures: []
  };
  parsedMaterial.defines["USE_TEX_LOD"] = true;
  const { imageBasedLightingEnvironment } = options;
  if (imageBasedLightingEnvironment) {
    parsedMaterial.bindings.pbr_diffuseEnvSampler = imageBasedLightingEnvironment.diffuseEnvSampler.texture;
    parsedMaterial.bindings.pbr_specularEnvSampler = imageBasedLightingEnvironment.specularEnvSampler.texture;
    parsedMaterial.bindings.pbr_brdfLUT = imageBasedLightingEnvironment.brdfLutTexture.texture;
    parsedMaterial.uniforms.IBLenabled = true;
    parsedMaterial.uniforms.scaleIBLAmbient = [1, 1];
  }
  if (options == null ? void 0 : options.pbrDebug) {
    parsedMaterial.defines["PBR_DEBUG"] = true;
    parsedMaterial.uniforms.scaleDiffBaseMR = [0, 0, 0, 0];
    parsedMaterial.uniforms.scaleFGDSpec = [0, 0, 0, 0];
  }
  if (attributes["NORMAL"])
    parsedMaterial.defines["HAS_NORMALS"] = true;
  if (attributes["TANGENT"] && (options == null ? void 0 : options.useTangents))
    parsedMaterial.defines["HAS_TANGENTS"] = true;
  if (attributes["TEXCOORD_0"])
    parsedMaterial.defines["HAS_UV"] = true;
  if (attributes["TEXCOORD_1"])
    parsedMaterial.defines["HAS_UV_1"] = true;
  if (attributes["JOINTS_0"] && attributes["WEIGHTS_0"])
    parsedMaterial.defines["HAS_SKIN"] = true;
  if (attributes["COLOR_0"])
    parsedMaterial.defines["HAS_COLORS"] = true;
  if (options == null ? void 0 : options.imageBasedLightingEnvironment)
    parsedMaterial.defines["USE_IBL"] = true;
  if (options == null ? void 0 : options.lights)
    parsedMaterial.defines["USE_LIGHTS"] = true;
  if (material) {
    if (options.validateAttributes !== false) {
      warnOnMissingExpectedAttributes(material, attributes);
    }
    parseMaterial(device, material, parsedMaterial, attributes, options.gltf);
  }
  return parsedMaterial;
}
function warnOnMissingExpectedAttributes(material, attributes) {
  var _a;
  const uvDependentTextureSlots = getUvDependentTextureSlots(material, 0);
  if (uvDependentTextureSlots.length > 0 && !attributes["TEXCOORD_0"]) {
    import_core2.log.warn(`glTF material uses ${uvDependentTextureSlots.join(", ")} but primitive is missing TEXCOORD_0; textured shading will sample the default UV coordinates`)();
  }
  const uv1DependentTextureSlots = getUvDependentTextureSlots(material, 1);
  if (uv1DependentTextureSlots.length > 0 && !attributes["TEXCOORD_1"]) {
    import_core2.log.warn(`glTF material uses ${uv1DependentTextureSlots.join(", ")} with TEXCOORD_1 but primitive is missing TEXCOORD_1; those textures will be skipped`)();
  }
  const isUnlitMaterial = Boolean(material.unlit || ((_a = material.extensions) == null ? void 0 : _a.KHR_materials_unlit));
  if (isUnlitMaterial || attributes["NORMAL"]) {
    return;
  }
  const missingNormalReason = material.normalTexture ? "lit PBR shading with normalTexture" : "lit PBR shading";
  import_core2.log.warn(`glTF primitive is missing NORMAL while using ${missingNormalReason}; shading will fall back to geometric normals`)();
}
function getUvDependentTextureSlots(material, textureCoordinateSet) {
  const uvDependentTextureSlots = [];
  for (const slotDefinition of getTextureTransformSlotDefinitions()) {
    const textureInfo = getNestedTextureInfo(material, slotDefinition.pathSegments);
    if (!textureInfo) {
      continue;
    }
    if (resolveTextureCoordinateSet(textureInfo) === textureCoordinateSet) {
      uvDependentTextureSlots.push(slotDefinition.displayName);
    }
  }
  return uvDependentTextureSlots;
}
function getNestedTextureInfo(material, pathSegments) {
  let value = material;
  for (const pathSegment of pathSegments) {
    value = value == null ? void 0 : value[pathSegment];
    if (!value) {
      return null;
    }
  }
  return value;
}
function parseMaterial(device, material, parsedMaterial, attributes, gltf) {
  var _a;
  parsedMaterial.uniforms.unlit = Boolean(material.unlit || ((_a = material.extensions) == null ? void 0 : _a.KHR_materials_unlit));
  if (material.pbrMetallicRoughness) {
    parsePbrMetallicRoughness(device, material.pbrMetallicRoughness, parsedMaterial, attributes, gltf);
  }
  if (material.normalTexture) {
    addTexture(device, material.normalTexture, "pbr_normalSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_NORMALMAP",
        enabledUniformName: "normalMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "normal"
    });
    const { scale = 1 } = material.normalTexture;
    parsedMaterial.uniforms.normalScale = scale;
  }
  if (material.occlusionTexture) {
    addTexture(device, material.occlusionTexture, "pbr_occlusionSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_OCCLUSIONMAP",
        enabledUniformName: "occlusionMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "occlusion"
    });
    const { strength = 1 } = material.occlusionTexture;
    parsedMaterial.uniforms.occlusionStrength = strength;
  }
  parsedMaterial.uniforms.emissiveFactor = material.emissiveFactor || [0, 0, 0];
  if (material.emissiveTexture) {
    addTexture(device, material.emissiveTexture, "pbr_emissiveSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_EMISSIVEMAP",
        enabledUniformName: "emissiveMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "emissive"
    });
  }
  parseMaterialExtensions(device, material.extensions, parsedMaterial, gltf, attributes);
  switch (material.alphaMode || "OPAQUE") {
    case "OPAQUE":
      break;
    case "MASK": {
      const { alphaCutoff = 0.5 } = material;
      parsedMaterial.defines["ALPHA_CUTOFF"] = true;
      parsedMaterial.uniforms.alphaCutoffEnabled = true;
      parsedMaterial.uniforms.alphaCutoff = alphaCutoff;
      break;
    }
    case "BLEND":
      import_core2.log.warn("glTF BLEND alphaMode might not work well because it requires mesh sorting")();
      applyAlphaBlendParameters(parsedMaterial);
      break;
  }
}
function applyAlphaBlendParameters(parsedMaterial) {
  parsedMaterial.parameters.blend = true;
  parsedMaterial.parameters.blendColorOperation = "add";
  parsedMaterial.parameters.blendColorSrcFactor = "src-alpha";
  parsedMaterial.parameters.blendColorDstFactor = "one-minus-src-alpha";
  parsedMaterial.parameters.blendAlphaOperation = "add";
  parsedMaterial.parameters.blendAlphaSrcFactor = "one";
  parsedMaterial.parameters.blendAlphaDstFactor = "one-minus-src-alpha";
  parsedMaterial.glParameters["blend"] = true;
  parsedMaterial.glParameters["blendEquation"] = GLEnum.FUNC_ADD;
  parsedMaterial.glParameters["blendFunc"] = [
    GLEnum.SRC_ALPHA,
    GLEnum.ONE_MINUS_SRC_ALPHA,
    GLEnum.ONE,
    GLEnum.ONE_MINUS_SRC_ALPHA
  ];
}
function applyTransmissionBlendApproximation(parsedMaterial) {
  parsedMaterial.parameters.blend = true;
  parsedMaterial.parameters.depthWriteEnabled = false;
  parsedMaterial.parameters.blendColorOperation = "add";
  parsedMaterial.parameters.blendColorSrcFactor = "one";
  parsedMaterial.parameters.blendColorDstFactor = "one-minus-src-alpha";
  parsedMaterial.parameters.blendAlphaOperation = "add";
  parsedMaterial.parameters.blendAlphaSrcFactor = "one";
  parsedMaterial.parameters.blendAlphaDstFactor = "one-minus-src-alpha";
  parsedMaterial.glParameters["blend"] = true;
  parsedMaterial.glParameters["depthMask"] = false;
  parsedMaterial.glParameters["blendEquation"] = GLEnum.FUNC_ADD;
  parsedMaterial.glParameters["blendFunc"] = [
    GLEnum.ONE,
    GLEnum.ONE_MINUS_SRC_ALPHA,
    GLEnum.ONE,
    GLEnum.ONE_MINUS_SRC_ALPHA
  ];
}
function parsePbrMetallicRoughness(device, pbrMetallicRoughness, parsedMaterial, attributes, gltf) {
  if (pbrMetallicRoughness.baseColorTexture) {
    addTexture(device, pbrMetallicRoughness.baseColorTexture, "pbr_baseColorSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_BASECOLORMAP",
        enabledUniformName: "baseColorMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "baseColor"
    });
  }
  parsedMaterial.uniforms.baseColorFactor = pbrMetallicRoughness.baseColorFactor || [1, 1, 1, 1];
  if (pbrMetallicRoughness.metallicRoughnessTexture) {
    addTexture(device, pbrMetallicRoughness.metallicRoughnessTexture, "pbr_metallicRoughnessSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_METALROUGHNESSMAP",
        enabledUniformName: "metallicRoughnessMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "metallicRoughness"
    });
  }
  const { metallicFactor = 1, roughnessFactor = 1 } = pbrMetallicRoughness;
  parsedMaterial.uniforms.metallicRoughnessValues = [metallicFactor, roughnessFactor];
}
function parseMaterialExtensions(device, extensions, parsedMaterial, gltf, attributes = {}) {
  if (!extensions) {
    return;
  }
  if (hasMaterialExtensionShading(extensions)) {
    parsedMaterial.defines["USE_MATERIAL_EXTENSIONS"] = true;
  }
  parseSpecularExtension(device, extensions.KHR_materials_specular, parsedMaterial, gltf, attributes);
  parseIorExtension(extensions.KHR_materials_ior, parsedMaterial);
  parseTransmissionExtension(device, extensions.KHR_materials_transmission, parsedMaterial, gltf, attributes);
  parseVolumeExtension(device, extensions.KHR_materials_volume, parsedMaterial, gltf, attributes);
  parseClearcoatExtension(device, extensions.KHR_materials_clearcoat, parsedMaterial, gltf, attributes);
  parseSheenExtension(device, extensions.KHR_materials_sheen, parsedMaterial, gltf, attributes);
  parseIridescenceExtension(device, extensions.KHR_materials_iridescence, parsedMaterial, gltf, attributes);
  parseAnisotropyExtension(device, extensions.KHR_materials_anisotropy, parsedMaterial, gltf, attributes);
  parseEmissiveStrengthExtension(extensions.KHR_materials_emissive_strength, parsedMaterial);
}
function hasMaterialExtensionShading(extensions) {
  return Boolean(extensions.KHR_materials_specular || extensions.KHR_materials_ior || extensions.KHR_materials_transmission || extensions.KHR_materials_volume || extensions.KHR_materials_clearcoat || extensions.KHR_materials_sheen || extensions.KHR_materials_iridescence || extensions.KHR_materials_anisotropy);
}
function parseSpecularExtension(device, extension, parsedMaterial, gltf, attributes = {}) {
  if (!extension) {
    return;
  }
  if (extension.specularColorFactor) {
    parsedMaterial.uniforms.specularColorFactor = extension.specularColorFactor;
  }
  if (extension.specularFactor !== void 0) {
    parsedMaterial.uniforms.specularIntensityFactor = extension.specularFactor;
  }
  if (extension.specularColorTexture) {
    addTexture(device, extension.specularColorTexture, "pbr_specularColorSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_SPECULARCOLORMAP",
        enabledUniformName: "specularColorMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "specularColor"
    });
  }
  if (extension.specularTexture) {
    addTexture(device, extension.specularTexture, "pbr_specularIntensitySampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_SPECULARINTENSITYMAP",
        enabledUniformName: "specularIntensityMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "specularIntensity"
    });
  }
}
function parseIorExtension(extension, parsedMaterial) {
  if ((extension == null ? void 0 : extension.ior) !== void 0) {
    parsedMaterial.uniforms.ior = extension.ior;
  }
}
function parseTransmissionExtension(device, extension, parsedMaterial, gltf, attributes = {}) {
  if (!extension) {
    return;
  }
  if (extension.transmissionFactor !== void 0) {
    parsedMaterial.uniforms.transmissionFactor = extension.transmissionFactor;
  }
  if (extension.transmissionTexture) {
    addTexture(device, extension.transmissionTexture, "pbr_transmissionSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_TRANSMISSIONMAP",
        enabledUniformName: "transmissionMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "transmission"
    });
  }
  if ((extension.transmissionFactor ?? 0) > 0 || extension.transmissionTexture) {
    import_core2.log.warn("KHR_materials_transmission uses a premultiplied-alpha blending approximation and may require mesh sorting")();
    applyTransmissionBlendApproximation(parsedMaterial);
  }
}
function parseVolumeExtension(device, extension, parsedMaterial, gltf, attributes = {}) {
  if (!extension) {
    return;
  }
  if (extension.thicknessFactor !== void 0) {
    parsedMaterial.uniforms.thicknessFactor = extension.thicknessFactor;
  }
  if (extension.thicknessTexture) {
    addTexture(device, extension.thicknessTexture, "pbr_thicknessSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_THICKNESSMAP"
      },
      gltf,
      attributes,
      textureTransformSlot: "thickness"
    });
  }
  if (extension.attenuationDistance !== void 0) {
    parsedMaterial.uniforms.attenuationDistance = extension.attenuationDistance;
  }
  if (extension.attenuationColor) {
    parsedMaterial.uniforms.attenuationColor = extension.attenuationColor;
  }
}
function parseClearcoatExtension(device, extension, parsedMaterial, gltf, attributes = {}) {
  if (!extension) {
    return;
  }
  if (extension.clearcoatFactor !== void 0) {
    parsedMaterial.uniforms.clearcoatFactor = extension.clearcoatFactor;
  }
  if (extension.clearcoatRoughnessFactor !== void 0) {
    parsedMaterial.uniforms.clearcoatRoughnessFactor = extension.clearcoatRoughnessFactor;
  }
  if (extension.clearcoatTexture) {
    addTexture(device, extension.clearcoatTexture, "pbr_clearcoatSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_CLEARCOATMAP",
        enabledUniformName: "clearcoatMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "clearcoat"
    });
  }
  if (extension.clearcoatRoughnessTexture) {
    addTexture(device, extension.clearcoatRoughnessTexture, "pbr_clearcoatRoughnessSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_CLEARCOATROUGHNESSMAP",
        enabledUniformName: "clearcoatRoughnessMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "clearcoatRoughness"
    });
  }
  if (extension.clearcoatNormalTexture) {
    addTexture(device, extension.clearcoatNormalTexture, "pbr_clearcoatNormalSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_CLEARCOATNORMALMAP"
      },
      gltf,
      attributes,
      textureTransformSlot: "clearcoatNormal"
    });
  }
}
function parseSheenExtension(device, extension, parsedMaterial, gltf, attributes = {}) {
  if (!extension) {
    return;
  }
  if (extension.sheenColorFactor) {
    parsedMaterial.uniforms.sheenColorFactor = extension.sheenColorFactor;
  }
  if (extension.sheenRoughnessFactor !== void 0) {
    parsedMaterial.uniforms.sheenRoughnessFactor = extension.sheenRoughnessFactor;
  }
  if (extension.sheenColorTexture) {
    addTexture(device, extension.sheenColorTexture, "pbr_sheenColorSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_SHEENCOLORMAP",
        enabledUniformName: "sheenColorMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "sheenColor"
    });
  }
  if (extension.sheenRoughnessTexture) {
    addTexture(device, extension.sheenRoughnessTexture, "pbr_sheenRoughnessSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_SHEENROUGHNESSMAP",
        enabledUniformName: "sheenRoughnessMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "sheenRoughness"
    });
  }
}
function parseIridescenceExtension(device, extension, parsedMaterial, gltf, attributes = {}) {
  if (!extension) {
    return;
  }
  if (extension.iridescenceFactor !== void 0) {
    parsedMaterial.uniforms.iridescenceFactor = extension.iridescenceFactor;
  }
  if (extension.iridescenceIor !== void 0) {
    parsedMaterial.uniforms.iridescenceIor = extension.iridescenceIor;
  }
  if (extension.iridescenceThicknessMinimum !== void 0 || extension.iridescenceThicknessMaximum !== void 0) {
    parsedMaterial.uniforms.iridescenceThicknessRange = [
      extension.iridescenceThicknessMinimum ?? 100,
      extension.iridescenceThicknessMaximum ?? 400
    ];
  }
  if (extension.iridescenceTexture) {
    addTexture(device, extension.iridescenceTexture, "pbr_iridescenceSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_IRIDESCENCEMAP",
        enabledUniformName: "iridescenceMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "iridescence"
    });
  }
  if (extension.iridescenceThicknessTexture) {
    addTexture(device, extension.iridescenceThicknessTexture, "pbr_iridescenceThicknessSampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_IRIDESCENCETHICKNESSMAP"
      },
      gltf,
      attributes,
      textureTransformSlot: "iridescenceThickness"
    });
  }
}
function parseAnisotropyExtension(device, extension, parsedMaterial, gltf, attributes = {}) {
  if (!extension) {
    return;
  }
  if (extension.anisotropyStrength !== void 0) {
    parsedMaterial.uniforms.anisotropyStrength = extension.anisotropyStrength;
  }
  if (extension.anisotropyRotation !== void 0) {
    parsedMaterial.uniforms.anisotropyRotation = extension.anisotropyRotation;
  }
  if (extension.anisotropyTexture) {
    addTexture(device, extension.anisotropyTexture, "pbr_anisotropySampler", parsedMaterial, {
      featureOptions: {
        define: "HAS_ANISOTROPYMAP",
        enabledUniformName: "anisotropyMapEnabled"
      },
      gltf,
      attributes,
      textureTransformSlot: "anisotropy"
    });
  }
}
function parseEmissiveStrengthExtension(extension, parsedMaterial) {
  if ((extension == null ? void 0 : extension.emissiveStrength) !== void 0) {
    parsedMaterial.uniforms.emissiveStrength = extension.emissiveStrength;
  }
}
function addTexture(device, gltfTexture, uniformName, parsedMaterial, textureParseOptions = {}) {
  var _a, _b, _c;
  const { featureOptions = {}, gltf, attributes = {}, textureTransformSlot } = textureParseOptions;
  const { define, enabledUniformName } = featureOptions;
  const textureCoordinateSet = resolveTextureCoordinateSet(gltfTexture);
  if (textureCoordinateSet > 1) {
    import_core2.log.warn(`Skipping ${String(uniformName)} because ${textureCoordinateSet} is not supported; only TEXCOORD_0 and TEXCOORD_1 are currently available`)();
    return;
  }
  if (textureCoordinateSet === 1 && !attributes["TEXCOORD_1"]) {
    import_core2.log.warn(`Skipping ${String(uniformName)} because it requires TEXCOORD_1 but the primitive does not provide TEXCOORD_1`)();
    return;
  }
  const resolvedTextureInfo = resolveTextureInfo(gltfTexture, gltf);
  const image = (_b = (_a = resolvedTextureInfo.texture) == null ? void 0 : _a.source) == null ? void 0 : _b.image;
  if (!image) {
    import_core2.log.warn(`Skipping unresolved glTF texture for ${String(uniformName)}`)();
    return;
  }
  const gltfSampler = {
    wrapS: 10497,
    // default REPEAT S (U) wrapping mode.
    wrapT: 10497,
    // default REPEAT T (V) wrapping mode.
    minFilter: 9729,
    // default LINEAR filtering
    magFilter: 9729,
    // default LINEAR filtering
    ...(_c = resolvedTextureInfo == null ? void 0 : resolvedTextureInfo.texture) == null ? void 0 : _c.sampler
  };
  const baseOptions = {
    id: resolvedTextureInfo.uniformName || resolvedTextureInfo.id,
    sampler: convertSampler(gltfSampler)
  };
  let texture;
  if (image.compressed) {
    texture = createCompressedTexture(device, image, baseOptions);
  } else {
    const { width, height } = device.getExternalImageSize(image);
    texture = device.createTexture({
      ...baseOptions,
      width,
      height,
      data: image
    });
  }
  parsedMaterial.bindings[uniformName] = texture;
  if (define)
    parsedMaterial.defines[define] = true;
  if (enabledUniformName) {
    parsedMaterial.uniforms[enabledUniformName] = true;
  }
  if (textureTransformSlot) {
    const textureTransformSlotDefinition = getTextureTransformSlotDefinition(textureTransformSlot);
    parsedMaterial.uniforms[textureTransformSlotDefinition.uvSetUniform] = textureCoordinateSet;
    parsedMaterial.uniforms[textureTransformSlotDefinition.uvTransformUniform] = getTextureTransformMatrix(resolveTextureTransform(gltfTexture));
  }
  parsedMaterial.generatedTextures.push(texture);
}
function resolveTextureInfo(gltfTexture, gltf) {
  if (gltfTexture.texture || gltfTexture.index === void 0 || !(gltf == null ? void 0 : gltf.textures)) {
    return gltfTexture;
  }
  const resolvedTextureEntry = gltf.textures[gltfTexture.index];
  if (!resolvedTextureEntry) {
    return gltfTexture;
  }
  if ("texture" in resolvedTextureEntry && resolvedTextureEntry.texture) {
    return {
      ...resolvedTextureEntry,
      ...gltfTexture,
      texture: resolvedTextureEntry.texture
    };
  }
  if (!("source" in resolvedTextureEntry)) {
    return gltfTexture;
  }
  return {
    ...gltfTexture,
    texture: resolvedTextureEntry
  };
}
function createCompressedTextureFallback(device, baseOptions) {
  return device.createTexture({
    ...baseOptions,
    format: "rgba8unorm",
    width: 1,
    height: 1,
    mipLevels: 1
  });
}
function resolveCompressedTextureFormat(level) {
  return level.textureFormat;
}
function getMaxCompressedMipLevels(baseWidth, baseHeight, format) {
  const { blockWidth = 1, blockHeight = 1 } = import_core2.textureFormatDecoder.getInfo(format);
  let count = 1;
  for (let i = 1; ; i++) {
    const w = Math.max(1, baseWidth >> i);
    const h = Math.max(1, baseHeight >> i);
    if (w < blockWidth || h < blockHeight)
      break;
    count++;
  }
  return count;
}
function createCompressedTexture(device, image, baseOptions) {
  var _a, _b;
  let levels;
  if (Array.isArray(image.data) && ((_a = image.data[0]) == null ? void 0 : _a.data)) {
    levels = image.data;
  } else if ("mipmaps" in image && Array.isArray(image.mipmaps)) {
    levels = image.mipmaps;
  } else {
    levels = [];
  }
  if (levels.length === 0 || !((_b = levels[0]) == null ? void 0 : _b.data)) {
    import_core2.log.warn("createCompressedTexture: compressed image has no valid mip levels, creating fallback")();
    return createCompressedTextureFallback(device, baseOptions);
  }
  const baseLevel = levels[0];
  const baseWidth = baseLevel.width ?? image.width ?? 0;
  const baseHeight = baseLevel.height ?? image.height ?? 0;
  if (baseWidth <= 0 || baseHeight <= 0) {
    import_core2.log.warn("createCompressedTexture: base level has invalid dimensions, creating fallback")();
    return createCompressedTextureFallback(device, baseOptions);
  }
  const format = resolveCompressedTextureFormat(baseLevel);
  if (!format) {
    import_core2.log.warn("createCompressedTexture: compressed image has no textureFormat, creating fallback")();
    return createCompressedTextureFallback(device, baseOptions);
  }
  const maxMipLevels = getMaxCompressedMipLevels(baseWidth, baseHeight, format);
  const levelLimit = Math.min(levels.length, maxMipLevels);
  let validLevelCount = 1;
  for (let i = 1; i < levelLimit; i++) {
    const level = levels[i];
    if (!level.data || level.width <= 0 || level.height <= 0) {
      import_core2.log.warn(`createCompressedTexture: mip level ${i} has invalid data/dimensions, truncating`)();
      break;
    }
    const levelFormat = resolveCompressedTextureFormat(level);
    if (levelFormat && levelFormat !== format) {
      import_core2.log.warn(`createCompressedTexture: mip level ${i} format '${levelFormat}' differs from base '${format}', truncating`)();
      break;
    }
    const expectedW = Math.max(1, baseWidth >> i);
    const expectedH = Math.max(1, baseHeight >> i);
    if (level.width !== expectedW || level.height !== expectedH) {
      import_core2.log.warn(`createCompressedTexture: mip level ${i} dimensions ${level.width}x${level.height} don't match expected ${expectedW}x${expectedH}, truncating`)();
      break;
    }
    validLevelCount++;
  }
  const texture = device.createTexture({
    ...baseOptions,
    format,
    usage: import_core2.Texture.TEXTURE | import_core2.Texture.COPY_DST,
    width: baseWidth,
    height: baseHeight,
    mipLevels: validLevelCount,
    data: baseLevel.data
  });
  for (let i = 1; i < validLevelCount; i++) {
    texture.writeData(levels[i].data, {
      width: levels[i].width,
      height: levels[i].height,
      mipLevel: i
    });
  }
  return texture;
}

// dist/parsers/parse-gltf-lights.js
var import_core3 = require("@math.gl/core");
var import_shadertools = require("@luma.gl/shadertools");
function parseGLTFLights(gltf, options = {}) {
  var _a, _b, _c, _d;
  const lightDefs = (
    // `postProcessGLTF()` moves KHR_lights_punctual into `gltf.lights`.
    gltf.lights || ((_b = (_a = gltf.extensions) == null ? void 0 : _a["KHR_lights_punctual"]) == null ? void 0 : _b["lights"])
  );
  if (!lightDefs || !Array.isArray(lightDefs) || lightDefs.length === 0) {
    return [];
  }
  const lights = [];
  const parentNodeById = createParentNodeMap(gltf.nodes || []);
  const worldMatrixByNodeId = /* @__PURE__ */ new Map();
  for (const node of gltf.nodes || []) {
    const lightIndex = node.light ?? ((_d = (_c = node.extensions) == null ? void 0 : _c.KHR_lights_punctual) == null ? void 0 : _d.light);
    if (typeof lightIndex !== "number") {
      continue;
    }
    const gltfLight = lightDefs[lightIndex];
    if (!gltfLight) {
      continue;
    }
    const color = normalizeGLTFLightColor(gltfLight.color || [1, 1, 1], options.useByteColors ?? true);
    const intensity = gltfLight.intensity ?? 1;
    const range = gltfLight.range;
    const worldMatrix = getNodeWorldMatrix(node, parentNodeById, worldMatrixByNodeId);
    switch (gltfLight.type) {
      case "directional":
        lights.push(parseDirectionalLight(worldMatrix, color, intensity));
        break;
      case "point":
        lights.push(parsePointLight(worldMatrix, color, intensity, range));
        break;
      case "spot":
        lights.push(parseSpotLight(worldMatrix, color, intensity, range, gltfLight.spot));
        break;
      default:
        break;
    }
  }
  return lights;
}
function normalizeGLTFLightColor(color, useByteColors) {
  if (useByteColors) {
    return color.map((component) => component * 255);
  }
  return (0, import_shadertools.normalizeByteColor3)(color, false);
}
function parsePointLight(worldMatrix, color, intensity, range) {
  const position = getLightPosition(worldMatrix);
  let attenuation = [1, 0, 0];
  if (range !== void 0 && range > 0) {
    attenuation = [1, 0, 1 / (range * range)];
  }
  return {
    type: "point",
    position,
    color,
    intensity,
    attenuation
  };
}
function parseDirectionalLight(worldMatrix, color, intensity) {
  const direction = getLightDirection(worldMatrix);
  return {
    type: "directional",
    direction,
    color,
    intensity
  };
}
function parseSpotLight(worldMatrix, color, intensity, range, spot = {}) {
  const position = getLightPosition(worldMatrix);
  const direction = getLightDirection(worldMatrix);
  let attenuation = [1, 0, 0];
  if (range !== void 0 && range > 0) {
    attenuation = [1, 0, 1 / (range * range)];
  }
  return {
    type: "spot",
    position,
    direction,
    color,
    intensity,
    attenuation,
    innerConeAngle: spot.innerConeAngle ?? 0,
    outerConeAngle: spot.outerConeAngle ?? Math.PI / 4
  };
}
function createParentNodeMap(nodes) {
  const parentNodeById = /* @__PURE__ */ new Map();
  for (const node of nodes) {
    for (const childNode of node.children || []) {
      parentNodeById.set(childNode.id, node);
    }
  }
  return parentNodeById;
}
function getNodeWorldMatrix(node, parentNodeById, worldMatrixByNodeId) {
  const cachedWorldMatrix = worldMatrixByNodeId.get(node.id);
  if (cachedWorldMatrix) {
    return cachedWorldMatrix;
  }
  const localMatrix = getNodeLocalMatrix(node);
  const parentNode = parentNodeById.get(node.id);
  const worldMatrix = parentNode ? new import_core3.Matrix4(getNodeWorldMatrix(parentNode, parentNodeById, worldMatrixByNodeId)).multiplyRight(localMatrix) : localMatrix;
  worldMatrixByNodeId.set(node.id, worldMatrix);
  return worldMatrix;
}
function getNodeLocalMatrix(node) {
  if (node.matrix) {
    return new import_core3.Matrix4(node.matrix);
  }
  const matrix = new import_core3.Matrix4();
  if (node.translation) {
    matrix.translate(node.translation);
  }
  if (node.rotation) {
    matrix.multiplyRight(new import_core3.Matrix4().fromQuaternion(node.rotation));
  }
  if (node.scale) {
    matrix.scale(node.scale);
  }
  return matrix;
}
function getLightPosition(worldMatrix) {
  return worldMatrix.transformAsPoint([0, 0, 0]);
}
function getLightDirection(worldMatrix) {
  return worldMatrix.transformDirection([0, 0, -1]);
}

// dist/parsers/parse-gltf.js
var import_engine4 = require("@luma.gl/engine");
var import_shadertools3 = require("@luma.gl/shadertools");

// dist/webgl-to-webgpu/convert-webgl-topology.js
function convertGLDrawModeToTopology(drawMode) {
  switch (drawMode) {
    case GLEnum.POINTS:
      return "point-list";
    case GLEnum.LINES:
      return "line-list";
    case GLEnum.LINE_STRIP:
      return "line-strip";
    case GLEnum.TRIANGLES:
      return "triangle-list";
    case GLEnum.TRIANGLE_STRIP:
      return "triangle-strip";
    default:
      throw new Error(String(drawMode));
  }
}

// dist/gltf/create-gltf-model.js
var import_core4 = require("@luma.gl/core");
var import_engine2 = require("@luma.gl/engine");
var import_shadertools2 = require("@luma.gl/shadertools");
var import_engine3 = require("@luma.gl/engine");
var SHADER = (
  /* WGSL */
  `
struct VertexInputs {
  @location(0) positions: vec3f,
#ifdef HAS_NORMALS
  @location(1) normals: vec3f,
#endif
#ifdef HAS_TANGENTS
  @location(2) TANGENT: vec4f,
#endif
#ifdef HAS_UV
  @location(3) texCoords: vec2f,
#endif
#ifdef HAS_UV_1
  @location(4) texCoords1: vec2f,
#endif
#ifdef HAS_SKIN
  @location(5) JOINTS_0: vec4u,
  @location(6) WEIGHTS_0: vec4f,
#endif
};

struct FragmentInputs {
  @builtin(position) position: vec4f,
  @location(0) pbrPosition: vec3f,
  @location(1) pbrUV0: vec2f,
  @location(2) pbrUV1: vec2f,
  @location(3) pbrNormal: vec3f,
#ifdef HAS_TANGENTS
  @location(4) pbrTangent: vec4f,
#endif
};

@vertex
fn vertexMain(inputs: VertexInputs) -> FragmentInputs {
  var outputs: FragmentInputs;
  var position = vec4f(inputs.positions, 1.0);
  var normal = vec3f(0.0, 0.0, 1.0);
  var tangent = vec4f(1.0, 0.0, 0.0, 1.0);
  var uv0 = vec2f(0.0, 0.0);
  var uv1 = vec2f(0.0, 0.0);

#ifdef HAS_NORMALS
  normal = inputs.normals;
#endif
#ifdef HAS_UV
  uv0 = inputs.texCoords;
#endif
#ifdef HAS_UV_1
  uv1 = inputs.texCoords1;
#endif
#ifdef HAS_TANGENTS
  tangent = inputs.TANGENT;
#endif
#ifdef HAS_SKIN
  let skinMatrix = getSkinMatrix(inputs.WEIGHTS_0, inputs.JOINTS_0);
  position = skinMatrix * position;
  normal = normalize((skinMatrix * vec4f(normal, 0.0)).xyz);
#ifdef HAS_TANGENTS
  tangent = vec4f(normalize((skinMatrix * vec4f(tangent.xyz, 0.0)).xyz), tangent.w);
#endif
#endif

  let worldPosition = pbrProjection.modelMatrix * position;

#ifdef HAS_NORMALS
  normal = normalize((pbrProjection.normalMatrix * vec4f(normal, 0.0)).xyz);
#endif
#ifdef HAS_TANGENTS
  let worldTangent = normalize((pbrProjection.modelMatrix * vec4f(tangent.xyz, 0.0)).xyz);
  outputs.pbrTangent = vec4f(worldTangent, tangent.w);
#endif

  outputs.position = pbrProjection.modelViewProjectionMatrix * position;
  outputs.pbrPosition = worldPosition.xyz / worldPosition.w;
  outputs.pbrUV0 = uv0;
  outputs.pbrUV1 = uv1;
  outputs.pbrNormal = normal;
  return outputs;
}

@fragment
fn fragmentMain(inputs: FragmentInputs) -> @location(0) vec4f {
  fragmentInputs.pbr_vPosition = inputs.pbrPosition;
  fragmentInputs.pbr_vUV0 = inputs.pbrUV0;
  fragmentInputs.pbr_vUV1 = inputs.pbrUV1;
  fragmentInputs.pbr_vNormal = inputs.pbrNormal;
#ifdef HAS_TANGENTS
  let tangent = normalize(inputs.pbrTangent.xyz);
  let bitangent = normalize(cross(inputs.pbrNormal, tangent)) * inputs.pbrTangent.w;
  fragmentInputs.pbr_vTBN = mat3x3f(tangent, bitangent, inputs.pbrNormal);
#endif
  return pbr_filterColor(vec4f(1.0));
}
`
);
var vs = (
  /* glsl */
  `#version 300 es

  // in vec4 POSITION;
  in vec4 positions;

  #ifdef HAS_NORMALS
    // in vec4 NORMAL;
    in vec4 normals;
  #endif

  #ifdef HAS_TANGENTS
    in vec4 TANGENT;
  #endif

  #ifdef HAS_UV
    // in vec2 TEXCOORD_0;
    in vec2 texCoords;
  #endif

  #ifdef HAS_UV_1
    in vec2 texCoords1;
  #endif

  #ifdef HAS_SKIN
    in uvec4 JOINTS_0;
    in vec4 WEIGHTS_0;
  #endif

  void main(void) {
    vec4 _NORMAL = vec4(0.);
    vec4 _TANGENT = vec4(0.);
    vec2 _TEXCOORD_0 = vec2(0.);
    vec2 _TEXCOORD_1 = vec2(0.);

    #ifdef HAS_NORMALS
      _NORMAL = normals;
    #endif

    #ifdef HAS_TANGENTS
      _TANGENT = TANGENT;
    #endif

    #ifdef HAS_UV
      _TEXCOORD_0 = texCoords;
    #endif

    #ifdef HAS_UV_1
      _TEXCOORD_1 = texCoords1;
    #endif

    vec4 pos = positions;

    #ifdef HAS_SKIN
      mat4 skinMat = getSkinMatrix(WEIGHTS_0, JOINTS_0);
      pos = skinMat * pos;
      _NORMAL = skinMat * _NORMAL;
      _TANGENT = vec4((skinMat * vec4(_TANGENT.xyz, 0.)).xyz, _TANGENT.w);
    #endif

    pbr_setPositionNormalTangentUV(pos, _NORMAL, _TANGENT, _TEXCOORD_0, _TEXCOORD_1);
    gl_Position = pbrProjection.modelViewProjectionMatrix * pos;
  }
`
);
var fs = (
  /* glsl */
  `#version 300 es
  out vec4 fragmentColor;

  void main(void) {
    vec3 pos = pbr_vPosition;
    fragmentColor = pbr_filterColor(vec4(1.0));
  }
`
);
function createGLTFMaterial(device, options) {
  const materialFactory = options.materialFactory || new import_engine3.MaterialFactory(device, { modules: [import_shadertools2.pbrMaterial] });
  const pbrMaterialProps = { ...options.parsedPPBRMaterial.uniforms };
  delete pbrMaterialProps.camera;
  const materialBindings = Object.fromEntries(Object.entries({
    ...pbrMaterialProps,
    ...options.parsedPPBRMaterial.bindings
  }).filter(([name, value]) => materialFactory.ownsBinding(name) && isMaterialBindingResource(value)));
  const material = materialFactory.createMaterial({
    id: options.id,
    bindings: materialBindings
  });
  material.setProps({ pbrMaterial: pbrMaterialProps });
  return material;
}
function createGLTFModel(device, options) {
  const { id, geometry, parsedPPBRMaterial, vertexCount, modelOptions = {} } = options;
  import_core4.log.info(4, "createGLTFModel defines: ", parsedPPBRMaterial.defines)();
  const managedResources = [];
  const parameters = {
    depthWriteEnabled: true,
    depthCompare: "less",
    depthFormat: "depth24plus",
    cullMode: "back"
  };
  const modelProps = {
    id,
    source: SHADER,
    vs,
    fs,
    geometry,
    topology: geometry.topology,
    vertexCount,
    modules: [import_shadertools2.pbrMaterial, import_shadertools2.skin],
    ...modelOptions,
    defines: { ...parsedPPBRMaterial.defines, ...modelOptions.defines },
    parameters: { ...parameters, ...parsedPPBRMaterial.parameters, ...modelOptions.parameters }
  };
  const material = options.material || createGLTFMaterial(device, {
    id: id ? `${id}-material` : void 0,
    parsedPPBRMaterial
  });
  modelProps.material = material;
  const model = new import_engine3.Model(device, modelProps);
  const sceneShaderInputValues = {
    ...parsedPPBRMaterial.uniforms,
    ...modelOptions.uniforms,
    ...parsedPPBRMaterial.bindings,
    ...modelOptions.bindings
  };
  const sceneShaderInputProps = getSceneShaderInputProps(model.shaderInputs.getModules(), material, sceneShaderInputValues);
  model.shaderInputs.setProps(sceneShaderInputProps);
  return new import_engine3.ModelNode({ managedResources, model });
}
function isMaterialBindingResource(value) {
  return value instanceof import_core4.Buffer || value instanceof import_engine2.DynamicTexture || value instanceof import_core4.Sampler || value instanceof import_core4.Texture || value instanceof import_core4.TextureView;
}
function getSceneShaderInputProps(modules, material, shaderInputValues) {
  const propertyToModuleNameMap = /* @__PURE__ */ new Map();
  for (const module2 of modules) {
    for (const uniformName of Object.keys(module2.uniformTypes || {})) {
      propertyToModuleNameMap.set(uniformName, module2.name);
    }
    for (const binding of module2.bindingLayout || []) {
      propertyToModuleNameMap.set(binding.name, module2.name);
    }
  }
  const sceneShaderInputProps = {};
  for (const [propertyName, value] of Object.entries(shaderInputValues)) {
    if (value === void 0) {
      continue;
    }
    const moduleName = propertyToModuleNameMap.get(propertyName);
    if (!moduleName || material.ownsModule(moduleName)) {
      continue;
    }
    sceneShaderInputProps[moduleName] ||= {};
    sceneShaderInputProps[moduleName][propertyName] = value;
  }
  return sceneShaderInputProps;
}

// dist/parsers/parse-gltf.js
var defaultOptions = {
  modelOptions: {},
  pbrDebug: false,
  imageBasedLightingEnvironment: void 0,
  lights: true,
  useTangents: false,
  useByteColors: true
};
function parseGLTF(device, gltf, options = {}) {
  const combinedOptions = { ...defaultOptions, ...options };
  const materialFactory = new import_engine4.MaterialFactory(device, { modules: [import_shadertools3.pbrMaterial] });
  const materials = (gltf.materials || []).map((gltfMaterial, materialIndex) => createGLTFMaterial(device, {
    id: getGLTFMaterialId(gltfMaterial, materialIndex),
    parsedPPBRMaterial: parsePBRMaterial(device, gltfMaterial, {}, {
      ...combinedOptions,
      gltf,
      validateAttributes: false
    }),
    materialFactory
  }));
  const gltfMaterialIdToMaterialMap = /* @__PURE__ */ new Map();
  (gltf.materials || []).forEach((gltfMaterial, materialIndex) => {
    gltfMaterialIdToMaterialMap.set(gltfMaterial.id, materials[materialIndex]);
  });
  const gltfMeshIdToNodeMap = /* @__PURE__ */ new Map();
  gltf.meshes.forEach((gltfMesh, idx) => {
    const newMesh = createNodeForGLTFMesh(device, gltfMesh, gltf, gltfMaterialIdToMaterialMap, combinedOptions);
    gltfMeshIdToNodeMap.set(gltfMesh.id, newMesh);
  });
  const gltfNodeIndexToNodeMap = /* @__PURE__ */ new Map();
  const gltfNodeIdToNodeMap = /* @__PURE__ */ new Map();
  gltf.nodes.forEach((gltfNode, idx) => {
    const newNode = createNodeForGLTFNode(device, gltfNode, combinedOptions);
    gltfNodeIndexToNodeMap.set(idx, newNode);
    gltfNodeIdToNodeMap.set(gltfNode.id, newNode);
  });
  gltf.nodes.forEach((gltfNode, idx) => {
    gltfNodeIndexToNodeMap.get(idx).add((gltfNode.children ?? []).map(({ id }) => {
      const child = gltfNodeIdToNodeMap.get(id);
      if (!child)
        throw new Error(`Cannot find child ${id} of node ${idx}`);
      return child;
    }));
    if (gltfNode.mesh) {
      const mesh = gltfMeshIdToNodeMap.get(gltfNode.mesh.id);
      if (!mesh) {
        throw new Error(`Cannot find mesh child ${gltfNode.mesh.id} of node ${idx}`);
      }
      gltfNodeIndexToNodeMap.get(idx).add(mesh);
    }
  });
  const scenes = gltf.scenes.map((gltfScene) => {
    const children = (gltfScene.nodes || []).map(({ id }) => {
      const child = gltfNodeIdToNodeMap.get(id);
      if (!child)
        throw new Error(`Cannot find child ${id} of scene ${gltfScene.name || gltfScene.id}`);
      return child;
    });
    return new import_engine4.GroupNode({
      id: gltfScene.name || gltfScene.id,
      children
    });
  });
  return { scenes, materials, gltfMeshIdToNodeMap, gltfNodeIdToNodeMap, gltfNodeIndexToNodeMap };
}
function createNodeForGLTFNode(device, gltfNode, options) {
  return new import_engine4.GroupNode({
    id: gltfNode.name || gltfNode.id,
    children: [],
    matrix: gltfNode.matrix,
    position: gltfNode.translation,
    rotation: gltfNode.rotation,
    scale: gltfNode.scale
  });
}
function createNodeForGLTFMesh(device, gltfMesh, gltf, gltfMaterialIdToMaterialMap, options) {
  const gltfPrimitives = gltfMesh.primitives || [];
  const primitives = gltfPrimitives.map((gltfPrimitive, i) => createNodeForGLTFPrimitive({
    device,
    gltfPrimitive,
    primitiveIndex: i,
    gltfMesh,
    gltf,
    gltfMaterialIdToMaterialMap,
    options
  }));
  const mesh = new import_engine4.GroupNode({
    id: gltfMesh.name || gltfMesh.id,
    children: primitives
  });
  return mesh;
}
function createNodeForGLTFPrimitive({ device, gltfPrimitive, primitiveIndex, gltfMesh, gltf, gltfMaterialIdToMaterialMap, options }) {
  const id = gltfPrimitive.name || `${gltfMesh.name || gltfMesh.id}-primitive-${primitiveIndex}`;
  const topology = convertGLDrawModeToTopology(gltfPrimitive.mode ?? 4);
  const vertexCount = gltfPrimitive.indices ? gltfPrimitive.indices.count : getVertexCount(gltfPrimitive.attributes);
  const geometry = createGeometry(id, gltfPrimitive, topology);
  const parsedPPBRMaterial = parsePBRMaterial(device, gltfPrimitive.material, geometry.attributes, {
    ...options,
    gltf
  });
  const modelNode = createGLTFModel(device, {
    id,
    geometry,
    material: gltfPrimitive.material ? gltfMaterialIdToMaterialMap.get(gltfPrimitive.material.id) || null : null,
    parsedPPBRMaterial,
    modelOptions: options.modelOptions,
    vertexCount
  });
  modelNode.bounds = [gltfPrimitive.attributes.POSITION.min, gltfPrimitive.attributes.POSITION.max];
  return modelNode;
}
function getVertexCount(attributes) {
  let vertexCount = Infinity;
  for (const attribute of Object.values(attributes)) {
    if (attribute) {
      const { value, size, components } = attribute;
      const attributeSize = size ?? components;
      if ((value == null ? void 0 : value.length) !== void 0 && attributeSize >= 1) {
        vertexCount = Math.min(vertexCount, value.length / attributeSize);
      }
    }
  }
  if (!Number.isFinite(vertexCount)) {
    throw new Error("Could not determine vertex count from attributes");
  }
  return vertexCount;
}
function createGeometry(id, gltfPrimitive, topology) {
  var _a;
  const attributes = {};
  for (const [attributeName, attribute] of Object.entries(gltfPrimitive.attributes)) {
    const { components, size, value, normalized } = attribute;
    attributes[attributeName] = { size: size ?? components, value, normalized };
  }
  return new import_engine4.Geometry({
    id,
    topology,
    indices: (_a = gltfPrimitive.indices) == null ? void 0 : _a.value,
    attributes
  });
}
function getGLTFMaterialId(gltfMaterial, materialIndex) {
  return gltfMaterial.name || gltfMaterial.id || `material-${materialIndex}`;
}

// dist/gltf/gltf-animator.js
var import_core7 = require("@luma.gl/core");

// dist/gltf/animations/interpolate.js
var import_core5 = require("@luma.gl/core");
var import_core6 = require("@math.gl/core");
function updateTargetPath(target, path, newValue) {
  switch (path) {
    case "translation":
      return target.setPosition(newValue).updateMatrix();
    case "rotation":
      return target.setRotation(newValue).updateMatrix();
    case "scale":
      return target.setScale(newValue).updateMatrix();
    default:
      import_core5.log.warn(`Bad animation path ${path}`)();
      return null;
  }
}
function interpolate(time, { input, interpolation, output }, target, path) {
  const value = evaluateSampler(time, { input, interpolation, output }, path);
  if (value) {
    updateTargetPath(target, path, value);
  }
}
function evaluateSampler(time, { input, interpolation, output }, path) {
  const maxTime = input[input.length - 1];
  if (!Number.isFinite(maxTime) || maxTime <= 0) {
    return output[0] || null;
  }
  const animationTime = time % maxTime;
  const nextIndex = input.findIndex((t) => t >= animationTime);
  if (nextIndex < 0) {
    return output[output.length - 1] || null;
  }
  const previousIndex = Math.max(0, nextIndex - 1);
  const previousTime = input[previousIndex];
  const nextTime = input[nextIndex];
  switch (interpolation) {
    case "STEP":
      return output[previousIndex];
    case "LINEAR":
      if (nextTime > previousTime) {
        const ratio = (animationTime - previousTime) / (nextTime - previousTime);
        return linearInterpolate(path, output[previousIndex], output[nextIndex], ratio);
      }
      return output[previousIndex] || null;
    case "CUBICSPLINE":
      if (nextTime > previousTime) {
        const ratio = (animationTime - previousTime) / (nextTime - previousTime);
        const tDiff = nextTime - previousTime;
        const p0 = output[3 * previousIndex + 1];
        const outTangent0 = output[3 * previousIndex + 2];
        const inTangent1 = output[3 * nextIndex + 0];
        const p1 = output[3 * nextIndex + 1];
        return cubicsplineInterpolate({ p0, outTangent0, inTangent1, p1, tDiff, ratio });
      }
      return output[3 * previousIndex + 1] || null;
    default:
      import_core5.log.warn(`Interpolation ${interpolation} not supported`)();
      return null;
  }
}
function linearInterpolate(path, start, stop, ratio) {
  if (path === "rotation") {
    return new import_core6.Quaternion().slerp({ start, target: stop, ratio });
  }
  const newVal = [];
  for (let i = 0; i < start.length; i++) {
    newVal[i] = ratio * stop[i] + (1 - ratio) * start[i];
  }
  return newVal;
}
function cubicsplineInterpolate({ p0, outTangent0, inTangent1, p1, tDiff, ratio: t }) {
  const newVal = [];
  for (let i = 0; i < p0.length; i++) {
    const m0 = outTangent0[i] * tDiff;
    const m1 = inTangent1[i] * tDiff;
    newVal[i] = (2 * Math.pow(t, 3) - 3 * Math.pow(t, 2) + 1) * p0[i] + (Math.pow(t, 3) - 2 * Math.pow(t, 2) + t) * m0 + (-2 * Math.pow(t, 3) + 3 * Math.pow(t, 2)) * p1[i] + (Math.pow(t, 3) - Math.pow(t, 2)) * m1;
  }
  return newVal;
}

// dist/gltf/gltf-animator.js
var GLTFSingleAnimator = class {
  /** Animation definition being played. */
  animation;
  /** Target scenegraph lookup table. */
  gltfNodeIdToNodeMap;
  /** Materials aligned with the source glTF materials array. */
  materials;
  /** Playback start time in seconds. */
  startTime = 0;
  /** Whether playback is currently enabled. */
  playing = true;
  /** Playback speed multiplier. */
  speed = 1;
  /** Mutable runtime texture-transform state for animated material slots. */
  materialTextureTransformState = /* @__PURE__ */ new Map();
  /** Creates a single-animation controller. */
  constructor(props) {
    this.animation = props.animation;
    this.gltfNodeIdToNodeMap = props.gltfNodeIdToNodeMap;
    this.materials = props.materials || [];
    this.animation.name ||= "unnamed";
    Object.assign(this, props);
    if (this.animation.channels.some((channel) => channel.type !== "node") && !this.materials.length) {
      throw new Error(`Animation ${this.animation.name} targets materials, but GLTFAnimator was created without a materials array`);
    }
  }
  /** Advances the animation to the supplied wall-clock time in milliseconds. */
  setTime(timeMs) {
    if (!this.playing) {
      return;
    }
    const absTime = timeMs / 1e3;
    const time = (absTime - this.startTime) * this.speed;
    this.animation.channels.forEach((channel) => {
      if (channel.type === "node") {
        const { sampler, targetNodeId, path } = channel;
        const targetNode = this.gltfNodeIdToNodeMap.get(targetNodeId);
        if (!targetNode) {
          throw new Error(`Cannot find animation target node ${targetNodeId}`);
        }
        interpolate(time, sampler, targetNode, path);
        return;
      }
      const material = this.materials[channel.targetMaterialIndex];
      if (!material) {
        throw new Error(`Cannot find animation target material ${channel.targetMaterialIndex} for ${channel.pointer}`);
      }
      const value = evaluateSampler(time, channel.sampler);
      if (value) {
        if (channel.type === "material") {
          applyMaterialAnimationValue(material, channel, value);
        } else {
          applyTextureTransformAnimationValue(material, channel, value, this.materialTextureTransformState);
        }
      }
    });
  }
};
var GLTFAnimator = class {
  /** Individual animation controllers. */
  animations;
  /** Creates an animator for the supplied glTF scenegraph. */
  constructor(props) {
    this.animations = props.animations.map((animation, index) => {
      const name = animation.name || `Animation-${index}`;
      return new GLTFSingleAnimator({
        gltfNodeIdToNodeMap: props.gltfNodeIdToNodeMap,
        materials: props.materials,
        animation: { name, channels: animation.channels }
      });
    });
  }
  /** @deprecated Use .setTime(). Will be removed (deck.gl is using this) */
  animate(time) {
    import_core7.log.warn("GLTFAnimator#animate is deprecated. Use GLTFAnimator#setTime instead")();
    this.setTime(time);
  }
  /** Advances every animation to the supplied wall-clock time in milliseconds. */
  setTime(time) {
    this.animations.forEach((animation) => animation.setTime(time));
  }
  /** Returns the per-animation controllers managed by this animator. */
  getAnimations() {
    return this.animations;
  }
};
function applyMaterialAnimationValue(material, channel, value) {
  const pbrMaterial3 = channel.component !== void 0 ? {
    [channel.property]: updateMaterialArrayComponent(getCurrentMaterialValue(material, channel.property), channel.component, value[0])
  } : {
    [channel.property]: value.length === 1 ? value[0] : value
  };
  material.setProps({ pbrMaterial: pbrMaterial3 });
}
function getCurrentMaterialValue(material, property) {
  var _a;
  const uniformValues = material.shaderInputs.getUniformValues();
  const currentValue = (_a = uniformValues["pbrMaterial"]) == null ? void 0 : _a[property];
  return Array.isArray(currentValue) ? [...currentValue] : [];
}
function updateMaterialArrayComponent(currentValue, component, nextValue) {
  const updatedValue = [...currentValue];
  updatedValue[component] = nextValue;
  return updatedValue;
}
function applyTextureTransformAnimationValue(material, channel, value, materialTextureTransformState) {
  const slotDefinition = getTextureTransformSlotDefinition(channel.textureSlot);
  const currentTransform = getCurrentTextureTransform(materialTextureTransformState, material, channel);
  switch (channel.path) {
    case "offset":
      if (channel.component !== void 0) {
        currentTransform.offset[channel.component] = value[0];
      } else {
        currentTransform.offset = [value[0], value[1]];
      }
      break;
    case "rotation":
      currentTransform.rotation = value[0];
      break;
    case "scale":
      if (channel.component !== void 0) {
        currentTransform.scale[channel.component] = value[0];
      } else {
        currentTransform.scale = [value[0], value[1]];
      }
      break;
  }
  material.setProps({
    pbrMaterial: {
      [slotDefinition.uvTransformUniform]: getTextureTransformDeltaMatrix(channel.baseTransform, currentTransform)
    }
  });
}
function getCurrentTextureTransform(materialTextureTransformState, material, channel) {
  const materialState = materialTextureTransformState.get(material) || {};
  let textureTransformState = materialState[channel.textureSlot];
  if (!textureTransformState) {
    textureTransformState = {
      offset: [...channel.baseTransform.offset],
      rotation: channel.baseTransform.rotation,
      scale: [...channel.baseTransform.scale]
    };
    materialState[channel.textureSlot] = textureTransformState;
    materialTextureTransformState.set(material, materialState);
  }
  return textureTransformState;
}

// dist/parsers/parse-gltf-animations.js
var import_core8 = require("@luma.gl/core");

// dist/gltf/gltf-extension-support.js
var UNKNOWN_EXTENSION_SUPPORT = {
  supportLevel: "none",
  comment: "Not currently listed in the luma.gl glTF extension support registry."
};
var GLTF_EXTENSION_SUPPORT_REGISTRY = {
  KHR_draco_mesh_compression: {
    supportLevel: "built-in",
    comment: "Decoded by loaders.gl before luma.gl builds the scenegraph."
  },
  EXT_meshopt_compression: {
    supportLevel: "built-in",
    comment: "Meshopt-compressed primitives are decoded during load."
  },
  KHR_mesh_quantization: {
    supportLevel: "built-in",
    comment: "Quantized accessors are unpacked before geometry creation."
  },
  KHR_lights_punctual: {
    supportLevel: "built-in",
    comment: "Parsed into luma.gl Light objects."
  },
  KHR_materials_unlit: {
    supportLevel: "built-in",
    comment: "Unlit materials bypass the default lighting path."
  },
  KHR_materials_emissive_strength: {
    supportLevel: "built-in",
    comment: "Applied by the stock PBR shader."
  },
  KHR_texture_basisu: {
    supportLevel: "built-in",
    comment: "BasisU / KTX2 textures pass through when the device supports them."
  },
  KHR_texture_transform: {
    supportLevel: "built-in",
    comment: "UV transforms are applied during load."
  },
  EXT_texture_webp: {
    supportLevel: "loader-only",
    comment: "Texture source is resolved during load; final support depends on browser and device decode support."
  },
  EXT_texture_avif: {
    supportLevel: "loader-only",
    comment: "Texture source is resolved during load; final support depends on browser and device decode support."
  },
  KHR_materials_specular: {
    supportLevel: "built-in",
    comment: "The stock shader now applies specular factors and textures to the dielectric F0 term."
  },
  KHR_materials_ior: {
    supportLevel: "built-in",
    comment: "The stock shader now drives dielectric reflectance from the glTF IOR value."
  },
  KHR_materials_transmission: {
    supportLevel: "built-in",
    comment: "The stock shader now applies transmission to the base layer and exposes transparency through alpha, without a scene-color refraction buffer."
  },
  KHR_materials_volume: {
    supportLevel: "built-in",
    comment: "Thickness and attenuation now tint transmitted light in the stock shader."
  },
  KHR_materials_clearcoat: {
    supportLevel: "built-in",
    comment: "The stock shader now adds a secondary clearcoat specular lobe."
  },
  KHR_materials_sheen: {
    supportLevel: "built-in",
    comment: "The stock shader now adds a sheen lobe for cloth-like materials."
  },
  KHR_materials_iridescence: {
    supportLevel: "built-in",
    comment: "The stock shader now tints specular response with a view-dependent thin-film iridescence approximation."
  },
  KHR_materials_anisotropy: {
    supportLevel: "built-in",
    comment: "The stock shader now shapes highlights and IBL response with an anisotropy-direction approximation."
  },
  KHR_materials_pbrSpecularGlossiness: {
    supportLevel: "loader-only",
    comment: "Extension data can be loaded, but it is not translated into the default metallic-roughness material path."
  },
  KHR_materials_variants: {
    supportLevel: "loader-only",
    comment: "Variant metadata can be loaded, but applications must choose and apply variants."
  },
  EXT_mesh_gpu_instancing: {
    supportLevel: "none",
    comment: "GPU instancing data is not yet converted into luma.gl instanced draw setup."
  },
  KHR_node_visibility: {
    supportLevel: "none",
    comment: "Node-visibility animations and toggles are not mapped onto runtime scenegraph state."
  },
  KHR_animation_pointer: {
    supportLevel: "parsed-and-wired",
    comment: "Selected node TRS, material factor, and KHR_texture_transform offset/rotation/scale pointers are wired to runtime updates; unsupported targets are skipped."
  },
  KHR_materials_diffuse_transmission: {
    supportLevel: "none",
    comment: "Diffuse-transmission shading is not implemented in the stock PBR shader."
  },
  KHR_materials_dispersion: {
    supportLevel: "none",
    comment: "Chromatic dispersion is not implemented in the stock PBR shader."
  },
  KHR_materials_volume_scatter: {
    supportLevel: "none",
    comment: "Volume scattering is not implemented in the stock PBR shader."
  },
  KHR_xmp: {
    supportLevel: "none",
    comment: "Metadata payloads remain in the loaded glTF, but luma.gl does not interpret them."
  },
  KHR_xmp_json_ld: {
    supportLevel: "none",
    comment: "Metadata is preserved in the glTF, but luma.gl does not interpret it."
  },
  EXT_lights_image_based: {
    supportLevel: "none",
    comment: "Use loadPBREnvironment() or custom environment setup instead."
  },
  EXT_texture_video: {
    supportLevel: "none",
    comment: "Video textures are not created automatically by the stock pipeline."
  },
  MSFT_lod: {
    supportLevel: "none",
    comment: "Level-of-detail switching is not implemented in the stock scenegraph loader."
  }
};
function getGLTFExtensionSupport(gltf) {
  const extensionNames = Array.from(collectGLTFExtensionNames(gltf)).sort();
  const extensionSupportEntries = extensionNames.map((extensionName) => {
    const extensionSupportDefinition = GLTF_EXTENSION_SUPPORT_REGISTRY[extensionName] || UNKNOWN_EXTENSION_SUPPORT;
    return [
      extensionName,
      {
        extensionName,
        supported: extensionSupportDefinition.supportLevel === "built-in" || extensionSupportDefinition.supportLevel === "parsed-and-wired",
        supportLevel: extensionSupportDefinition.supportLevel,
        comment: extensionSupportDefinition.comment
      }
    ];
  });
  return new Map(extensionSupportEntries);
}
function getRegisteredGLTFExtensionSupport(extensionName) {
  return GLTF_EXTENSION_SUPPORT_REGISTRY[extensionName] || null;
}
function collectGLTFExtensionNames(gltf) {
  var _a;
  const gltfWithRemovedExtensions = gltf;
  const extensionNames = /* @__PURE__ */ new Set();
  addExtensionNames(extensionNames, gltf.extensionsUsed);
  addExtensionNames(extensionNames, gltf.extensionsRequired);
  addExtensionNames(extensionNames, gltfWithRemovedExtensions.extensionsRemoved);
  addExtensionNames(extensionNames, Object.keys(gltf.extensions || {}));
  if (((_a = gltfWithRemovedExtensions.lights) == null ? void 0 : _a.length) || (gltf.nodes || []).some((node) => "light" in node)) {
    extensionNames.add("KHR_lights_punctual");
  }
  if ((gltf.materials || []).some((material) => {
    var _a2;
    const gltfMaterial = material;
    return gltfMaterial.unlit || ((_a2 = gltfMaterial.extensions) == null ? void 0 : _a2.KHR_materials_unlit);
  })) {
    extensionNames.add("KHR_materials_unlit");
  }
  return extensionNames;
}
function addExtensionNames(extensionNames, newExtensionNames = []) {
  for (const extensionName of newExtensionNames) {
    extensionNames.add(extensionName);
  }
}

// dist/webgl-to-webgpu/convert-webgl-attribute.js
var ATTRIBUTE_TYPE_TO_COMPONENTS = {
  SCALAR: 1,
  VEC2: 2,
  VEC3: 3,
  VEC4: 4,
  MAT2: 4,
  MAT3: 9,
  MAT4: 16
};
var ATTRIBUTE_COMPONENT_TYPE_TO_ARRAY = {
  5120: Int8Array,
  5121: Uint8Array,
  5122: Int16Array,
  5123: Uint16Array,
  5125: Uint32Array,
  5126: Float32Array
};
function accessorToTypedArray(accessor) {
  var _a;
  const ArrayType = ATTRIBUTE_COMPONENT_TYPE_TO_ARRAY[accessor.componentType];
  const components = ATTRIBUTE_TYPE_TO_COMPONENTS[accessor.type];
  const length = components * accessor.count;
  const { buffer, byteOffset = 0 } = ((_a = accessor.bufferView) == null ? void 0 : _a.data) ?? {};
  const typedArray = new ArrayType(buffer, byteOffset + (accessor.byteOffset || 0), length);
  return { typedArray, components };
}

// dist/parsers/parse-gltf-animations.js
function parseGLTFAnimations(gltf) {
  const gltfAnimations = gltf.animations || [];
  const accessorCache1D = /* @__PURE__ */ new Map();
  const accessorCache2D = /* @__PURE__ */ new Map();
  return gltfAnimations.flatMap((animation, index) => {
    const name = animation.name || `Animation-${index}`;
    const samplerCache = /* @__PURE__ */ new Map();
    const channels = animation.channels.flatMap(({ sampler, target }) => {
      let parsedSampler = samplerCache.get(sampler);
      if (!parsedSampler) {
        const gltfSampler = animation.samplers[sampler];
        if (!gltfSampler) {
          throw new Error(`Cannot find animation sampler ${sampler}`);
        }
        const { input, interpolation = "LINEAR", output } = gltfSampler;
        parsedSampler = {
          input: accessorToJsArray1D(gltf.accessors[input], accessorCache1D),
          interpolation,
          output: accessorToJsArray2D(gltf.accessors[output], accessorCache2D)
        };
        samplerCache.set(sampler, parsedSampler);
      }
      const parsedChannel = parseAnimationChannel(gltf, target, parsedSampler);
      return parsedChannel ? [parsedChannel] : [];
    });
    return channels.length ? [{ name, channels }] : [];
  });
}
function parseAnimationChannel(gltf, target, sampler) {
  if (target.path === "pointer") {
    return parseAnimationPointerChannel(gltf, target, sampler);
  }
  const path = getNodeAnimationPath(target.path);
  if (!path) {
    return null;
  }
  const targetNode = gltf.nodes[target.node ?? 0];
  if (!targetNode) {
    throw new Error(`Cannot find animation target ${target.node}`);
  }
  return {
    type: "node",
    sampler,
    targetNodeId: targetNode.id,
    path
  };
}
function parseAnimationPointerChannel(gltf, target, sampler) {
  var _a, _b;
  const pointer = (_b = (_a = target.extensions) == null ? void 0 : _a["KHR_animation_pointer"]) == null ? void 0 : _b.pointer;
  if (typeof pointer !== "string" || !pointer.startsWith("/")) {
    import_core8.log.warn("KHR_animation_pointer channel is missing a valid JSON pointer and will be skipped")();
    return null;
  }
  const pointerSegments = splitJsonPointer(pointer);
  switch (pointerSegments[0]) {
    case "nodes":
      return parseNodePointerAnimationChannel(gltf, pointerSegments, sampler, pointer);
    case "materials":
      return parseMaterialPointerAnimationChannel(gltf, pointerSegments, sampler, pointer);
    default:
      warnUnsupportedAnimationPointer(pointer, `top-level target "${pointerSegments[0]}" has no runtime animation mapping`);
      return null;
  }
}
function parseNodePointerAnimationChannel(gltf, pointerSegments, sampler, pointer) {
  if (pointerSegments.length !== 3) {
    warnUnsupportedAnimationPointer(pointer, "node pointers must use /nodes/{index}/{translation|rotation|scale|weights}");
    return null;
  }
  const nodeIndex = Number(pointerSegments[1]);
  const targetNode = gltf.nodes[nodeIndex];
  if (!Number.isInteger(nodeIndex) || !targetNode) {
    import_core8.log.warn(`KHR_animation_pointer target ${pointer} references a missing node and will be skipped`)();
    return null;
  }
  const path = getNodeAnimationPath(pointerSegments[2]);
  if (!path) {
    warnUnsupportedAnimationPointer(pointer, `node property "${pointerSegments[2]}" has no runtime animation mapping`);
    return null;
  }
  if (path === "weights") {
    import_core8.log.warn(`KHR_animation_pointer target ${pointer} will be skipped because morph weights are not implemented in GLTFAnimator`)();
    return null;
  }
  return {
    type: "node",
    sampler,
    targetNodeId: targetNode.id,
    path
  };
}
function parseMaterialPointerAnimationChannel(gltf, pointerSegments, sampler, pointer) {
  if (pointerSegments.length < 3) {
    warnUnsupportedAnimationPointer(pointer, "material pointers must include a material index and target property path");
    return null;
  }
  const materialIndex = Number(pointerSegments[1]);
  const material = gltf.materials[materialIndex];
  if (!Number.isInteger(materialIndex) || !material) {
    import_core8.log.warn(`KHR_animation_pointer target ${pointer} references a missing material and will be skipped`)();
    return null;
  }
  const materialTarget = resolveMaterialAnimationTarget(material, pointerSegments.slice(2));
  if ("reason" in materialTarget) {
    warnUnsupportedAnimationPointer(pointer, materialTarget.reason);
    return null;
  }
  return {
    sampler,
    pointer,
    targetMaterialIndex: materialIndex,
    ...materialTarget
  };
}
function getNodeAnimationPath(path) {
  switch (path) {
    case "translation":
    case "rotation":
    case "scale":
    case "weights":
      return path;
    default:
      return null;
  }
}
function resolveMaterialAnimationTarget(material, pointerSegments) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n, _o, _p, _q, _r;
  const textureTransformTarget = resolveTextureTransformAnimationTarget(material, pointerSegments);
  if (!("reason" in textureTransformTarget)) {
    return textureTransformTarget;
  }
  if (textureTransformTarget.reason !== "not-a-texture-transform-target") {
    return textureTransformTarget;
  }
  const pointerPath = pointerSegments.join("/");
  switch (pointerPath) {
    case "pbrMetallicRoughness/baseColorFactor":
      return material["pbrMetallicRoughness"] ? { type: "material", property: "baseColorFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "pbrMetallicRoughness/metallicFactor":
      return material["pbrMetallicRoughness"] ? { type: "material", property: "metallicRoughnessValues", component: 0 } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "pbrMetallicRoughness/roughnessFactor":
      return material["pbrMetallicRoughness"] ? { type: "material", property: "metallicRoughnessValues", component: 1 } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "normalTexture/scale":
      return material["normalTexture"] ? { type: "material", property: "normalScale" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "occlusionTexture/strength":
      return material["occlusionTexture"] ? { type: "material", property: "occlusionStrength" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "emissiveFactor":
      return { type: "material", property: "emissiveFactor" };
    case "alphaCutoff":
      return { type: "material", property: "alphaCutoff" };
    case "extensions/KHR_materials_specular/specularFactor":
      return ((_a = material["extensions"]) == null ? void 0 : _a["KHR_materials_specular"]) ? { type: "material", property: "specularIntensityFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_specular/specularColorFactor":
      return ((_b = material["extensions"]) == null ? void 0 : _b["KHR_materials_specular"]) ? { type: "material", property: "specularColorFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_ior/ior":
      return ((_c = material["extensions"]) == null ? void 0 : _c["KHR_materials_ior"]) ? { type: "material", property: "ior" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_transmission/transmissionFactor":
      return ((_d = material["extensions"]) == null ? void 0 : _d["KHR_materials_transmission"]) ? { type: "material", property: "transmissionFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_volume/thicknessFactor":
      return ((_e = material["extensions"]) == null ? void 0 : _e["KHR_materials_volume"]) ? { type: "material", property: "thicknessFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_volume/attenuationDistance":
      return ((_f = material["extensions"]) == null ? void 0 : _f["KHR_materials_volume"]) ? { type: "material", property: "attenuationDistance" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_volume/attenuationColor":
      return ((_g = material["extensions"]) == null ? void 0 : _g["KHR_materials_volume"]) ? { type: "material", property: "attenuationColor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_clearcoat/clearcoatFactor":
      return ((_h = material["extensions"]) == null ? void 0 : _h["KHR_materials_clearcoat"]) ? { type: "material", property: "clearcoatFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_clearcoat/clearcoatRoughnessFactor":
      return ((_i = material["extensions"]) == null ? void 0 : _i["KHR_materials_clearcoat"]) ? { type: "material", property: "clearcoatRoughnessFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_sheen/sheenColorFactor":
      return ((_j = material["extensions"]) == null ? void 0 : _j["KHR_materials_sheen"]) ? { type: "material", property: "sheenColorFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_sheen/sheenRoughnessFactor":
      return ((_k = material["extensions"]) == null ? void 0 : _k["KHR_materials_sheen"]) ? { type: "material", property: "sheenRoughnessFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_iridescence/iridescenceFactor":
      return ((_l = material["extensions"]) == null ? void 0 : _l["KHR_materials_iridescence"]) ? { type: "material", property: "iridescenceFactor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_iridescence/iridescenceIor":
      return ((_m = material["extensions"]) == null ? void 0 : _m["KHR_materials_iridescence"]) ? { type: "material", property: "iridescenceIor" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_iridescence/iridescenceThicknessMinimum":
      return ((_n = material["extensions"]) == null ? void 0 : _n["KHR_materials_iridescence"]) ? { type: "material", property: "iridescenceThicknessRange", component: 0 } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_iridescence/iridescenceThicknessMaximum":
      return ((_o = material["extensions"]) == null ? void 0 : _o["KHR_materials_iridescence"]) ? { type: "material", property: "iridescenceThicknessRange", component: 1 } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_anisotropy/anisotropyStrength":
      return ((_p = material["extensions"]) == null ? void 0 : _p["KHR_materials_anisotropy"]) ? { type: "material", property: "anisotropyStrength" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_anisotropy/anisotropyRotation":
      return ((_q = material["extensions"]) == null ? void 0 : _q["KHR_materials_anisotropy"]) ? { type: "material", property: "anisotropyRotation" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    case "extensions/KHR_materials_emissive_strength/emissiveStrength":
      return ((_r = material["extensions"]) == null ? void 0 : _r["KHR_materials_emissive_strength"]) ? { type: "material", property: "emissiveStrength" } : { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
    default:
      return { reason: getUnsupportedMaterialPointerReason(pointerSegments) };
  }
}
function resolveTextureTransformAnimationTarget(material, pointerSegments) {
  const extensionIndex = pointerSegments.lastIndexOf("extensions");
  if (extensionIndex < 0 || pointerSegments[extensionIndex + 1] !== "KHR_texture_transform" || extensionIndex < 1) {
    return { reason: "not-a-texture-transform-target" };
  }
  const textureSlotDefinition = resolveTextureTransformSlot(pointerSegments.slice(0, extensionIndex));
  if (!textureSlotDefinition) {
    return {
      reason: getUnsupportedTextureTransformSlotReason(pointerSegments.slice(0, extensionIndex))
    };
  }
  const textureInfo = getNestedMaterialValue(material, textureSlotDefinition.pathSegments);
  if (!textureInfo) {
    return {
      reason: `texture-transform target "${pointerSegments.slice(0, extensionIndex).join("/")}" does not exist on the referenced material`
    };
  }
  const textureTransformPath = pointerSegments[extensionIndex + 2];
  if (textureTransformPath === "texCoord") {
    return {
      reason: "animated KHR_texture_transform.texCoord is unsupported because texCoord selection is structural, not a runtime float/vector update"
    };
  }
  if (textureTransformPath !== "offset" && textureTransformPath !== "rotation" && textureTransformPath !== "scale") {
    return {
      reason: `KHR_texture_transform property "${textureTransformPath}" is not animatable; supported properties are offset, rotation, and scale`
    };
  }
  const componentSegment = pointerSegments[extensionIndex + 3];
  if (pointerSegments.length > extensionIndex + 4) {
    return {
      reason: `KHR_texture_transform.${textureTransformPath} does not support nested property paths`
    };
  }
  let component;
  if (componentSegment !== void 0) {
    component = Number(componentSegment);
    if (textureTransformPath === "rotation") {
      return {
        reason: "KHR_texture_transform.rotation does not support component indices"
      };
    }
    if (!Number.isInteger(component) || component < 0 || component > 1) {
      return {
        reason: `KHR_texture_transform.${textureTransformPath} component index "${componentSegment}" is invalid; only 0 and 1 are supported`
      };
    }
  }
  return {
    type: "textureTransform",
    textureSlot: textureSlotDefinition.slot,
    path: textureTransformPath,
    component,
    baseTransform: resolveTextureTransform(textureInfo)
  };
}
function getNestedMaterialValue(material, pathSegments) {
  let value = material;
  for (const pathSegment of pathSegments) {
    value = value == null ? void 0 : value[pathSegment];
    if (!value) {
      return null;
    }
  }
  return value;
}
function splitJsonPointer(pointer) {
  return pointer.slice(1).split("/").map((segment) => segment.replace(/~1/g, "/").replace(/~0/g, "~"));
}
function getUnsupportedMaterialPointerReason(pointerSegments) {
  const extensionName = getPointerExtensionName(pointerSegments);
  if (extensionName) {
    const extensionSupport = getRegisteredGLTFExtensionSupport(extensionName);
    if ((extensionSupport == null ? void 0 : extensionSupport.supportLevel) === "none") {
      return `${extensionName} is referenced by this pointer, but ${extensionSupport.comment.charAt(0).toLowerCase()}${extensionSupport.comment.slice(1)}`;
    }
  }
  return `no runtime target exists for material property "${pointerSegments.join("/")}"`;
}
function getUnsupportedTextureTransformSlotReason(pointerSegments) {
  const extensionName = getPointerExtensionName(pointerSegments);
  if (extensionName) {
    const extensionSupport = getRegisteredGLTFExtensionSupport(extensionName);
    if ((extensionSupport == null ? void 0 : extensionSupport.supportLevel) === "none") {
      return `${extensionName} is referenced by this pointer, but ${extensionSupport.comment.charAt(0).toLowerCase()}${extensionSupport.comment.slice(1)}`;
    }
  }
  return `texture-transform target "${pointerSegments.join("/")}" has no runtime texture-slot mapping`;
}
function getPointerExtensionName(pointerSegments) {
  const extensionIndex = pointerSegments.indexOf("extensions");
  const extensionName = pointerSegments[extensionIndex + 1];
  return extensionIndex >= 0 && extensionName ? extensionName : null;
}
function warnUnsupportedAnimationPointer(pointer, reason) {
  import_core8.log.warn(`KHR_animation_pointer target ${pointer} will be skipped because ${reason}`)();
}
function accessorToJsArray1D(accessor, accessorCache) {
  if (accessorCache.has(accessor)) {
    return accessorCache.get(accessor);
  }
  const { typedArray: array, components } = accessorToTypedArray(accessor);
  assert(components === 1, "accessorToJsArray1D must have exactly 1 component");
  const result = Array.from(array);
  accessorCache.set(accessor, result);
  return result;
}
function accessorToJsArray2D(accessor, accessorCache) {
  if (accessorCache.has(accessor)) {
    return accessorCache.get(accessor);
  }
  const { typedArray: array, components } = accessorToTypedArray(accessor);
  assert(components >= 1, "accessorToJsArray2D must have at least 1 component");
  const result = [];
  for (let i = 0; i < array.length; i += components) {
    result.push(Array.from(array.slice(i, i + components)));
  }
  accessorCache.set(accessor, result);
  return result;
}
function assert(condition, message) {
  if (!condition) {
    throw new Error(message);
  }
}

// dist/gltf/create-scenegraph-from-gltf.js
function createScenegraphsFromGLTF(device, gltf, options) {
  const { scenes, materials, gltfMeshIdToNodeMap, gltfNodeIdToNodeMap, gltfNodeIndexToNodeMap } = parseGLTF(device, gltf, options);
  const animations = parseGLTFAnimations(gltf);
  const animator = new GLTFAnimator({ animations, gltfNodeIdToNodeMap, materials });
  const lights = parseGLTFLights(gltf, { useByteColors: (options == null ? void 0 : options.useByteColors) ?? true });
  const extensionSupport = getGLTFExtensionSupport(gltf);
  const sceneBounds = scenes.map((scene) => getScenegraphBounds(scene.getBounds()));
  const modelBounds = getCombinedScenegraphBounds(sceneBounds);
  return {
    scenes,
    materials,
    animator,
    lights,
    extensionSupport,
    sceneBounds,
    modelBounds,
    gltfMeshIdToNodeMap,
    gltfNodeIdToNodeMap,
    gltfNodeIndexToNodeMap,
    gltf
  };
}
function getScenegraphBounds(bounds) {
  if (!bounds) {
    return {
      bounds: null,
      center: [0, 0, 0],
      size: [0, 0, 0],
      radius: 0.5,
      recommendedOrbitDistance: 1
    };
  }
  const normalizedBounds = [
    [bounds[0][0], bounds[0][1], bounds[0][2]],
    [bounds[1][0], bounds[1][1], bounds[1][2]]
  ];
  const size = [
    normalizedBounds[1][0] - normalizedBounds[0][0],
    normalizedBounds[1][1] - normalizedBounds[0][1],
    normalizedBounds[1][2] - normalizedBounds[0][2]
  ];
  const center = [
    normalizedBounds[0][0] + size[0] * 0.5,
    normalizedBounds[0][1] + size[1] * 0.5,
    normalizedBounds[0][2] + size[2] * 0.5
  ];
  const maxHalfExtent = Math.max(size[0], size[1], size[2]) * 0.5;
  const radius = Math.max(0.5 * Math.hypot(size[0], size[1], size[2]), 1e-3);
  return {
    bounds: normalizedBounds,
    center,
    size,
    radius,
    recommendedOrbitDistance: Math.max(Math.max(maxHalfExtent, 1e-3) / Math.tan(Math.PI / 6) * 1.15, radius * 1.1)
  };
}
function getCombinedScenegraphBounds(sceneBounds) {
  let combinedBounds = null;
  for (const sceneBoundInfo of sceneBounds) {
    if (!sceneBoundInfo.bounds) {
      continue;
    }
    if (!combinedBounds) {
      combinedBounds = [
        [...sceneBoundInfo.bounds[0]],
        [...sceneBoundInfo.bounds[1]]
      ];
      continue;
    }
    for (let axis = 0; axis < 3; axis++) {
      combinedBounds[0][axis] = Math.min(combinedBounds[0][axis], sceneBoundInfo.bounds[0][axis]);
      combinedBounds[1][axis] = Math.max(combinedBounds[1][axis], sceneBoundInfo.bounds[1][axis]);
    }
  }
  return getScenegraphBounds(combinedBounds);
}
