// @loaders.gl/loader-utils/dist/lib/sources/data-source.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { mergeOptions } from "../option-utils/merge-options.js";
import { resolvePath } from "../path-utils/file-aliases.js";
/** base class of all data sources */
export class DataSource {
    static defaultOptions = {
        core: {
            type: 'auto',
            attributions: [],
            loadOptions: {},
            loaders: []
        }
    };
    optionsType;
    options;
    data;
    url;
    /** The actual load options, if calling a loaders.gl loader */
    loadOptions;
    /** A resolved fetch function extracted from loadOptions prop */
    fetch;
    _needsRefresh = true;
    constructor(data, options, defaultOptions) {
        if (defaultOptions) {
            // @ts-expect-error Typescript gets confused
            this.options = mergeOptions({ ...defaultOptions, core: DataSource.defaultOptions }, options);
        }
        else {
            // @ts-expect-error
            this.options = { ...options };
        }
        this.data = data;
        this.url = typeof data === 'string' ? resolvePath(data) : '';
        this.loadOptions = normalizeDirectLoaderOptions(this.options.core?.loadOptions);
        this.fetch = getFetchFunction(this.loadOptions);
    }
    setProps(options) {
        this.options = Object.assign(this.options, options);
        // TODO - add a shallow compare to avoid setting refresh if no change?
        this.setNeedsRefresh();
    }
    /** Mark this data source as needing a refresh (redraw) */
    setNeedsRefresh() {
        this._needsRefresh = true;
    }
    /**
     * Does this data source need refreshing?
     * @note The specifics of the refresh mechanism depends on type of data source
     */
    getNeedsRefresh(clear = true) {
        const needsRefresh = this._needsRefresh;
        if (clear) {
            this._needsRefresh = false;
        }
        return needsRefresh;
    }
}
/**
 * Gets the current fetch function from options
 * @todo - move to loader-utils module
 * @todo - use in core module counterpart
 * @param options
 * @param context
 */
export function getFetchFunction(options) {
    const fetchFunction = options?.core?.fetch;
    // options.fetch can be a function
    if (fetchFunction && typeof fetchFunction === 'function') {
        return (url, fetchOptions) => fetchFunction(url, fetchOptions);
    }
    // options.fetch can be an options object, use global fetch with those options
    const fetchOptions = options?.fetch;
    if (fetchOptions && typeof fetchOptions !== 'function') {
        return (url) => fetch(url, fetchOptions);
    }
    // else return the global fetch function
    return (url) => fetch(url);
}
function normalizeDirectLoaderOptions(options) {
    const loadOptions = { ...options };
    if (options?.core) {
        loadOptions.core = { ...options.core };
    }
    const topLevelBaseUri = typeof loadOptions.baseUri === 'string' ? loadOptions.baseUri : undefined;
    const topLevelBaseUrl = typeof loadOptions.baseUrl === 'string' ? loadOptions.baseUrl : undefined;
    if (topLevelBaseUri !== undefined || topLevelBaseUrl !== undefined) {
        loadOptions.core ||= {};
        if (loadOptions.core.baseUrl === undefined) {
            loadOptions.core.baseUrl = topLevelBaseUrl ?? topLevelBaseUri;
        }
        delete loadOptions.baseUri;
        delete loadOptions.baseUrl;
    }
    return loadOptions;
}
